{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_25022020/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_25022020/sim_samples/\"\n",
    "hg38_fa = \"/home/avaraby1/genomes/human/hg38/hg38_p8.fa\"\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "tmapcols=[\"ref_gene_id\",\"ref_id\",\"class_code\",\"qry_gene_id\",\"qry_id\",\"num_exons\",\"FPKM\",\"TPM\",\"cov\",\"len\",\"major_iso_id\",\"ref_match_len\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first need to cluster the loci\n",
    "cluster_cmd = \"gffread \"\"\"+base_dir_data+\"\"\"ALL.sub.gtf --cluster-only -T -o \"\"\"+base_dir_data+\"\"\"ALL.merged.gtf\"\"\"\n",
    "subprocess.call(cluster_cmd,shell=True)\n",
    "\n",
    "# next run the comparison of ALL with CHESS\n",
    "compare_cmd = \"gffcompare -o \"\"\"+base_dir_data+\"\"\"chess2ALL -r \"\"\"+base_dir_data+\"\"\"chess2.2_assembly.gff \"\"\"+base_dir_data+\"\"\"ALL.merged.gtf\"\"\"\n",
    "subprocess.call(compare_cmd,shell=True)\n",
    "\n",
    "# next we can isolate the intergenic noise\n",
    "# get all the real loci codes\n",
    "realgtf_cmd = \"\"\"awk -F '\\t|\\\"' 'FNR==NR{ids[$1]++;next} {if($10 in ids) print $0}' <(awk -v pat='=' -F '\\t' '$3~pat {print $5}' \"\"\"+base_dir_data+\"\"\"chess2ALL.ALL.merged.gtf.tmap) \"\"\"+base_dir_data+\"\"\"ALL.merged.gtf > \"\"\"+base_dir_data+\"\"\"real.gtf\"\"\"\n",
    "subprocess.call(realgtf_cmd,shell=True)\n",
    "\n",
    "# next get a list of all loci that contain real transcripts\n",
    "realloci_cmd = \"\"\"awk -F '\\t|locus \"' '$3==\"transcript\" {print $10}' \"\"\"+base_dir_data+\"\"\"real.gtf | cut -d'\"' -f1 | sort | uniq > \"\"\"+base_dir_data+\"\"\"real.locs\"\"\"\n",
    "subprocess.call(realloci_cmd,shell=True)\n",
    "\n",
    "# extract all transcript IDs that are in the set of real loci\n",
    "realtids_cmd = \"\"\"grep -w -F -f \"\"\"+base_dir_data+\"\"\"real.locs \"\"\"+base_dir_data+\"\"\"ALL.merged.gtf | awk -F '\\t|\\\"' '$3==\"transcript\" {print $10}' | sort | uniq > \"\"\"+base_dir_data+\"\"\"real_loc.all_tids\"\"\"\n",
    "subprocess.call(realtids_cmd.split(\" \"))\n",
    "\n",
    "# extract all transcript IDs\n",
    "alltids_cmd = \"\"\"awk -F '\\t|\\\"' '$3==\"transcript\" {print $10}' \"\"\"+base_dir_data+\"\"\"ALL.merged.gtf | sort | uniq > \"\"\"+base_dir_data+\"\"\"all.tids\"\"\"\n",
    "subprocess.call(alltids_cmd,shell=True)\n",
    "\n",
    "# get tids of potentially RNApol\n",
    "poltids_cmd = \"\"\"diff --new-line-format=\"\" --unchanged-line-format=\"\"  \"\"\"+base_dir_data+\"\"\"all.tids \"\"\"+base_dir_data+\"\"\"real_loc.all_tids > \"\"\"+base_dir_data+\"\"\"RNApol.tids\"\"\"\n",
    "subprocess.call(poltids_cmd,shell=True)\n",
    "\n",
    "# get related RNApol tmap\n",
    "poltmap_cmd = \"\"\"grep -w -F -f \"\"\"+base_dir_data+\"\"\"RNApol.tids \"\"\"+base_dir_data+\"\"\"chess2ALL.ALL.merged.gtf.tmap > \"\"\"+base_dir_data+\"\"\"RNApol.tmap\"\"\"\n",
    "subprocess.call(poltmap_cmd.split(\" \"))\n",
    "\n",
    "# next, everything that is not in this set is RNApol noise\n",
    "polgtf_cmd = \"\"\"awk -F '\\t|\\\"' 'FNR==NR{ids[$1]++;next} {if($10 in ids) print $0}' <(awk -v pat='u' -F '\\t' '$3~pat {print $5}' RNApol.tmap) ALL.merged.gtf > \"\"\"+base_dir_data+\"\"\"RNApol.gtf\"\"\"\n",
    "subprocess.call(polgtf_cmd,shell=True)\n",
    "\n",
    "# next we need to separate out intronic and splicing noise\n",
    "# first get a set of real tids\n",
    "noise_cmd=\"\"\"awk -F '\\t|\\\"' '$3==\"transcript\" {print $10}' \"\"\"+base_dir_data+\"\"\"real.gtf | sort | uniq > \"\"\"+base_dir_data+\"\"\"real.tids\"\"\"\n",
    "subprocess.call(noise_cmd,shell=True)\n",
    "\n",
    "# next we can subtract this from the set of all tids in real locs\n",
    "all_cmd = \"\"\"diff --new-line-format=\"\" --unchanged-line-format=\"\"  \"\"\"+base_dir_data+\"\"\"real_loc.all_tids \"\"\"+base_dir_data+\"\"\"real.tids > \"\"\"+base_dir_data+\"\"\"real_loc.nonreal_tids\"\"\"\n",
    "subprocess.call(all_cmd,shell=True)\n",
    "\n",
    "# now we can generate the corresponding tmap file\n",
    "tmap_cmd = \"\"\"grep -w -F -f \"\"\"+base_dir_data+\"\"\"real_loc.nonreal_tids \"\"\"+base_dir_data+\"\"\"chess2ALL.ALL.merged.gtf.tmap > \"\"\"+base_dir_data+\"\"\"nonreal.tmap\"\"\"\n",
    "subprocess.call(tmap_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we can focus on figuring out how to extract real transcripts/loci\n",
    "# and adding noise transcripts to them\n",
    "\n",
    "# begin by extracting a list of chess IDs of the transcripts that match from ALL to CHESS\n",
    "# and convert chess transcript IDs to gene IDs\n",
    "chess_realloc_cmd = \"\"\"awk -v pat='=' -F '\\t' '$3~pat {print}' \"\"\"+base_dir_data+\"\"\"chess2ALL.ALL.merged.gtf.tmap > \"\"\"+base_dir_data+\"real.chess.locs\"\n",
    "subprocess.call(chess_realloc_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of transcripts in real: 87659\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_id</th>\n",
       "      <th>qry_id</th>\n",
       "      <th>gid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.6.1</td>\n",
       "      <td>ALL_00000003</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.8.1</td>\n",
       "      <td>ALL_00012984</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.7.28</td>\n",
       "      <td>ALL_00012993</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>ALL_00013034</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_id        qry_id  gid\n",
       "0   CHS.2.1  ALL_00012961    2\n",
       "1   CHS.6.1  ALL_00000003    6\n",
       "2   CHS.8.1  ALL_00012984    8\n",
       "3  CHS.7.28  ALL_00012993    7\n",
       "4  CHS.15.1  ALL_00013034   15"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract a subset of CHESS comprised of these loci only - this will serve as a truth set\n",
    "\n",
    "# first read in all the chess geneIDs\n",
    "real_tm = pd.read_csv(base_dir_data+\"real.chess.locs\",names=tmapcols,sep=\"\\t\",usecols=[\"ref_id\",\"qry_id\"])\n",
    "# only include those with CHS type IDs in the matching reference transcript\n",
    "real_tm = real_tm[real_tm[\"ref_id\"].str.contains(\"CHS\\.[0-9]*\\.[0-9]*\")].reset_index(drop=True)\n",
    "real_tm[\"gid\"] = real_tm[\"ref_id\"].str.split(\"\\.\",expand=True)[1].astype(int)\n",
    "# convert chess GFF with matching transcript to GTF as real.gtf\n",
    "assert len(real_tm[real_tm.duplicated(\"qry_id\",keep=\"first\")])==0,\"duplicated ALL IDs in real\"\n",
    "real_tm.drop_duplicates(\"ref_id\",keep=\"first\",inplace=True)\n",
    "assert len(real_tm[real_tm.duplicated(\"ref_id\",keep=\"first\")])==0,\"duplicated CHESS IDs in real\"\n",
    "print(\"number of transcripts in real: \"+str(len(set(real_tm[\"ref_id\"]))))\n",
    "real_tm.to_csv(base_dir_data+\"real_chess2all.tgids\",index=False)\n",
    "real_tm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['rRNA', 'exon', 'transcript', 'tRNA']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gid_x</th>\n",
       "      <th>tid</th>\n",
       "      <th>ref_id</th>\n",
       "      <th>qry_id</th>\n",
       "      <th>gid_y</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>2</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>2</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>2</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>2</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>2</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>ALL_00012961</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008298</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>56954255</td>\n",
       "      <td>56954656</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.59383.1\"; gene_id \"CHS.5938...</td>\n",
       "      <td>59383</td>\n",
       "      <td>CHS.59383.1</td>\n",
       "      <td>CHS.59383.1</td>\n",
       "      <td>ALL_00441799</td>\n",
       "      <td>59383</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008299</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>56960286</td>\n",
       "      <td>56968979</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.59383.1\"; gene_id \"CHS.5938...</td>\n",
       "      <td>59383</td>\n",
       "      <td>CHS.59383.1</td>\n",
       "      <td>CHS.59383.1</td>\n",
       "      <td>ALL_00441799</td>\n",
       "      <td>59383</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008300</th>\n",
       "      <td>chrY</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201135</td>\n",
       "      <td>57203577</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...</td>\n",
       "      <td>59386</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>ALL_00441903</td>\n",
       "      <td>59386</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008301</th>\n",
       "      <td>chrY</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201135</td>\n",
       "      <td>57202145</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...</td>\n",
       "      <td>59386</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>ALL_00441903</td>\n",
       "      <td>59386</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008302</th>\n",
       "      <td>chrY</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203182</td>\n",
       "      <td>57203577</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...</td>\n",
       "      <td>59386</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>CHS.59386.1</td>\n",
       "      <td>ALL_00441903</td>\n",
       "      <td>59386</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1008303 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid      source        type     start       end score strand phase  \\\n",
       "0        chr1  BestRefSeq  transcript     14362     29370     .      -     .   \n",
       "1        chr1  BestRefSeq        exon     14362     14829     .      -     .   \n",
       "2        chr1  BestRefSeq        exon     14970     15038     .      -     .   \n",
       "3        chr1  BestRefSeq        exon     15796     15947     .      -     .   \n",
       "4        chr1  BestRefSeq        exon     16607     16765     .      -     .   \n",
       "...       ...         ...         ...       ...       ...   ...    ...   ...   \n",
       "1008298  chrY  BestRefSeq        exon  56954255  56954656     .      +     .   \n",
       "1008299  chrY  BestRefSeq        exon  56960286  56968979     .      +     .   \n",
       "1008300  chrY      Gnomon  transcript  57201135  57203577     .      -     .   \n",
       "1008301  chrY      Gnomon        exon  57201135  57202145     .      -     .   \n",
       "1008302  chrY      Gnomon        exon  57203182  57203577     .      -     .   \n",
       "\n",
       "                                                attributes  gid_x  \\\n",
       "0        transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...      2   \n",
       "1        transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...      2   \n",
       "2        transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...      2   \n",
       "3        transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...      2   \n",
       "4        transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...      2   \n",
       "...                                                    ...    ...   \n",
       "1008298  transcript_id \"CHS.59383.1\"; gene_id \"CHS.5938...  59383   \n",
       "1008299  transcript_id \"CHS.59383.1\"; gene_id \"CHS.5938...  59383   \n",
       "1008300  transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...  59386   \n",
       "1008301  transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...  59386   \n",
       "1008302  transcript_id \"CHS.59386.1\"; gene_id \"CHS.5938...  59386   \n",
       "\n",
       "                 tid       ref_id        qry_id  gid_y _merge  \n",
       "0            CHS.2.1      CHS.2.1  ALL_00012961      2   both  \n",
       "1            CHS.2.1      CHS.2.1  ALL_00012961      2   both  \n",
       "2            CHS.2.1      CHS.2.1  ALL_00012961      2   both  \n",
       "3            CHS.2.1      CHS.2.1  ALL_00012961      2   both  \n",
       "4            CHS.2.1      CHS.2.1  ALL_00012961      2   both  \n",
       "...              ...          ...           ...    ...    ...  \n",
       "1008298  CHS.59383.1  CHS.59383.1  ALL_00441799  59383   both  \n",
       "1008299  CHS.59383.1  CHS.59383.1  ALL_00441799  59383   both  \n",
       "1008300  CHS.59386.1  CHS.59386.1  ALL_00441903  59386   both  \n",
       "1008301  CHS.59386.1  CHS.59386.1  ALL_00441903  59386   both  \n",
       "1008302  CHS.59386.1  CHS.59386.1  ALL_00441903  59386   both  \n",
       "\n",
       "[1008303 rows x 15 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chessDF = pd.read_csv(base_dir_data+\"chess2.2_assembly.gff\",sep=\"\\t\",comment=\"#\",names=gff3cols)\n",
    "chess_types = [x for x in set(chessDF[\"type\"]) if x not in [\"gene\",\"CDS\"]]\n",
    "print(chess_types)\n",
    "chessDF = chessDF[chessDF[\"type\"].isin(chess_types)].reset_index(drop=True)\n",
    "chessDF[\"gid\"] = chessDF[\"attributes\"].str.split(\"CHS.\",expand=True,n=1)[1].str.split(\"\\.|;\",expand=True,n=1)[0].astype(int)\n",
    "chessDF[\"tid\"] = np.where(chessDF[\"type\"]!=\"exon\",\n",
    "            chessDF[\"attributes\"].str.split(\"ID=\",expand=True,n=1)[1].str.split(\";\",expand=True,n=1)[0],\n",
    "            chessDF[\"attributes\"].str.split(\"Parent=\",expand=True,n=1)[1].str.split(\";\",expand=True,n=1)[0])\n",
    "chessDF = chessDF[chessDF[\"gid\"].isin(real_tm[\"gid\"])].reset_index(drop=True)\n",
    "chessDF = chessDF.merge(real_tm,left_on=\"tid\",right_on=\"ref_id\",how=\"right\",indicator=True)\n",
    "assert len(chessDF[chessDF[\"_merge\"]==\"both\"])==len(chessDF),\"not all are merged for real\"\n",
    "assert len(real_tm[\"ref_id\"])==len(set(chessDF[\"tid\"])),\"length of the dataframe has changed in real\"\n",
    "chessDF[\"attributes\"] = \"transcript_id \\\"\"+chessDF[\"tid\"]+\\\n",
    "                        \"\\\"; gene_id \\\"CHS.\"+chessDF[\"gid_x\"].astype(str)+\\\n",
    "                        \"\\\"; old_transcript_id_old \\\"\"+chessDF[\"qry_id\"]+\"\\\";\"\n",
    "chessDF[gff3cols].to_csv(base_dir_data+\"chess_real_matches.gtf\",header=False,index=False,sep=\"\\t\",quoting=csv.QUOTE_NONE)\n",
    "chessDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# next we need to perform the same operation for the splicing noise\n",
    "# first need to also retrieve chess IDs\n",
    "chess_spliceloc_cmd = \"\"\"awk -v pat='c|k|m|n|j|e' -F '\\t' '$3~pat {print}' \"\"\"+base_dir_data+\"\"\"nonreal.tmap > \"\"\"+base_dir_data+\"splice.chess.locs\"\n",
    "subprocess.call(chess_spliceloc_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_id</th>\n",
       "      <th>qry_id</th>\n",
       "      <th>gid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.16.3</td>\n",
       "      <td>ALL_00034985</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.24.2</td>\n",
       "      <td>ALL_00013071</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.26.2</td>\n",
       "      <td>ALL_00000025</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.26.2</td>\n",
       "      <td>ALL_00000026</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.26.2</td>\n",
       "      <td>ALL_00000030</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_id        qry_id  gid\n",
       "0  CHS.16.3  ALL_00034985   16\n",
       "1  CHS.24.2  ALL_00013071   24\n",
       "2  CHS.26.2  ALL_00000025   26\n",
       "3  CHS.26.2  ALL_00000026   26\n",
       "4  CHS.26.2  ALL_00000030   26"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to obtain a mapping from chess to ALL for the splicing noise transcripts\n",
    "sp_tm = pd.read_csv(base_dir_data+\"splice.chess.locs\",names=tmapcols,sep=\"\\t\",usecols=[\"ref_id\",\"qry_id\"])\n",
    "# only include those that belong to valid transcripts in CHESS\n",
    "sp_tm = sp_tm[sp_tm[\"ref_id\"].isin(real_tm[\"ref_id\"])].reset_index(drop=True)\n",
    "# only include those with CHS type IDs in the matching reference transcript\n",
    "sp_tm = sp_tm[sp_tm[\"ref_id\"].str.contains(\"CHS\\.[0-9]*\\.[0-9]*\")].reset_index(drop=True)\n",
    "sp_tm[\"gid\"] = sp_tm[\"ref_id\"].str.split(\"\\.\",expand=True)[1].astype(int)\n",
    "# make sure the IDs exist in the set of real genes\n",
    "sp_tm = sp_tm[sp_tm[\"gid\"].isin(real_tm[\"gid\"])]\n",
    "sp_tm.to_csv(base_dir_data+\"splicing_chess2all.tgids\",index=False)\n",
    "sp_tm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# next we need to perform the same operation for the intronic noise\n",
    "# first need to also retrieve chess IDs\n",
    "chess_intronloc_cmd = \"\"\"awk -v pat='i' -F '\\t' '$3~pat {print}' \"\"\"+base_dir_data+\"\"\"nonreal.tmap > \"\"\"+base_dir_data+\"intron.chess.locs\"\n",
    "subprocess.call(chess_intronloc_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_id</th>\n",
       "      <th>qry_id</th>\n",
       "      <th>gid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.41.9</td>\n",
       "      <td>ALL_00028752</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.138.5</td>\n",
       "      <td>ALL_00013566</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.146.7</td>\n",
       "      <td>ALL_00013616</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.161.1</td>\n",
       "      <td>ALL_00000580</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.180.8</td>\n",
       "      <td>ALL_00000614</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ref_id        qry_id  gid\n",
       "0   CHS.41.9  ALL_00028752   41\n",
       "1  CHS.138.5  ALL_00013566  138\n",
       "2  CHS.146.7  ALL_00013616  146\n",
       "3  CHS.161.1  ALL_00000580  161\n",
       "4  CHS.180.8  ALL_00000614  180"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to obtain a mapping from chess to ALL for the intronic noise transcripts\n",
    "int_tm = pd.read_csv(base_dir_data+\"intron.chess.locs\",names=tmapcols,sep=\"\\t\",usecols=[\"ref_id\",\"qry_id\"])\n",
    "# only include those that belong to valid transcripts in CHESS\n",
    "int_tm = int_tm[int_tm[\"ref_id\"].isin(real_tm[\"ref_id\"])].reset_index(drop=True)\n",
    "# only include those with CHS type IDs in the matching reference transcript\n",
    "int_tm = int_tm[int_tm[\"ref_id\"].str.contains(\"CHS\\.[0-9]*\\.[0-9]*\")].reset_index(drop=True)\n",
    "int_tm[\"gid\"] = int_tm[\"ref_id\"].str.split(\"\\.\",expand=True)[1].astype(int)\n",
    "int_tm.to_csv(base_dir_data+\"intronic_chess2all.tgids\",index=False)\n",
    "int_tm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162177"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_qry_ids = set(real_tm[\"qry_id\"]).union(set(sp_tm[\"qry_id\"]).union(int_tm[\"qry_id\"]))\n",
    "len(all_qry_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to build a dictionary from qry_id to ref_id\n",
    "all_tm = pd.concat([real_tm,sp_tm,int_tm],axis=0)\n",
    "id_dict = pd.Series(all_tm.ref_id.values,index=all_tm.qry_id).to_dict()\n",
    "real_dict = pd.Series(real_tm.ref_id.values,index=real_tm.qry_id).to_dict()\n",
    "sp_dict = pd.Series(sp_tm.ref_id.values,index=sp_tm.qry_id).to_dict()\n",
    "int_dict = pd.Series(int_tm.ref_id.values,index=int_tm.qry_id).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert set(id_dict)==set(all_tm[\"qry_id\"]),\"qry sets not identical\"\n",
    "assert set(id_dict.values())==set(all_tm[\"ref_id\"]),\"ref sets not identical\"\n",
    "assert len(list(id_dict))==len(list(all_tm[\"qry_id\"].tolist())),\"qry lists of different lengths\"\n",
    "assert len(list(id_dict.values()))==len(list(all_tm[\"ref_id\"].tolist())),\"ref lists of different lengths\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n",
      "2.0\n"
     ]
    }
   ],
   "source": [
    "count=0\n",
    "\n",
    "real_gtf_fp = open(base_dir_data+\"real_1.gtf\",\"w+\")\n",
    "sp_gtf_fp = open(base_dir_data+\"splicing_1.gtf\",\"w+\")\n",
    "int_gtf_fp = open(base_dir_data+\"intronic_1.gtf\",\"w+\")\n",
    "\n",
    "with open(base_dir_data+\"ALL.merged.gtf\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        line = line.strip()\n",
    "        lineCols = line.split(\"\\t\")\n",
    "        count+=1\n",
    "        if count%1000000==0:\n",
    "            print(count/1000000)\n",
    "            \n",
    "        # get transcript and gene IDs\n",
    "        attrs = lineCols[8].split(\"\\\"\")\n",
    "        tid = attrs[1]\n",
    "        gid = attrs[3]\n",
    "        ntid = \"\"\n",
    "        ngid = \"\"\n",
    "        # now find the corresponding chess IDs\n",
    "        if not tid in id_dict: # nothing found\n",
    "            continue\n",
    "            \n",
    "        if tid in real_dict:\n",
    "            ntid = real_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = \"\\t\".join(lineCols[:8])\\\n",
    "                             +\"\\ttranscript_id \\\"\"+ntid\\\n",
    "                             +\"\\\"; gene_id \\\"\"+ngid\\\n",
    "                             +\"\\\"; old_transcript_id_old \\\"\"+tid\\\n",
    "                             +\"\\\"; old_gene_id \\\"\"+gid+\"\\\";\\n\"\n",
    "#             print(new_line)\n",
    "            real_gtf_fp.write(new_line)\n",
    "            \n",
    "        \n",
    "        elif tid in sp_dict:\n",
    "            ntid = sp_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = \"\\t\".join(lineCols[:8])\\\n",
    "                             +\"\\ttranscript_id \\\"\"+tid\\\n",
    "                             +\"\\\"; gene_id \\\"\"+ngid\\\n",
    "                             +\"\\\"; old_transcript_id_old \\\"\"+tid\\\n",
    "                             +\"\\\"; old_gene_id \\\"\"+gid+\"\\\";\\n\"\n",
    "#             print(new_line)\n",
    "            sp_gtf_fp.write(new_line)\n",
    "        \n",
    "        elif tid in int_dict:\n",
    "            ntid = int_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = \"\\t\".join(lineCols[:8])\\\n",
    "                             +\"\\ttranscript_id \\\"\"+tid\\\n",
    "                             +\"\\\"; gene_id \\\"\"+ngid\\\n",
    "                             +\"\\\"; old_transcript_id_old \\\"\"+tid\\\n",
    "                             +\"\\\"; old_gene_id \\\"\"+gid+\"\\\";\\n\"\n",
    "#             print(new_line)\n",
    "            int_gtf_fp.write(new_line)\n",
    "            \n",
    "        else: # nothing found\n",
    "            print(\"ERROR\")\n",
    "            break;\n",
    "            \n",
    "#         break;\n",
    "        \n",
    "real_gtf_fp.close()\n",
    "sp_gtf_fp.close()\n",
    "int_gtf_fp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly need to concatenate all 4 to create a new set of transcripts\n",
    "# interesting to know how many remain in total\n",
    "cat_cmd = \"cat \"+base_dir_data+\"real_1.gtf \"+base_dir_data+\"splicing_1.gtf \"+base_dir_data+\"intronic_1.gtf \"+base_dir_data+\"RNApol_1.gtf > \"+base_dir_data+\"ALL.merged.filtered.gtf\"\n",
    "subprocess.call(cat_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we need to modify intergenic set to replcae all undefined strands with real strand information\n",
    "polDF = pd.read_csv(base_dir_data+\"RNApol_1.gtf\",sep=\"\\t\",comment=\"#\",names=gff3cols)\n",
    "polDF[\"tid\"] = polDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "undef_strand_tids = polDF[(polDF[\"type\"]==\"transcript\")&(polDF[\"strand\"]==\".\")].tid.tolist()\n",
    "def_strand_tids = polDF[(polDF[\"type\"]==\"transcript\")&~(polDF[\"strand\"]==\".\")].tid.tolist()\n",
    "assert len(undef_strand_tids)==len(set(undef_strand_tids)),\"duplicates in undefined\"\n",
    "assert len(def_strand_tids)==len(set(def_strand_tids)),\"duplicates in defined\"\n",
    "assert len(set(undef_strand_tids).intersection(set(def_strand_tids)))==0,\"same transcripts in both groups\"\n",
    "random_strands = random.choices([\"+\",\"-\"],k=len(undef_strand_tids))\n",
    "undef_strand_df = pd.concat([pd.DataFrame(undef_strand_tids,columns=[\"tid\"]),\\\n",
    "                             pd.DataFrame(random_strands,columns=[\"strand_new\"])],axis=1)\n",
    "polDF = polDF.merge(undef_strand_df,on=\"tid\",how=\"left\")\n",
    "polDF[\"strand\"] = np.where(polDF[\"strand\"]==\".\",polDF[\"strand_new\"],polDF[\"strand\"])\n",
    "polDF[gff3cols].to_csv(base_dir_data+\"RNApol.gtf\",header=False,index=False,sep=\"\\t\",quoting=csv.QUOTE_NONE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first we need to load the RNApol set in order to know which transcripts to output for that set\n",
    "polDF = pd.read_csv(base_dir_data+\"RNApol.gtf\",sep=\"\\t\",comment=\"#\",names=gff3cols)\n",
    "polDF = polDF[polDF[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "polDF[\"tid\"] = polDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "pol_set = set(polDF[\"tid\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we need to modify any other files with the updated information\n",
    "# namely, we need to update the tracking file with the updated information\n",
    "count=0\n",
    "\n",
    "real_tr_fp = open(base_dir_data+\"real.tracking\",\"w+\")\n",
    "sp_tr_fp = open(base_dir_data+\"splicing.tracking\",\"w+\")\n",
    "int_tr_fp = open(base_dir_data+\"intronic.tracking\",\"w+\")\n",
    "pol_tr_fp = open(base_dir_data+\"RNApol.tracking\",\"w+\")\n",
    "\n",
    "with open(base_dir_data+\"ALL.tracking\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        line = line.strip()\n",
    "        lineCols = line.split(\"\\t\")\n",
    "        count+=1\n",
    "        if count%1000000==0:\n",
    "            print(count/1000000)\n",
    "            \n",
    "        # get transcript and gene IDs\n",
    "        tid = lineCols[0]\n",
    "        # now find the corresponding chess IDs\n",
    "        if not tid in id_dict and not tid in pol_set: # nothing found\n",
    "            continue\n",
    "            \n",
    "        if tid in real_dict:\n",
    "            ntid = real_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = ntid+\"\\t\"+ngid+\"\\t\"+\"\\t\".join(lineCols[2:])+\"\\n\"\n",
    "#             print(new_line)\n",
    "            real_tr_fp.write(new_line)\n",
    "            \n",
    "        \n",
    "        elif tid in sp_dict:\n",
    "            ntid = sp_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = tid+\"\\t\"+ngid+\"\\t\"+\"\\t\".join(lineCols[2:])+\"\\n\"\n",
    "#             print(new_line)\n",
    "            sp_tr_fp.write(new_line)\n",
    "        \n",
    "        elif tid in int_dict:\n",
    "            ntid = int_dict[tid]\n",
    "            ngid = \"CHS.\"+ntid.split(\".\")[1]\n",
    "            new_line = tid+\"\\t\"+ngid+\"\\t\"+\"\\t\".join(lineCols[2:])+\"\\n\"\n",
    "#             print(new_line)\n",
    "            int_tr_fp.write(new_line)\n",
    "    \n",
    "        elif tid in pol_set:\n",
    "            pol_tr_fp.write(line+\"\\n\")\n",
    "            \n",
    "        else: # nothing found\n",
    "            print(\"ERROR\")\n",
    "            break;\n",
    "            \n",
    "#         break;\n",
    "        \n",
    "real_tr_fp.close()\n",
    "sp_tr_fp.close()\n",
    "int_tr_fp.close()\n",
    "pol_tr_fp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly need to concatenate all 4 to create a new tracking file\n",
    "cat_tr_cmd = \"cat \"+base_dir_data+\"real.tracking \"+\\\n",
    "          base_dir_data+\"splicing.tracking \"+\\\n",
    "          base_dir_data+\"intronic.tracking \"+\\\n",
    "          base_dir_data+\"RNApol.tracking > \"+base_dir_data+\"ALL.merged.filtered.tracking\"\n",
    "subprocess.call(cat_tr_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>loading base annotations\n",
      ">>>getting loci IDs\n"
     ]
    }
   ],
   "source": [
    "# deal with missing strand for real loci now\n",
    "# load base annotations\n",
    "print(\">>>loading base annotations\")\n",
    "real_baseDF = pd.read_csv(base_dir_data+\"real_1.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "splice_baseDF = pd.read_csv(base_dir_data+\"splicing_1.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "int_baseDF = pd.read_csv(base_dir_data+\"intronic_1.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "\n",
    "# get all loci and transcript IDs\n",
    "print(\">>>getting loci IDs\")\n",
    "real_baseDF[\"lid\"] = real_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "splice_baseDF[\"lid\"] = splice_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "int_baseDF[\"lid\"] = int_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>lid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.1.1;Parent=CHS.1;gene_name=DDX11L1;Dbxr...</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid      source        type  start    end score strand phase  \\\n",
       "0  chr1      RefSeq        gene  11874  14409     .      +     .   \n",
       "1  chr1  BestRefSeq  transcript  11874  14409     .      +     .   \n",
       "2  chr1  BestRefSeq        exon  11874  12227     .      +     .   \n",
       "3  chr1  BestRefSeq        exon  12613  12721     .      +     .   \n",
       "4  chr1  BestRefSeq        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes    lid  \n",
       "0  ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...  CHS.1  \n",
       "1  ID=CHS.1.1;Parent=CHS.1;gene_name=DDX11L1;Dbxr...  CHS.1  \n",
       "2                                     Parent=CHS.1.1  CHS.1  \n",
       "3                                     Parent=CHS.1.1  CHS.1  \n",
       "4                                     Parent=CHS.1.1  CHS.1  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chessDF = pd.read_csv(base_dir_data+\"chess2.2_assembly.gff\",sep=\"\\t\",comment=\"#\",names=gff3cols)\n",
    "chessDF[chessDF[\"type\"]==\"gene\"].reset_index(drop=True)\n",
    "chessDF[\"lid\"] = \"CHS.\"+chessDF[\"attributes\"].str.split(\"CHS.\",expand=True,n=1)[1].str.split(\"\\.|;\",expand=True,n=1)[0]\n",
    "chessDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "getting real strands\n"
     ]
    }
   ],
   "source": [
    "print(\"getting real strands\")\n",
    "st_df = chessDF[[\"lid\",\"strand\"]].drop_duplicates(keep=\"first\")\n",
    "assert set(st_df[\"strand\"])==set([\"+\",\"-\"]),\"wrong strands\"\n",
    "st_df.columns = [\"lid\",\"strand_new\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "setting real\n",
      "setting intronic\n",
      "setting splicing\n"
     ]
    }
   ],
   "source": [
    "print(\"setting real\")\n",
    "real_baseDF2=real_baseDF.merge(st_df,on=\"lid\",how=\"inner\")\n",
    "real_baseDF2[\"strand\"] = real_baseDF2[\"strand_new\"]\n",
    "assert set(real_baseDF2[\"strand\"]==set([\"+\",\"-\"])),\"wrong strands real\"\n",
    "\n",
    "print(\"setting intronic\")\n",
    "int_baseDF2=int_baseDF.merge(st_df,on=\"lid\",how=\"inner\")\n",
    "int_baseDF2[\"strand\"] = int_baseDF2[\"strand_new\"]\n",
    "assert set(int_baseDF2[\"strand\"])==set([\"+\",\"-\"]),\"wrong strands int\"\n",
    "\n",
    "print(\"setting splicing\")\n",
    "splice_baseDF2=splice_baseDF.merge(st_df,on=\"lid\",how=\"inner\")\n",
    "splice_baseDF2[\"strand\"] = splice_baseDF2[\"strand_new\"]\n",
    "assert set(splice_baseDF2[\"strand\"])==set([\"+\",\"-\"]),\"wrong strands splice\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mv_cmd = \"\"\"mv \"\"\"+base_dir_data+\"\"\"splicing.gtf \"\"\"+base_dir_data+\"\"\"splicing_old.gtf\"\"\"\n",
    "# subprocess.call(mv_cmd,shell=True)\n",
    "# mv_cmd = \"\"\"mv \"\"\"+base_dir_data+\"\"\"intronic.gtf \"\"\"+base_dir_data+\"\"\"intronic_old.gtf\"\"\"\n",
    "# subprocess.call(mv_cmd,shell=True)\n",
    "mv_cmd = \"\"\"mv \"\"\"+base_dir_data+\"\"\"ALL.merged.filtered.gtf \"\"\"+base_dir_data+\"\"\"ALL.merged.filtered_old.gtf\"\"\"\n",
    "subprocess.call(mv_cmd,shell=True)\n",
    "\n",
    "real_baseDF2[gff3cols].to_csv(base_dir_data+\"real.gtf\",header=False,index=False,sep=\"\\t\",quoting=csv.QUOTE_NONE)\n",
    "splice_baseDF2[gff3cols].to_csv(base_dir_data+\"splicing.gtf\",header=False,index=False,sep=\"\\t\",quoting=csv.QUOTE_NONE)\n",
    "int_baseDF2[gff3cols].to_csv(base_dir_data+\"intronic.gtf\",header=False,index=False,sep=\"\\t\",quoting=csv.QUOTE_NONE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly need to concatenate all 4 to create a new set of transcripts again\n",
    "cat_cmd = \"cat \"+base_dir_data+\"real.gtf \"+base_dir_data+\"splicing.gtf \"+base_dir_data+\"intronic.gtf \"+base_dir_data+\"RNApol.gtf > \"+base_dir_data+\"ALL.merged.filtered.gtf\"\n",
    "subprocess.call(cat_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_cmd = \"/ccb/salz8-1/avaraby/tx_noise/soft/gtex_stats/gtex_stats \\\n",
    "           -t \"+base_dir_data+\"tissues/tissues.lst \\\n",
    "           -a \"+base_dir_data+\"ALL.merged.filtered.tracking \\\n",
    "           -r \"+base_dir_data+\"real.gtf \\\n",
    "           -g \"+base_dir_data+\"ALL.merged.filtered.gtf \\\n",
    "           -s \"+base_dir_data+\"splicing.gtf \\\n",
    "           -i \"+base_dir_data+\"intronic.gtf \\\n",
    "           -p \"+base_dir_data+\"RNApol.gtf \\\n",
    "           -o \"+base_dir_out+\"res\"\n",
    "# print(agg_cmd)\n",
    "subprocess.call(agg_cmd,shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
