{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/sim_samples/\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "real_gtf = pd.read_csv(base_dir_data+\"real.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "real_gtf[\"tid_real\"] = real_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "splice_gtf = pd.read_csv(base_dir_data+\"splicing.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "splice_gtf[\"tid_splicing\"] = splice_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "int_gtf = pd.read_csv(base_dir_data+\"intronic.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "int_gtf[\"tid_intronic\"] = int_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "pol_gtf = pd.read_csv(base_dir_data+\"RNApol.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "pol_gtf[\"tid_polymerase\"] = pol_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# based on the number of loci per sample\n",
    "s_loc = pd.read_csv(base_dir_out+\"res.num_locs_sample\")\n",
    "\n",
    "for label in s_loc.columns:\n",
    "    # now we shall remove any outliers from the data\n",
    "    q25,q50,q75 = s_loc[label].quantile([0.25,0.5,0.75])\n",
    "    iqr = q75-q25\n",
    "    thw = q75+1.5*iqr\n",
    "    tlw = q25-1.5*iqr\n",
    "    ahw = s_loc[s_loc[label]<=thw][label].max()\n",
    "    alw = s_loc[s_loc[label]>=tlw][label].min()\n",
    "    s_loc = s_loc[(s_loc[label]<=ahw)&(s_loc[label]>=alw)].reset_index(drop=True)\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "ax.boxplot(s_loc.T,labels=s_loc.T.index.tolist())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    realDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.real_t\"+str(tissue_num))\n",
    "    spliceDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.splice_t\"+str(tissue_num))\n",
    "    intDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.int_t\"+str(tissue_num))\n",
    "    polDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.pol_t\"+str(tissue_num))\n",
    "    \n",
    "    print(\"starting number of real loci: \"+str(len(set(realDF[\"lid\"]))))\n",
    "    print(\"starting number of splicing loci: \"+str(len(set(spliceDF[\"lid\"]))))\n",
    "    print(\"starting number of intronic loci: \"+str(len(set(intDF[\"lid\"]))))\n",
    "    print(\"starting number of intergenic loci: \"+str(len(set(polDF[\"lid\"]))))\n",
    "    \n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        # ==================================\n",
    "        # get numbers\n",
    "        # ==================================\n",
    "        realn,splicen,intn,real_splicen,real_intn,splice_intn,real_splice_intn,all_realn,poln = s_loc.sample()[[\"real\",\\\n",
    "                                                                                                                \"splicing\",\\\n",
    "                                                                                                                \"intronic\",\\\n",
    "                                                                                                                \"real_splicing\",\\\n",
    "                                                                                                                \"real_intronic\",\\\n",
    "                                                                                                                \"splicing_intronic\",\\\n",
    "                                                                                                                \"real_splicing_intronic\",\\\n",
    "                                                                                                                \"all_real\",\\\n",
    "                                                                                                                \"intergenic\"]].iloc[0].tolist()\n",
    "        \n",
    "        # ==================================\n",
    "        # Now to generate sample from tissue\n",
    "        # for intergenic set\n",
    "        # ==================================\n",
    "        \n",
    "        # now let's split this into samples the old way\n",
    "        # now let's split this into samples the old way\n",
    "        realDF_g = realDF.groupby(\"lid\").agg({\"tid_real\":{\"tids_real\":lambda x:list(x),\n",
    "                                                                \"count_real\":\"count\"},\n",
    "                                              \"real_tpms\":\"first\"}).reset_index()\n",
    "        realDF_g.columns = [\"lid\",\"tids_real\",\"count_real\",\"real_tpms\"]\n",
    "        realDF_g[\"real_tpms\"] = realDF_g[\"real_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        spliceDF_g = spliceDF.groupby(\"lid\").agg({\"tid_splicing\":{\"tids_splicing\":lambda x:list(x),\n",
    "                                                                \"count_splicing\":\"count\"},\n",
    "                                              \"splicing_tpms\":\"first\"}).reset_index()\n",
    "        spliceDF_g.columns = [\"lid\",\"tids_splicing\",\"count_splicing\",\"splicing_tpms\"]\n",
    "        spliceDF_g[\"splicing_tpms\"] = spliceDF_g[\"splicing_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        intDF_g = intDF.groupby(\"lid\").agg({\"tid_intronic\":{\"tids_intronic\":lambda x:list(x),\n",
    "                                                                \"count_intronic\":\"count\"},\n",
    "                                              \"intronic_tpms\":\"first\"}).reset_index()\n",
    "        intDF_g.columns = [\"lid\",\"tids_intronic\",\"count_intronic\",\"intronic_tpms\"]\n",
    "        intDF_g[\"intronic_tpms\"] = intDF_g[\"intronic_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        polDF_g = polDF.groupby(\"lid\").agg({\"tid_intergenic\":{\"tids_intergenic\":lambda x:list(x),\n",
    "                                                                \"count_intergenic\":\"count\"},\n",
    "                                              \"intergenic_tpms\":\"first\"}).reset_index()\n",
    "        polDF_g.columns = [\"lid\",\"tids_intergenic\",\"count_intergenic\",\"intergenic_tpms\"]\n",
    "        polDF_g[\"intergenic_tpms\"] = polDF_g[\"intergenic_tpms\"].str.split(\";\")\n",
    "\n",
    "        all_g_lid = realDF_g.merge(spliceDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid = all_g_lid.merge(intDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid = all_g_lid.merge(polDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid[\"count_real\"] = all_g_lid[\"count_real\"].fillna(0)\n",
    "        all_g_lid[\"count_splicing\"] = all_g_lid[\"count_splicing\"].fillna(0)\n",
    "        all_g_lid[\"count_intronic\"] = all_g_lid[\"count_intronic\"].fillna(0)\n",
    "        all_g_lid[\"count_intergenic\"] = all_g_lid[\"count_intergenic\"].fillna(0)\n",
    "\n",
    "        all_g_lid[\"count_real\"]=all_g_lid[\"count_real\"].astype(int)\n",
    "        all_g_lid[\"count_splicing\"]=all_g_lid[\"count_splicing\"].astype(int)\n",
    "        all_g_lid[\"count_intronic\"]=all_g_lid[\"count_intronic\"].astype(int)\n",
    "        all_g_lid[\"count_intergenic\"]=all_g_lid[\"count_intergenic\"].astype(int)\n",
    "\n",
    "        all_g_lid.fillna(\"\",inplace=True)\n",
    "\n",
    "        # need a function to get the samples from this now\n",
    "        lids_intergenic=[]\n",
    "        tids_intergenic=[]\n",
    "        tpms_intergenic=[]\n",
    "\n",
    "        lids_real=[]\n",
    "        tids_real=[]\n",
    "        tpms_real=[]\n",
    "\n",
    "        lids_splicing=[]\n",
    "        tids_splicing=[]\n",
    "        tpms_splicing=[]\n",
    "\n",
    "        lids_intronic=[]\n",
    "        tids_intronic=[]\n",
    "        tpms_intronic=[]\n",
    "\n",
    "        def get_samples(row):\n",
    "            global lids_intergenic\n",
    "            global tids_intergenic\n",
    "            global tpms_intergenic\n",
    "            if(row[\"count_intergenic\"]>0):\n",
    "                all_intergenic_tids = np.array(row[\"tids_intergenic\"])\n",
    "                all_intergenic_tpms = row[\"intergenic_tpms\"]\n",
    "                # randomly choose a sample\n",
    "                rand_sample_idx = random.randint(0,len(all_intergenic_tpms)-1)\n",
    "                tmp_tpms = all_intergenic_tpms[rand_sample_idx].split(\"-\")\n",
    "                tid_ids = []\n",
    "                res_tpms = []\n",
    "                for i in range(len(tmp_tpms)):\n",
    "                    if not tmp_tpms[i]==\"0\":\n",
    "                        tid_ids.append(i)\n",
    "                        res_tpms.append(float(tmp_tpms[i]))\n",
    "                res_tids = all_intergenic_tids[tid_ids]\n",
    "                lids_intergenic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                tids_intergenic.extend(res_tids)\n",
    "                tpms_intergenic.extend(res_tpms)\n",
    "            else:\n",
    "                all_real_tpms = row[\"real_tpms\"]\n",
    "                all_splicing_tpms = row[\"splicing_tpms\"]\n",
    "                all_intronic_tpms = row[\"intronic_tpms\"]\n",
    "\n",
    "                all_real_tids = np.array(row[\"tids_real\"])\n",
    "                all_splicing_tids = np.array(row[\"tids_splicing\"])\n",
    "                all_intronic_tids = np.array(row[\"tids_intronic\"])\n",
    "\n",
    "                num_observations = max(len(all_real_tpms),len(all_splicing_tpms),len(all_intronic_tpms))\n",
    "                rand_sample_idx = random.randint(0,num_observations-1)\n",
    "\n",
    "                if row[\"count_real\"]>0:\n",
    "                    assert len(all_real_tpms)==num_observations\n",
    "                    tmp_tpms = all_real_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_real_tids[tid_ids]\n",
    "                    lids_real.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_real.extend(res_tids)\n",
    "                    tpms_real.extend(res_tpms)\n",
    "                if row[\"count_splicing\"]>0:\n",
    "                    assert len(all_splicing_tpms)==num_observations\n",
    "                    tmp_tpms = all_splicing_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_splicing_tids[tid_ids]\n",
    "                    lids_splicing.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_splicing.extend(res_tids)\n",
    "                    tpms_splicing.extend(res_tpms)\n",
    "                if row[\"count_intronic\"]>0:\n",
    "                    assert len(all_intronic_tpms)==num_observations\n",
    "                    tmp_tpms = all_intronic_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_intronic_tids[tid_ids]\n",
    "                    lids_intronic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_intronic.extend(res_tids)\n",
    "                    tpms_intronic.extend(res_tpms)\n",
    "\n",
    "        all_g_lid.apply(lambda row: get_samples(row),axis=1)\n",
    "        \n",
    "        print(\"number of real transcripts: \"+str(len(set(tids_real))))\n",
    "        print(\"number of splicing transcripts: \"+str(len(set(tids_splicing))))\n",
    "        print(\"number of intronic transcripts: \"+str(len(set(tids_intronic))))\n",
    "        print(\"number of intergenic transcripts: \"+str(len(set(tids_intergenic))))\n",
    "        print(\"number of real genes: \"+str(len(set(lids_real))))\n",
    "        print(\"number of solicing genes: \"+str(len(set(lids_splicing))))\n",
    "        print(\"number of intronic genes: \"+str(len(set(lids_intronic))))\n",
    "        print(\"number of intergenic genes: \"+str(len(set(lids_intergenic))))\n",
    "\n",
    "        print(\"targeting \"+str(realn)+\" real only loci\")\n",
    "        print(\"targeting \"+str(splicen)+\" splicing only loci\")\n",
    "        print(\"targeting \"+str(intn)+\" intronic only loci\")\n",
    "        print(\"targeting \"+str(real_splicen)+\" real and splicing loci\")\n",
    "        print(\"targeting \"+str(real_intn)+\" real and intronic loci\")\n",
    "        print(\"targeting \"+str(splice_intn)+\" splicing and intronic loci\")\n",
    "        print(\"targeting \"+str(real_splice_intn)+\" real and splicing and intronic loci\")\n",
    "        print(\"targeting \"+str(poln)+\" intergenic loci\")\n",
    "\n",
    "        real_sample_locs = set(lids_real)\n",
    "        splicing_sample_locs = set(lids_splicing)\n",
    "        intronic_sample_locs = set(lids_intronic)\n",
    "        intergenic_sample_locs = list(set(lids_intergenic))\n",
    "\n",
    "        real_only_sample_locs = list(real_sample_locs-(splicing_sample_locs.union(intronic_sample_locs)))\n",
    "        splicing_only_sample_locs = list(splicing_sample_locs-(real_sample_locs.union(intronic_sample_locs)))\n",
    "        intronic_only_sample_locs = list(intronic_sample_locs-(real_sample_locs.union(splicing_sample_locs)))\n",
    "\n",
    "        real_splicing_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs)-intronic_sample_locs)\n",
    "        real_intronic_sample_locs = list(real_sample_locs.intersection(intronic_sample_locs)-splicing_sample_locs)\n",
    "        splicing_intronic_sample_locs = list(splicing_sample_locs.intersection(intronic_sample_locs)-real_sample_locs)\n",
    "\n",
    "        real_splicing_intronic_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs.intersection(intronic_sample_locs)))\n",
    "\n",
    "        print(\"available real only loci: \"+str(len(real_only_sample_locs)))\n",
    "        print(\"available splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "        print(\"available intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "        print(\"available real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "        print(\"available real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "        print(\"available splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "        print(\"available real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "        print(\"available intergenic only loci: \"+str(len(intergenic_sample_locs)))\n",
    "\n",
    "        random.shuffle(real_only_sample_locs)\n",
    "        min_realn = min(len(real_only_sample_locs),realn)\n",
    "        real_only_sample_locs = real_only_sample_locs[:min_realn]\n",
    "        random.shuffle(splicing_only_sample_locs)\n",
    "        min_splicen = min(len(splicing_only_sample_locs),splicen)\n",
    "        splicing_only_sample_locs = splicing_only_sample_locs[:min_splicen]\n",
    "        random.shuffle(intronic_only_sample_locs)\n",
    "        min_intn = min(len(intronic_only_sample_locs),intn)\n",
    "        intronic_only_sample_locs = intronic_only_sample_locs[:min_intn]\n",
    "        random.shuffle(intergenic_sample_locs)\n",
    "        min_poln = min(len(intergenic_sample_locs),poln)\n",
    "        intergenic_sample_locs = intergenic_sample_locs[:min_poln]\n",
    "\n",
    "        random.shuffle(real_splicing_sample_locs)\n",
    "        min_real_splicen = min(len(real_splicing_sample_locs),real_splicen)\n",
    "        real_splicing_sample_locs = real_splicing_sample_locs[:min_real_splicen]\n",
    "        random.shuffle(real_intronic_sample_locs)\n",
    "        min_real_intn = min(len(real_intronic_sample_locs),real_intn)\n",
    "        real_intronic_sample_locs = real_intronic_sample_locs[:min_real_intn]\n",
    "        random.shuffle(splicing_intronic_sample_locs)\n",
    "        min_splice_intn = min(len(splicing_intronic_sample_locs),splice_intn)\n",
    "        splicing_intronic_sample_locs = splicing_intronic_sample_locs[:min_splice_intn]\n",
    "\n",
    "        random.shuffle(real_splicing_intronic_sample_locs)\n",
    "        min_real_splice_intn = min(len(real_splicing_intronic_sample_locs),real_splice_intn)\n",
    "        real_splicing_intronic_sample_locs = real_splicing_intronic_sample_locs[:min_real_splice_intn]\n",
    "\n",
    "        print(\"selecting real only loci: \"+str(len(real_only_sample_locs)))\n",
    "        print(\"selecting splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "        print(\"selecting intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "        print(\"selecting real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "        print(\"selecting real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "        print(\"selecting splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "        print(\"selecting real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "        print(\"selecting intergenic loci: \"+str(len(intergenic_sample_locs)))\n",
    "        \n",
    "        real_sub = pd.concat([pd.DataFrame(lids_real,columns=[\"lid\"]),pd.DataFrame(tids_real,columns=[\"tid\"]),pd.DataFrame(tpms_real,columns=[\"tpm\"])],axis=1)\n",
    "        splicing_sub = pd.concat([pd.DataFrame(lids_splicing,columns=[\"lid\"]),pd.DataFrame(tids_splicing,columns=[\"tid\"]),pd.DataFrame(tpms_splicing,columns=[\"tpm\"])],axis=1)\n",
    "        intronic_sub = pd.concat([pd.DataFrame(lids_intronic,columns=[\"lid\"]),pd.DataFrame(tids_intronic,columns=[\"tid\"]),pd.DataFrame(tpms_intronic,columns=[\"tpm\"])],axis=1)\n",
    "        intergenic_sub = pd.concat([pd.DataFrame(lids_intergenic,columns=[\"lid\"]),pd.DataFrame(tids_intergenic,columns=[\"tid\"]),pd.DataFrame(tpms_intergenic,columns=[\"tpm\"])],axis=1)\n",
    "        \n",
    "        real_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_only_sample_locs)].reset_index(drop=True)\n",
    "        splicing_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_only_sample_locs)].reset_index(drop=True)\n",
    "        intronic_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(intronic_only_sample_locs)].reset_index(drop=True)\n",
    "        intergenic_sub2 = intergenic_sub[intergenic_sub[\"lid\"].isin(intergenic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_splicing_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "        splicing_real_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_real_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        splicing_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_splicing_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        splicing_real_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_real_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_sub2 = pd.concat([real_only_sample_df,\\\n",
    "                               real_splicing_only_sample_df,\\\n",
    "                               real_intronic_only_sample_df,\\\n",
    "                               real_splicing_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "        splicing_sub2 = pd.concat([splicing_only_sample_df,\\\n",
    "                                   splicing_real_only_sample_df,\\\n",
    "                                   splicing_intronic_only_sample_df,\\\n",
    "                                   splicing_real_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "        intronic_sub2 = pd.concat([intronic_only_sample_df,\\\n",
    "                                   intronic_real_only_sample_df,\\\n",
    "                                   intronic_splicing_only_sample_df,\\\n",
    "                                   intronic_real_splicing_only_sample_df],axis=0).reset_index(drop=True)\n",
    "\n",
    "        print(\"final number of real transcripts: \"+str(len(set(real_sub2[\"tid\"]))))\n",
    "        print(\"final number of splicing transcripts: \"+str(len(set(splicing_sub2[\"tid\"]))))\n",
    "        print(\"final number of intronic transcripts: \"+str(len(set(intronic_sub2[\"tid\"]))))\n",
    "        print(\"final number of intergenic transcripts: \"+str(len(set(intergenic_sub2[\"tid\"]))))\n",
    "        print(\"final number of real loci: \"+str(len(set(real_sub2[\"lid\"]))))\n",
    "        print(\"final number of splicing loci: \"+str(len(set(splicing_sub2[\"lid\"]))))\n",
    "        print(\"final number of intronic loci: \"+str(len(set(intronic_sub2[\"lid\"]))))\n",
    "        print(\"final number of intergenic loci: \"+str(len(set(intergenic_sub2[\"lid\"]))))\n",
    "\n",
    "        ttx_real = list(real_sub2[\"tid\"])\n",
    "        tpm_real = list(real_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_splice = list(splicing_sub2[\"tid\"])\n",
    "        tpm_splice = list(splicing_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_int = list(intronic_sub2[\"tid\"])\n",
    "        tpm_int = list(intronic_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_intergenic = list(intergenic_sub2[\"tid\"])\n",
    "        tpm_intergenic = list(intergenic_sub2[\"tpm\"])\n",
    "\n",
    "        # now we can save the results\n",
    "        res_real = pd.concat([pd.DataFrame(ttx_real,columns=[\"tid_real\"]),pd.DataFrame(tpm_real,columns=[\"tpm\"])],axis=1)\n",
    "        realDF_f = realDF[[\"lid\",\"tid_real\"]].merge(res_real,on=\"tid_real\",how=\"inner\")\n",
    "        realDF_f = real_gtf.merge(realDF_f[[\"tid_real\",\"tpm\"]],on=\"tid_real\",how=\"inner\")\n",
    "        realDF_f[\"attributes\"] = realDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "                                    \";sim_tpm=\"+realDF_f[\"tpm\"].astype(str)\n",
    "        realDF_f[gff3cols].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        realDF_f[realDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "\n",
    "        res_splicing = pd.concat([pd.DataFrame(ttx_splice,columns=[\"tid_splicing\"]),pd.DataFrame(tpm_splice,columns=[\"tpm\"])],axis=1)\n",
    "        splicingDF_f = spliceDF[[\"lid\",\"tid_splicing\"]].merge(res_splicing,on=\"tid_splicing\",how=\"inner\")\n",
    "        splicingDF_f = splice_gtf.merge(splicingDF_f[[\"tid_splicing\",\"tpm\"]],on=\"tid_splicing\",how=\"inner\")\n",
    "        splicingDF_f[\"attributes\"] = splicingDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "            \";sim_tpm=\"+splicingDF_f[\"tpm\"].astype(str)\n",
    "        splicingDF_f[gff3cols].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        splicingDF_f[splicingDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "        splicingDF_f.head()\n",
    "\n",
    "        res_intronic = pd.concat([pd.DataFrame(ttx_int,columns=[\"tid_intronic\"]),pd.DataFrame(tpm_int,columns=[\"tpm\"])],axis=1)\n",
    "        intronicDF_f = intDF[[\"lid\",\"tid_intronic\"]].merge(res_intronic,on=\"tid_intronic\",how=\"inner\")\n",
    "        intronicDF_f = int_gtf.merge(intronicDF_f[[\"tid_intronic\",\"tpm\"]],on=\"tid_intronic\",how=\"inner\")\n",
    "        intronicDF_f[\"attributes\"] = intronicDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "            \";sim_tpm=\"+intronicDF_f[\"tpm\"].astype(str)\n",
    "        intronicDF_f[gff3cols].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        intronicDF_f[intronicDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "\n",
    "        res_intergenic = pd.concat([pd.DataFrame(ttx_intergenic,columns=[\"tid_intergenic\"]),pd.DataFrame(tpm_intergenic,columns=[\"tpm\"])],axis=1)\n",
    "        intergenicDF_f = polDF[[\"lid\",\"tid_intergenic\"]].merge(res_intergenic,on=\"tid_intergenic\",how=\"inner\")\n",
    "        intergenicDF_f = pol_gtf.merge(intergenicDF_f[[\"tid_intergenic\",\"tpm\"]],on=\"tid_intergenic\",how=\"inner\")\n",
    "        intergenicDF_f[\"attributes\"] = intergenicDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "                                    \";sim_tpm=\"+intergenicDF_f[\"tpm\"].astype(str)\n",
    "        intergenicDF_f[gff3cols].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        intergenicDF_f[intergenicDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
