{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the goal here is to create a better way of pre-selecting a pool of transcripts for the tissue\n",
    "# from within which the samples are going to be generated\n",
    "\n",
    "# currently, we have way too many transcripts in the tissue pool (while the number of loci is ok)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's begin putting together the transcriptome simulation set\n",
    "\n",
    "# What we need\n",
    "# number of transcripts to simulate (what is an average number of transcripts per tissue?)\n",
    "# number of loci (average for a tissue)\n",
    "# for each locus need to draw the number of transcripts of each category, such that the final distributions\n",
    "#     are close to tissue averages\n",
    "\n",
    "# then for each transcript need to draw from the distribution of total contribution to expression\n",
    "#     such that the final distributions resemble those observed in real data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10/sim_samples/\"\n",
    "\n",
    "num_tissues = 3\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>loading base annotations\n"
     ]
    }
   ],
   "source": [
    "# STAGE - 1 - generating sample specific gtfs\n",
    "\n",
    "# load base annotations\n",
    "print(\">>>loading base annotations\")\n",
    "real_baseDF = pd.read_csv(base_dir_data+\"real.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "real_baseDF = real_baseDF[(real_baseDF[\"type\"]==\"transcript\") & (real_baseDF[\"strand\"].isin([\"+\",\"-\"]))].reset_index(drop=True)\n",
    "\n",
    "splice_baseDF = pd.read_csv(base_dir_data+\"splicing.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "splice_baseDF = splice_baseDF[(splice_baseDF[\"type\"]==\"transcript\") & (splice_baseDF[\"strand\"].isin([\"+\",\"-\"]))].reset_index(drop=True)\n",
    "\n",
    "int_baseDF = pd.read_csv(base_dir_data+\"intronic.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "int_baseDF = int_baseDF[(int_baseDF[\"type\"]==\"transcript\") & (int_baseDF[\"strand\"].isin([\"+\",\"-\"]))].reset_index(drop=True)\n",
    "\n",
    "pol_baseDF = pd.read_csv(base_dir_data+\"RNApol.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "pol_baseDF = pol_baseDF[pol_baseDF[\"type\"]==\"transcript\"].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>getting loci IDs\n",
      "starting number of real noise loci: 40206\n",
      "starting number of splicing noise loci: 39058\n",
      "starting number of intronic loci: 27192\n",
      "starting number of polymerase loci: 638709\n"
     ]
    }
   ],
   "source": [
    "# get all loci and transcript IDs\n",
    "print(\">>>getting loci IDs\")\n",
    "real_baseDF[\"lid\"] = real_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "splice_baseDF[\"lid\"] = splice_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "int_baseDF[\"lid\"] = int_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "pol_baseDF[\"lid\"] = pol_baseDF[\"attributes\"].str.split(\"gene_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "real_locs = set(real_baseDF[\"lid\"])\n",
    "splice_locs = set(splice_baseDF[\"lid\"])\n",
    "int_locs = set(int_baseDF[\"lid\"])\n",
    "pol_locs = set(pol_baseDF[\"lid\"])\n",
    "print(\"starting number of real noise loci: \"+str(len(real_locs)))\n",
    "print(\"starting number of splicing noise loci: \"+str(len(splice_locs)))\n",
    "print(\"starting number of intronic loci: \"+str(len(int_locs)))\n",
    "print(\"starting number of polymerase loci: \"+str(len(pol_locs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of real transcripts: 301626\n",
      "number of real genes: 40206\n",
      "number of splicing transcripts: 11497993\n",
      "number of splicing genes: 39058\n",
      "number of intronic transcripts: 5839526\n",
      "number of intronic genes: 27192\n",
      "number of intergenic transcripts: 3109133\n",
      "number of intergenic genes: 638709\n"
     ]
    }
   ],
   "source": [
    "# perform cleanup, by removing any loci in \n",
    "#   1. int that are also in real\n",
    "#   2. splice that are not in real\n",
    "#   3. pol that are in real\n",
    "int_locs = int_locs - int_locs.difference(real_locs)\n",
    "assert(len(int_locs.difference(real_locs))==0),\"something wrong intronic\"\n",
    "int_baseDF = int_baseDF[int_baseDF[\"lid\"].isin(int_locs)].reset_index(drop=True)\n",
    "\n",
    "splice_locs = splice_locs - splice_locs.difference(real_locs)\n",
    "assert(len(splice_locs) == len(real_locs.intersection(splice_locs))),\"something wrong non-intronic\"\n",
    "splice_baseDF = splice_baseDF[splice_baseDF[\"lid\"].isin(splice_locs)].reset_index(drop=True)\n",
    "\n",
    "pol_locs = pol_locs - real_locs.intersection(pol_locs)\n",
    "assert(len(real_locs.intersection(pol_locs))==0),\"something wrong polymerase\"\n",
    "pol_baseDF = pol_baseDF[pol_baseDF[\"lid\"].isin(pol_locs)].reset_index(drop=True)\n",
    "\n",
    "real_baseDF[\"tid\"] = real_baseDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "real_baseDF = real_baseDF[[\"lid\",\"tid\"]]\n",
    "splice_baseDF[\"tid\"] = splice_baseDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "splice_baseDF = splice_baseDF[[\"lid\",\"tid\"]]\n",
    "int_baseDF[\"tid\"] = int_baseDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "int_baseDF = int_baseDF[[\"lid\",\"tid\"]]\n",
    "pol_baseDF[\"tid\"] = pol_baseDF[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "pol_baseDF = pol_baseDF[[\"lid\",\"tid\"]]\n",
    "\n",
    "print(\"number of real transcripts: \"+str(len(real_baseDF[\"tid\"])))\n",
    "print(\"number of real genes: \"+str(len(set(real_baseDF[\"lid\"]))))\n",
    "\n",
    "print(\"number of splicing transcripts: \"+str(len(splice_baseDF[\"tid\"])))\n",
    "print(\"number of splicing genes: \"+str(len(set(splice_baseDF[\"lid\"]))))\n",
    "\n",
    "print(\"number of intronic transcripts: \"+str(len(int_baseDF[\"tid\"])))\n",
    "print(\"number of intronic genes: \"+str(len(set(int_baseDF[\"lid\"]))))\n",
    "\n",
    "print(\"number of intergenic transcripts: \"+str(len(pol_baseDF[\"tid\"])))\n",
    "print(\"number of intergenic genes: \"+str(len(set(pol_baseDF[\"lid\"]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30161.48275862069 2415.8275862068963\n",
      "31389.793103448275 9445.264367816091\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD3CAYAAAAdfCMIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAQIklEQVR4nO3df6xfdX3H8edLUFsnxV5pxzJCb/hlEKugX3Udg8lEiVSWqJsbiajTrdnm2AiWgYhQnPxMQcGYabcaf2WbGLdobCZEh2HrCnKrdEwwyBCcZtCLFAoTEOp7f3w/d3xpeu/9trn3fu+P5yP55p7zOedzvp/TnJxXz+dzvuekqpAk6TmDboAkaXYwECRJgIEgSWoMBEkSYCBIkpr9B92AfXXQQQfV8PDwoJshSXPK1q1bH6yqZXtaNmcDYXh4mJGRkUE3Q5LmlCT3jbfMLiNJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWrm7A/TJM0vSfapnu90mToGgqRZYbwTexJP+jPELiNJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgT0+U7lJOcAw8CDwJHAe4HFwOXAPa3s/Kp6oGf9JcBS4Iaq+morPxZ4H/BDYDmwtqqeTrIIWA/8pG3r8qq6a4r2UZLUh0kDIcnBwAeAg6rqF0m+ArwVOAH4RlVdl+Q0uif0M5K8Fjipqk5N8lzgjiQ3AY8AXwBOrqr7k1wFvAvYCJwF/KiqrkyyspWdMPW7K0kaTz9dRj8Dfk73f/wALwS+B6wGtrSyzW0e4M1j5VX1FHAncCJwGLC4qu7fQ53VPXVuB16RZOz7/l+SNUlGkoyMjo72u4+SpD5MGghVtRM4B/hiks8APwbuptvl82hbbSewNMn+u5WPLVs+QTmTLOtty4aq6lRVZ9myZZPunCSpf5MGQuv3PwdYXVXvpjuOcCGwHTigrbYE2FFVT+9WPrZs+wTlTLJMkjQD+uky+lXgoXayB/gfYBGwCVjVyo5v8wBfGytvVwwvBW6iO/j8eBuT2L3Opp46K4Ft7cpEkjRD+rnL6OvAqW0Q+GHgZXQHgZ8ErkhyFHA4sBagqm5JcmOSS+neZXR2VT0MkOQdwCVJ7gP2Az7bvuMaYH2SC4Aj6N7FJEmaQamqQbdhn3Q6nRoZGRl0MyRNsyTM1fPUbJRka1V19rTMH6ZJkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJgP37WSnJS4DTgceB3wTWAduBDwF3A8PA+6vqsSTPAS4FHgNWABur6ua2nZOBt7a6VVUXt/Ih4HLgHuBI4PyqemBqdlGS1I9JAyHJfsDVwGlV9YsknwOeBj4PXFhV305yJnAu3YB4O7Ckqs5rJ/qbkxwNPB/4JHBMVT2Z5MtJXl9V36QbIN+oquuSnAasB86Yhv2VJI2jny6jVwMBzkzyAeA04GHgJODWts5mYHWbXg1sAaiqh4AngGOAVcB9VfXkRHV2K5ckzZB+uoxW0D2Zn15VjyT5AvBi4PGqqrbOTmB5m14OPNpTf2zZsnHKd6+zE1iaZP+qerq3IUnWAGsADj300D6aLknqVz9XCDuB71fVI23+34CXAYuTpJUtoTsuQPt7QE/9sWXjle9eZwmwY/cwAKiqDVXVqarOsmXL+mi6JKlf/QTCLcCL21gCdK8YvgfcSLc7CeB4YFOb3kT3imJssHhRW38LsCLJ8yeqs1u5JGmGTNplVFUPJTkX+FiSUbpdPx8G/g64MMkbgUOBs1uV64DjklzUyt9ZVbuAnyX5E+Datp3/aAPKAOcDVyQ5CjgcWDt1uyhJ6keeGQaYWzqdTo2MjAy6GZKmWRLm6nlqNkqytao6e1rmD9MkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCpBk2NDREkr4/wF6tn4ShoaEB7+XctP+gGyBpYdmxYwdVNa3fMRYk2jteIUiSAANBktQYCJIkwECQJDV9DyonWQzcAtxQVWuTLALWAz8BjgQur6q72rrvAI4DdgH/VVWfauXDwIeAu4Fh4P1V9ViS5wCXAo8BK4CNVXXzVOygJKk/e3OX0UeA7/bMnwX8qKquTLIS2AickOQQYC1wXFVVkluT/EtV/QD4JHBhVX07yZnAuXQD4u3Akqo6L8kQcHOSo6tq1xTsoySpD311GSU5A9gM/LCneDWwBaCqbgdekWQJcAqwtZ65r2wL8KYkzwVOAm5t5ZvbNnbf1kPAE8Ax+7hPkqR9MGkgJHkpcHRV/eNui5YDj/bM72xl45UfBDzeExRj5RNta/e2rEkykmRkdHR0sqZLkvZCP1cIbwGeSHIe8BvAa5KcBWwHDuhZb0krG6/8QWBxnvnFyFg5E9R5lqraUFWdquosW7asj6ZLkvo16RhCVV0yNt0Gkl9YVR9r06uAf21jCNuqameS64Ezk6RdDawCPl5VTyW5EXg18G3geGBT2/Qm4ETg820MYRHwvanbTUnSZPbmLqO30T1pPy/J6cA1wPokFwBHAO8FqKofJ1kPfDTJLuBv24AywB8DFyZ5I3AocHYrvw44LslFrfydDihL0szKdD9TZLp0Op0aGRkZdDMk7aUkM/Iso7l6bptuSbZWVWdPy/xhmiQJMBAkSY2BIEkCDARJUmMgSJIA35i24OzLm6S8W0NaGAyEBWa8k7u36Umyy0iSBBgIkqTGQJAkAQaCJKkxEOahoaEhkuzVB9ir9YeGhga8l5KmmncZzUM7duyYkYeHSZpfvEKQJAEGgiSpsctoHqqLlsC6A6f/OyTNKwbCPJSLd87MC0jWTetXSJphdhlJkgADQZLUGAiSJMBAkCQ1BoIkCfAuo3lrun9JvHTp0mndvqSZZyDMQ/tyy6kvyJFkl5EkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2PrlhgJnrG0XjLfKSFtDAYCAuMJ3dJ47HLSJIEGAiSpGbSLqMkhwMfAb4DHAL8tKo+nGQIuBy4BzgSOL+qHmh1zgGWAEuBG6rqq638WOB9wA+B5cDaqno6ySJgPfCTtq3Lq+quKd1TSdKE+hlDGAL+oaq+ApDkjiSbgD8CvlFV1yU5je4J/YwkrwVOqqpTkzwXuCPJTcAjwBeAk6vq/iRXAe8CNgJnAT+qqiuTrGxlJ0zxvkqSJjBpl1FV3ToWBj11/hdYDWxpZZvbPMCbx8qr6ingTuBE4DBgcVXdv4c6q3vq3A68IsmS3duSZE2SkSQjo6Ojfe+kJGlye3WXUZK3ANdX1feTLAcebYt2AkuT7E+3K+jOnmo7W9loz/q95bS/e1q2s/f7q2oDsAGg0+l4u4w0B9VFS2DdgdP/HdprfQdCkpOAk+h27wBsBw4AHqY7XrCjjQeMlY9Z0tYdr5xJlkmaR3Lxzmm//TkJtW5av2Je6usuoySrgVOAvwAOTrIK2ASsaqsc3+YBvjZW3q4YXgrcRHfw+fEkB++hzqaeOiuBbVX1rKsDSdL06ucuo1cBXwRGgBuBXwI+AZwPXJHkKOBwYC1AVd2S5MYkl9K9y+jsqnq4besdwCVJ7gP2Az7bvuYaYH2SC4AjgPdO3S5KkvqRufrL1U6nUyMjI4NuhqS9lGRmuozm6LltuiXZWlWdPS3zh2mSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUrNXr9CUpKmQZFq3v3Tp0mnd/nxlIEiaUXv7ngLfbTBz7DKSJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBvjFN0iwx0Ws1J1rm29SmjoEgaVbwxD54dhlJkgADQZLUzJouoyQnA28FtgNVVRcPuEmStKDMikBI8gLgk8AxVfVkki8neX1VfXPQbZOkhWK2dBmtAu6rqifb/GZg9QDbI0kLzmwJhOXAoz3zO1vZsyRZk2Qkycjo6OiMNU6SFoLZEgjbgQN65pe0smepqg1V1amqzrJly2ascZK0EMyWQNgCrEjy/DZ/PLBpgO2RpAUns+XHIEneAPwOMAo8NdldRklGgftmom0LxEHAg4NuhLQHHptTa0VV7bGLZdYEggYryUhVdQbdDml3HpszZ7Z0GUmSBsxAkCQBBoKesWHQDZDG4bE5QxxDkCQBXiFIkhoDQZIEGAiaRJI/TXLvoNuh2as9Tma/CZa/KMm7Z7BJ47XjPUnOHHQ7ZjPHEDSpJPdW1fCg26HZKUlqghNJkmHgM1X1uplq03gma+tCNysef63pk+QPgMuAvwYOAV4F/DvdZ0UdCNxWVZ9PchhwdVu2Eriqqm4bTKs1VyT5beDaJKfTPcY2Az8Hfg04s6pGgDXAcJJ1wNeBO4CPAT+ge0x+taquT3IFcDrwN8BrgbuBa9pnG92HXn4Q+POq+lySD9M9h+0CHq2qK3uO948ChwFHA29uzb22/X13kgOAq4B7gF8GHqqqv5qGf6K5par8zPMP8C3g1DZ9LrChTQf4Pt0nyx4CvLKVvxL4Uk/9ewe9D35m76cdX8PAOuCyVva7wMfb9DDwrZ71LwPOb9OL6T6CZv82/wSwFNgPOBb4EvB7bdkRY8cicApww25tOLZn+pQ2/QngbW36dXSvVMba8Jc99d8z6H/H2fDxCmHhuLP9PRT4lSTntfn/BA4GHgB+P8mb6D5t1sfJal/c1f6O8uwnGPd6OfDTnmPwdmCI7lXrA1W1o5XfluQYuidv6P5vvncbL+jZxn/z7GN2sna8HNg4NlNVn55opxYKA2HhGOs33QY8WVWXAyR5C3AvcDGwo6ouSXIU3Ut2aW/tqX9+F92rUZIcS/cYvL+qrm1lZwA/Haf+HcBRwHfodgGN2Qas6jmOf4tuF9NE7ei1DTi81Q2wpqo+NUmdec9AmOfaU2RXAH/W+mg3Ale2/tznAY9X1T8l+TJwWXsE+fPoPo789cBLgAOTnF5Vfz+YvdBslWQ13ePrauBFwMok1wNnAC9P0gFuA55IcjXdE/xldI/BC+iOY91TVbuS/CHdY+3sqrq6fcW5wMeTrKR7FVAAVXVDktckuQx4GlgEnNdzvL8nyWeAE1ubvtnTpl9vbbgqyQdbu/95Ov+d5grvMpI0ayU5ku6V64NJDgU+XVUnD7pd85VXCJJmsxXAxUm+S3dQ+ewBt2de8wpBkgT4S2VJUmMgSJIAA0GS1BgIkiTAQJAkNf8H0RBPXBRo3VIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the distribution of the number of real and noise loci per tissue first\n",
    "t_loc = pd.read_csv(base_dir_out+\"res.num_locs_tissue\")\n",
    "\n",
    "# now we shall remove any outliers from the data\n",
    "q25,q50,q75 = t_loc['real'].quantile([0.25,0.5,0.75])\n",
    "iqr = q75-q25\n",
    "thw = q75+1.5*iqr\n",
    "tlw = q25-1.5*iqr\n",
    "ahw = t_loc[t_loc[\"real\"]<=thw][\"real\"].max()\n",
    "alw = t_loc[t_loc[\"real\"]>=tlw][\"real\"].min()\n",
    "t_loc = t_loc[(t_loc['real']<=ahw)&(t_loc['real']>=alw)]\n",
    "\n",
    "q25,q50,q75 = t_loc['intergenic'].quantile([0.25,0.5,0.75])\n",
    "iqr = q75-q25\n",
    "thw = q75+1.5*iqr\n",
    "tlw = q25-1.5*iqr\n",
    "ahw = t_loc[t_loc[\"intergenic\"]<=thw][\"intergenic\"].max()\n",
    "alw = t_loc[t_loc[\"intergenic\"]>=tlw][\"intergenic\"].min()\n",
    "t_loc = t_loc[(t_loc['intergenic']<=ahw)&(t_loc['intergenic']>=alw)]\n",
    "\n",
    "std_real = (t_loc[\"real\"].mean()-t_loc[\"real\"].min())/3\n",
    "std_intergenic = (t_loc[\"intergenic\"].mean()-t_loc[\"intergenic\"].min())/3\n",
    "print(t_loc[\"real\"].mean(),std_real)\n",
    "print(t_loc[\"intergenic\"].mean(),std_intergenic)\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "ax.boxplot(t_loc.T,labels=t_loc.T.index.tolist())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "selecting 31218 real loci\n",
      "selecting 29016 intergenic loci\n",
      "number of all real loci: 40206\n",
      "number of real transcripts: 234697\n",
      "number of real genes: 31218\n",
      "number of splicing transcripts: 8916670\n",
      "number of splicing genes: 30323\n",
      "number of intronic transcripts: 4507952\n",
      "number of intronic genes: 21180\n",
      "number of intergenic transcripts: 148704\n",
      "number of intergenic genes: 29016\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "selecting 29592 real loci\n",
      "selecting 39563 intergenic loci\n",
      "number of all real loci: 40206\n",
      "number of real transcripts: 223113\n",
      "number of real genes: 29592\n",
      "number of splicing transcripts: 8510062\n",
      "number of splicing genes: 28777\n",
      "number of intronic transcripts: 4339048\n",
      "number of intronic genes: 20000\n",
      "number of intergenic transcripts: 182621\n",
      "number of intergenic genes: 39563\n",
      "\n",
      "=================\n",
      "Tissue #2\n",
      "=================\n",
      "\n",
      "selecting 25849 real loci\n",
      "selecting 23659 intergenic loci\n",
      "number of all real loci: 40206\n",
      "number of real transcripts: 192745\n",
      "number of real genes: 25849\n",
      "number of splicing transcripts: 7317687\n",
      "number of splicing genes: 25111\n",
      "number of intronic transcripts: 3789525\n",
      "number of intronic genes: 17469\n",
      "number of intergenic transcripts: 123172\n",
      "number of intergenic genes: 23659\n"
     ]
    }
   ],
   "source": [
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    # instead of sampling from the actual values we can get it from a distribution (assuming 2 independent gaussians)\n",
    "    ctlr = int(np.random.normal(t_loc[\"real\"].mean(),std_real,1)[0])\n",
    "    ctln = int(np.random.normal(t_loc[\"intergenic\"].mean(),std_intergenic,1)[0])\n",
    "\n",
    "    print(\"selecting \"+str(ctlr)+\" real loci\")\n",
    "    print(\"selecting \"+str(ctln)+\" intergenic loci\")\n",
    "    \n",
    "    all_real_locs = real_locs.union(splice_locs.union(int_locs))\n",
    "    print(\"number of all real loci: \"+str(len(all_real_locs)))\n",
    "\n",
    "    all_real_locs_rand = np.random.choice(list(all_real_locs),ctlr, replace=False)\n",
    "    pol_locs_rand = np.random.choice(list(pol_locs),ctln, replace=False)\n",
    "\n",
    "    real_baseDF_sub1 = real_baseDF[real_baseDF[\"lid\"].isin(all_real_locs_rand)].reset_index(drop=True)\n",
    "    splice_baseDF_sub1 = splice_baseDF[splice_baseDF[\"lid\"].isin(all_real_locs_rand)].reset_index(drop=True)\n",
    "    int_baseDF_sub1 = int_baseDF[int_baseDF[\"lid\"].isin(all_real_locs_rand)].reset_index(drop=True)\n",
    "    pol_baseDF_sub1 = pol_baseDF[pol_baseDF[\"lid\"].isin(pol_locs_rand)].reset_index(drop=True)\n",
    "\n",
    "    real_baseDF_sub1.to_csv(out_dir+\"stage1_tissue_locs.real_t\"+str(tissue_num),index=False)\n",
    "    splice_baseDF_sub1.to_csv(out_dir+\"stage1_tissue_locs.splice_t\"+str(tissue_num),index=False)\n",
    "    int_baseDF_sub1.to_csv(out_dir+\"stage1_tissue_locs.int_t\"+str(tissue_num),index=False)\n",
    "    pol_baseDF_sub1.to_csv(out_dir+\"stage1_tissue_locs.pol_t\"+str(tissue_num),index=False)\n",
    "\n",
    "    print(\"number of real transcripts: \"+str(len(real_baseDF_sub1[\"tid\"])))\n",
    "    print(\"number of real genes: \"+str(len(set(real_baseDF_sub1[\"lid\"]))))\n",
    "\n",
    "    print(\"number of splicing transcripts: \"+str(len(splice_baseDF_sub1[\"tid\"])))\n",
    "    print(\"number of splicing genes: \"+str(len(set(splice_baseDF_sub1[\"lid\"]))))\n",
    "\n",
    "    print(\"number of intronic transcripts: \"+str(len(int_baseDF_sub1[\"tid\"])))\n",
    "    print(\"number of intronic genes: \"+str(len(set(int_baseDF_sub1[\"lid\"]))))\n",
    "\n",
    "    print(\"number of intergenic transcripts: \"+str(len(pol_baseDF_sub1[\"tid\"])))\n",
    "    print(\"number of intergenic genes: \"+str(len(set(pol_baseDF_sub1[\"lid\"]))))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
