{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10/sim_samples/\"\n",
    "hg38_fa = \"/home/avaraby1/genomes/human/hg38/hg38_p8.fa\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "readlen = 101\n",
    "\n",
    "gffread_path = \"gffread\"\n",
    "genRNAseq_path = \"/ccb/salz8-1/avaraby/tx_noise/genRNAseq_noFold_cov2.R\"\n",
    "shuffleReads_path = \"/ccb/salz8-1/avaraby/tx_noise/shuffleReads.sh\"\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# first extract fasta sequences for each annotation\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        cmd_real = [gffread_path,\n",
    "                    \"-w\",out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    \"-g\",hg38_fa,\n",
    "                    out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        subprocess.call(cmd_real)\n",
    "\n",
    "        cmd_nonint = [gffread_path,\n",
    "                    \"-w\",out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    \"-g\",hg38_fa,\n",
    "                    out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        subprocess.call(cmd_nonint)\n",
    "\n",
    "        cmd_int = [gffread_path,\n",
    "                    \"-w\",out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    \"-g\",hg38_fa,\n",
    "                    out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        subprocess.call(cmd_int)\n",
    "\n",
    "        cmd_pol = [gffread_path,\n",
    "                    \"-w\",out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    \"-g\",hg38_fa,\n",
    "                    out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        subprocess.call(cmd_pol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# simulate reads\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        if not os.path.exists(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        cmd_real = [genRNAseq_path,\n",
    "                    out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",\n",
    "                    str(readlen),\n",
    "                    out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/\",\n",
    "                    \"0\"]\n",
    "        subprocess.call(cmd_real)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        cmd_nonint = [genRNAseq_path,\n",
    "                    out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",\n",
    "                    str(readlen),\n",
    "                    out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/\",\n",
    "                    \"0\"]\n",
    "        subprocess.call(cmd_nonint)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        cmd_int = [genRNAseq_path,\n",
    "                    out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",\n",
    "                    str(readlen),\n",
    "                    out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/\",\n",
    "                    \"0\"]\n",
    "        subprocess.call(cmd_int)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        cmd_pol = [genRNAseq_path,\n",
    "                    out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".fasta\",\n",
    "                    out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",\n",
    "                    str(readlen),\n",
    "                    out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/\",\n",
    "                    \"0\"]\n",
    "        subprocess.call(cmd_pol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        shuffle_cmd = [shuffleReads,\n",
    "                       out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                       out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(shuffle_cmd)\n",
    "\n",
    "        # create a combination of real and nonint\n",
    "        if not os.path.exists(out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        with open(out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\", 'w+') as outfile:\n",
    "            for fname in [out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                          out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\"]:\n",
    "                with open(fname) as infile:\n",
    "                    for line in infile:\n",
    "                        outfile.write(line)\n",
    "        shuffle_cmd = [shuffleReads,\n",
    "                       out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                       out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(shuffle_cmd)\n",
    "\n",
    "        # create a combination of real and int\n",
    "        if not os.path.exists(out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        with open(out_dir+\"real_intronic.sample\"+str(i)+\"/sample_01.fasta\", 'w+') as outfile:\n",
    "                for fname in [out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                              out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\"]:\n",
    "                    with open(fname) as infile:\n",
    "                        for line in infile:\n",
    "                            outfile.write(line)\n",
    "        shuffle_cmd = [shuffleReads,\n",
    "                       out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                       out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(shuffle_cmd)\n",
    "\n",
    "        # create a combination of real and polymerase\n",
    "        if not os.path.exists(out_dir+\"real_intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"real_intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        with open(out_dir+\"real_intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\", 'w+') as outfile:\n",
    "            for fname in [out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                          out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\"]:\n",
    "                with open(fname) as infile:\n",
    "                    for line in infile:\n",
    "                        outfile.write(line)\n",
    "        shuffle_cmd = [shuffleReads,\n",
    "                       out_dir+\"real_intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                       out_dir+\"real_intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(shuffle_cmd)\n",
    "\n",
    "        # create a combination of all reads\n",
    "        if not os.path.exists(out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        with open(out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\", 'w+') as outfile:\n",
    "            for fname in [out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                          out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                          out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                          out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\"]:\n",
    "                with open(fname) as infile:\n",
    "                    for line in infile:\n",
    "                        outfile.write(line)\n",
    "        shuffle_cmd = [shuffleReads,\n",
    "                       out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.fasta\",\n",
    "                       out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(shuffle_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine annotations for processing with sim2sam\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        merge_cmd = \"cat \"+out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           \" > \"+out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"\n",
    "        subprocess.call(merge_cmd,shell=True)\n",
    "        \n",
    "        merge_cmd = \"cat \"+out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           \" > \"+out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"\n",
    "        subprocess.call(merge_cmd,shell=True)\n",
    "        \n",
    "        merge_cmd = \"cat \"+out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"+\\\n",
    "                           \" > \"+out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"\n",
    "        subprocess.call(merge_cmd,shell=True)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10/sim_samples/\"\n",
    "idx_dir = \"/ccb/salz8-1/avaraby/tx_noise/indices/\"\n",
    "annotation_path = \"/ccb/salz3/avaraby/chess/versions/curRelease/chess2.2_assembly.gtf\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "readlen = 101\n",
    "\n",
    "hisat_path = \"/ccb/salz8-1/avaraby/agar_manuscript/soft/AGAR2/external/hisat2/hisat2\"\n",
    "kallisto_path = \"/home/avaraby1/soft/kallisto/kallisto\"\n",
    "salmon_path = \"/home/avaraby1/soft/salmon-latest_linux_x86_64/bin/salmon\"\n",
    "sim2sam_path = \"/ccb/salz8-1/avaraby/tx_noise/sim2sam/sim2sam\"\n",
    "stringtie2_path = \"stringtie\"\n",
    "\n",
    "num_threads = 24\n",
    "\n",
    "# convert with sim2sam\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        if not os.path.exists(out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        plst_cmd = [sim2sam_path,\n",
    "                    \"-i\",hg38_fa+\".fai\",\n",
    "                    \"-o\",out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sam\",\n",
    "                    \"-s\",out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\",\n",
    "                    \"-g\",out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        plst_cmd.extend([\"-t\",\"polyester\"])\n",
    "        subprocess.call(plst_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(num_threads),\n",
    "                    \"-o\",out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sorted.bam\",\n",
    "                    out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        plst_cmd = [sim2sam_path,\n",
    "                    \"-i\",hg38_fa+\".fai\",\n",
    "                    \"-o\",out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sam\",\n",
    "                    \"-s\",out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\",\n",
    "                    \"-g\",out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        plst_cmd.extend([\"-t\",\"polyester\"])\n",
    "        subprocess.call(plst_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(num_threads),\n",
    "                    \"-o\",out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sorted.bam\",\n",
    "                    out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sim2sam.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        plst_cmd = [sim2sam_path,\n",
    "                    \"-i\",hg38_fa+\".fai\",\n",
    "                    \"-o\",out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sam\",\n",
    "                    \"-s\",out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\",\n",
    "                    \"-g\",out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        plst_cmd.extend([\"-t\",\"polyester\"])\n",
    "        subprocess.call(plst_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(num_threads),\n",
    "                    \"-o\",out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sorted.bam\",\n",
    "                    out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "\n",
    "        if not os.path.exists(out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)):\n",
    "            os.makedirs(out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num))\n",
    "        plst_cmd = [sim2sam_path,\n",
    "                    \"-i\",hg38_fa+\".fai\",\n",
    "                    \"-o\",out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sam\",\n",
    "                    \"-s\",out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\",\n",
    "                    \"-g\",out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\"]\n",
    "        plst_cmd.extend([\"-t\",\"polyester\"])\n",
    "        subprocess.call(plst_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(num_threads),\n",
    "                    \"-o\",out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sorted.bam\",\n",
    "                    out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"sim2sam.sam\"]\n",
    "        subprocess.call(sort_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align with hisat2\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        hisat_cmd = [hisat,\n",
    "                     \"-x\",idx_dir+\"annotation.hisat\",\n",
    "                     \"-p\",str(threads),\n",
    "                     \"--rna-sensitive\",\"-f\",\n",
    "                     \"-S\",out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\",\n",
    "                     \"-U\",out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(hisat_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(threads),\n",
    "                    \"-o\",out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sorted.bam\",\n",
    "                    out_dir+\"strg.real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "        \n",
    "        hisat_cmd = [hisat,\n",
    "                     \"-x\",idx_dir+\"annotation.hisat\",\n",
    "                     \"-p\",str(threads),\n",
    "                     \"--rna-sensitive\",\"-f\",\n",
    "                     \"-S\",out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\",\n",
    "                     \"-U\",out_dir+\"real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(hisat_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(threads),\n",
    "                    \"-o\",out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sorted.bam\",\n",
    "                    out_dir+\"strg.real_splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "        \n",
    "        hisat_cmd = [hisat,\n",
    "                     \"-x\",idx_dir+\"annotation.hisat\",\n",
    "                     \"-p\",str(threads),\n",
    "                     \"--rna-sensitive\",\"-f\",\n",
    "                     \"-S\",out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\",\n",
    "                     \"-U\",out_dir+\"real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(hisat_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(threads),\n",
    "                    \"-o\",out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sorted.bam\",\n",
    "                    out_dir+\"strg.real_intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "        \n",
    "        hisat_cmd = [hisat,\n",
    "                     \"-x\",idx_dir+\"annotation.hisat\",\n",
    "                     \"-p\",str(threads),\n",
    "                     \"--rna-sensitive\",\"-f\",\n",
    "                     \"-S\",out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\",\n",
    "                     \"-U\",out_dir+\"all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(hisat_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(threads),\n",
    "                    \"-o\",out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sorted.bam\",\n",
    "                    out_dir+\"strg.all.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\"/hisat2.sam\"]\n",
    "        subprocess.call(sort_cmd)        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
