{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/sim_samples/\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "real_gtf = pd.read_csv(base_dir_data+\"real.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "real_gtf[\"tid_real\"] = real_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "splice_gtf = pd.read_csv(base_dir_data+\"splicing.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "splice_gtf[\"tid_splicing\"] = splice_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "int_gtf = pd.read_csv(base_dir_data+\"intronic.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "int_gtf[\"tid_intronic\"] = int_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]\n",
    "pol_gtf = pd.read_csv(base_dir_data+\"RNApol.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "pol_gtf[\"tid_intergenic\"] = pol_gtf[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True,n=1)[1].str.split(\"\\\"\",expand=True,n=1)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAD4CAYAAAAdIcpQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de5xddXnv8c+XSCVcAjNlYrAQIgItTCYEGAoYCgSQtATDsbYqCEpJT+qRRiqCoYRLUAKo4KXRqrSxqMgU0XMOlFQTdMaDICgTC8FwE0MUqJCJiYZE0DA+54/128mayZ5Ze2b2ZO+Z+b5fr3nN2r91e9ZvXZ51X4oIzMzM+rNLrQMwM7P652RhZmaFnCzMzKyQk4WZmRVysjAzs0JOFmZmVug1tQ5gsPbdd9+YMmVKrcMwMxtRVq5cuT4imgba34hNFlOmTKGzs7PWYZiZjSiSfjaY/gqThaQ3AtcCPwL2B34ZER+W1AjcAKwBDgEuj4gXUz+XAhOABmBFRNyVyqcDFwLPABOBSyLiVUm7ATcCz6dh3RARTw1mgszMrPoqObJoBP49Iu4EkPSYpGXA/wS+HRFfk/QWso39eZKOBWZGxBmSdgUek3Qv8GvgVuC0iHhB0k3Ae4ClwD8AP4+Ij0lqSWV/VuVpNTOzQSq8wB0RD5USRa6fLcBs4IFUdn/6DXBmqTwitgKPAycCBwHjI+KFMv3MzvXzKHCEpAmDnCYzM6uyAd0NJemtwPKIeILsNNJLqdUmoEHSa3qVl9pN7Kecgnb58c+T1Cmps6urayChm5nZEFScLCTNBGYCH0hF64C9UvMEYGNEvNqrvNRuXT/lvYfVu902EXFzRLRGRGtT04Av5puZ2SBVlCwkzQZmARcBkyQdDywDjk+dzEi/Ae4ulacjjcOBe8kuhL8saVKZfpbl+mkBHomITYOfLDOzgWlra2Pq1KmMGzeOqVOn0tbWVuuQ6kold0MdDdwOdAIdwB7AZ4HLgY9KOhR4I3AJQET8QFKHpOvI7oa6OCJ+lYZ1LrA43bo1DvhSGs2ngRslXQEcDMyt3iSamfWvra2NhQsXsnTpUk444QTuu+8+5s7NNkNnn312jaOrDxqp37NobW0NP2dhZtUwdepUlixZwsyZM7eVdXR0MH/+fH784x/XMLLqk7QyIloH2t+IfSjPzKwaJAFwyimn9Nt+pO5YV4vfDWVmY1pE0NzcTHt7+7aEEBG0t7fT3NxMRIz5RAFOFmZmLFy4kLlz59LR0QFkp6Dmzp3LwoULaxxZ/fBpKDMb80oXsefPn7/t/+LFi31xO8cXuM3MciSN6tNOg73A7dNQZmZWyMnCzMwKOVmYmVkhJwszMyvkZGFmZoWcLMzMrJCThZmZFXKyMDOzQk4WZmZWyMnCzMwKOVmYmVkhJwszMytUyWdVJwHXAkdExDGpbCnZp1RLpgFHRcRaSWuBtan8+Yh4V+pnCnAl8DQwBfhgRGyWtAtwHbAZOBBYGhEPDnXCzMyseip5RfkJwJ3A9FzZioi4HUDSBOCWiFib2t0SEYvKDOfzwFUR8UNJ84EFZMnj7cCEiLhMUiPwoKTDIqJ7UFNkZmZVV3gaKiK+DrzUq+z23M+5wBdzv0+U9CFJH5H0JgBJuwIzgYdSN/cDs1PzbOCBNNwNwCtA88AnxczMhsuQPn6UTiHNAj6VK74sHT3sDvxI0pnAFuDl2P6S+E3AxNQ8kZ7JKN+u9/jmAfMAJk+ePJTQzcxsAIZ6gfss4O5cEiAifpj+/wZ4GJgBrAfGq/Tlc5gArEvN64C9csPMt+shIm6OiNaIaG1qahpi6GZmVqmhJovzgVtKPySdKunPc+0PBn4aEVuBDuCYVD4DWJaalwHHp/4bgd2A1UOMy8zMqqiSu6FOAs4D9pN0BXBTRLwsaTrwVERsznW+Dlgk6Sjg9cA3IuK+1O69wFWSTgcmAxen8q8BR0q6OpW/2xe3zczqi7/BbWZjRmNjIxs3bqzKsBoaGtiwYUNVhrUzDfYb3EO6wG1mNpJs3LiRau0gb78EOzb4CW4zMyvkZGFmZoWcLMzMrJCThZmZFXKyMDOzQk4WZmZWyLfOmtmYEVdPgEV7V29YY4iThZmNGbpmU1Wfsyj7MYZRyqehzMyskJOFmZkVcrIwM7NCThZmZlbIycLMzAo5WZiZWSHfOmtmY0q1Xi3e0NBQleGMFE4WZjZmVPKMhaSqPYsxmlTyWdVJwLXAERFxTCo7n+wzqa+kzpZGxFdSu3OBI4Fusu9vfyGVTwGuBJ4GpgAfjIjNknYBrgM2AwemYT1YnckzM7NqqOTI4gTgTmB6r/J3RsTafIGk/YFLgCMjIiQ9JKk9In4CfB64KiJ+KGk+sIAsebwdmBARl0lqBB6UdJi/w21mVj8KL3BHxNeBl8q0+ntJl0i6Km3kAWYBK2P7MdwDwF9I2hWYCTyUyu8HZqfm2ak7ImID2dFK82AmxszMhsdgr1n8P2BZRHRJOgO4AzgVmEjPxLIple0LvJxLIqVy+ulnB5LmAfMAJk+ePMjQzcxsoAZ162xEPBMRXelnO3CSpHHAOmCvXKcTUtl6YLy234ZQKqeffsqN9+aIaI2I1qampsGEbmZmgzCoZCHpekmlo5JDgGfSNYblwNG5pHA88M2I2Ap0AMek8hnAstS8LHVHOp21G7B6MHGZmdnwqORuqJOA84D9JF0B3AS8AHxO0jNAS2pPRDwn6Ubgk5K6gX9NF7chu3vqKkmnA5OBi1P514AjJV2dyt/ti9tmZvVFI/V+4tbW1ujs7Kx1GGY2yoz25ywkrYyI1oH259d9mJlZIScLMzMr5GRhZmaFnCzMzKyQk4WZmRVysjAzs0JOFmZmVsjJwszMCjlZmJlZIScLMzMr5GRhZmaFnCzMzKzQYD9+ZGY2qmz/ssL25tH8QsGB8pGFmY15+URRSflY5GRhZpaMGzeux3/bzsnCzMa0/NFDd3d3j/+l9j7C8DULMxvjIqLsNQpft+jJRxZmZokkvvrVr/pIooxKvsE9CbgWOCIijkllC4BJZN/iPhq4KiKeSO3WAmtT789HxLtS+RTgSuBpYArwwYjYLGkX4DpgM3AgsDQiHqzK1JmZDdC5555b6xDqUiWnoU4A7gSm58r2BC6OiJD0DuDjwFtSu1siYlGZ4XyeLKn8UNJ8YAFZ8ng7MCEiLpPUCDwo6bCI6C4zDDOzqsufiupdbpnC01AR8XXgpV5lV8b2WtyF7Kig5ERJH5L0EUlvApC0KzATeCh1cz8wOzXPBh5Iw90AvAI0D25yzMwGJyJ2+LPthnSBW9IfAO8BLswVX5aOHnYHfiTpTGAL8HIuwWwCJqbmifRMRvl2vcc3D5gHMHny5KGEbmZmAzDoC9wpUXwOWBgRPy2VR8QP0//fAA8DM4D1wHhtP86bAKxLzeuAvXKDzrfrISJujojWiGhtamoabOhmZjZAg0oWksYDXwA+ERErJb0tlZ8q6c9znR4M/DQitgIdwDGpfAawLDUvA45P/TcCuwGrBxOXmZkNj0ruhjoJOA/YT9IVwE3AV4GpwBvSwcIewDfIjggWSToKeD3wjYi4Lw3qvcBVkk4HJgMXp/KvAUdKujqVv9sXt83M6otG6kWc1tbW6OzsrHUYZmYjiqSVEdE60P78UJ6ZmRVysjAzs0JOFmZmVsjJwszMCjlZmJlZIScLMzMr5GRhZmaFnCzMzKyQk4WZmRVysjAzs0JOFmZmVmhMJ4u2tjamTp3KuHHjmDp1Km1tbbUOycysLo3ZZNHW1sZFF13Eli1biAi2bNnCRRdd5IRhZlbGmH3r7AEHHMAvfvELuru3vw193Lhx7Lfffjz77LPVCNHMrO74rbMD9Nxzz9Hd3c2cOXPo6upizpw5dHd389xzz9U6NDOzujNmkwXAcccdx5133sm+++7LnXfeyXHHHVfrkMzM6tKYThZr1qyho6ODrVu30tHRwZo1a2odkplZXarks6qTgGuBIyLimFS2G3Aj8DxwCHBDRDyV2p0LHAl0k31/+wupfApwJfA0MAX4YERslrQLcB2wGTgQWBoRD1ZvEneYnm3N69at45RTTinbzUi9lmNmNhwqObI4AbgTUK7sH4CfR8T1wCeBpQCS9gcuAS6JiA8BfyvpkNTP54EvpH5+DCxI5W8HJkTEtansy5LGDW2y+hYRRAQtLS0A7Lnnnj3+t7S0OFGYmfVSmCwi4uvAS72KZwMPpPaPAkdImgDMAlbG9q3tA8BfSNoVmAk8lMrvT8PoPawNwCtA82AnqFKrVq2ipaWFzZs3A7B582ZaWlpYtWrVcI/azGzEKTwN1YeJ9Ewgm1JZX+X7Ai/nkkipvL9h7UDSPGAewOTJkwcZ+nalxODTTmZm/RvsBe51wF653xNSWV/l64Hx2n7BoFTe37B2EBE3R0RrRLQ2NTUNMnQzMxuowSaLZcDxAJJagEciYhOwHDg6lxSOB74ZEVuBDuCYVD4jDaP3sBqB3YDVg4zLzMyGQSV3Q50EnAfsJ+kK4Cbg08CN6ffBwFyAiHhO0o3AJyV1A/8aET9Jg3ovcJWk04HJwMWp/GvAkZKuTuXvjojtj1WbmVnNjdnXfeT5moWZjRV+3YeZmQ0bJwszMyvkZGFmZoWcLMzMrJCThZmZFXKyMDOzQqM+WTQ2NiKp3z+gsJvGxsYaT4mZWe0M9t1QI8bGjRur8gxF/tXmZmZjzag/sjAzs6FzsjAzs0JOFmZmVmjUX7OIqyfAor2rMxwzszFq1CcLXbOpahe4Y9HQ4zEzG4l8GsrMzAo5WZiZWSEnCzMzKzTqr1lAdR6oa2hoqEIkZmYj06hPFpVc3PaX8szM+jekZCFpCvAd4NlUNAFYBawFTs51ujgi7kn9XJq6awBWRMRdqXw6cCHwDDARuCQiXh1KfGZmVh1DPbJ4Cfi7iPg2gKRrgHuA0yLi5N4dSzoWmBkRZ0jaFXhM0r3Ar4FbU38vSLoJeA+wdIjxmZlZFQzpAndE/DKXKF4LtEbEfen3QkmXSFogaffUy5nAA6nfrcDjwInAQcD4iHghdXc/MHsosZmZWfVU826oc4C21HwH8KmIuJHs6GNJKp+YfpdsSmV9lfcgaZ6kTkmdXV1dVQzdzMz6U81k8dfA7QARsToitqTyduCU1LwO2CvXz4RU1ld5DxFxc0S0RkRrU1NTFUM3M7P+VCVZSJoJfD+dWkLSx3OtDwGeTs13A8enbl4DHA7cC6wBXpY0KXU3A1hWjdjMzGzoqnXr7Dxgfu73q5I+TXZ00EJ2lxMR8QNJHZKuI7sb6uKI+BWApHOBxZJ+BowDvlSl2MzMbIg0Up8vaG1tjc7OzqoMy89ZmNlYIWllRLQOtD+/7sPMzAo5WZiZWaExnSxmzZrFLrtkVbDLLrswa9asGkdkZlafxmyymDVrFitWrNh2rSIiWLFihROGmVkZYzZZrFixYkDlZmZj2ZhLFpIKX1lejVeam5mNJmMuWURE4W2yvo3WzKynMZcszMxs4Eb9x49sdKn0FKGPDs2qy8nCRpTeScBP35vtHD4NZWZmhZwszMyskJOFmZkVcrIwM7NCThZmZlbIycLMzAqN2WRx+umnD6jczGwsG/JzFpIeBF5JP7sj4lRJjcANZN/WPgS4PCJeTN1fCkwg+6zqioi4K5VPJ/v86jPAROCSiHh1qPH1Zfny5cyaNYt77rmHiEASb37zm1m+fPlwjdLMbMSqxkN534qIRb3KrgO+HRFfk/QW4EbgPEnHAjMj4gxJuwKPSboX+DVwK3BaRLwg6SbgPcDSKsTXJycGM7PKVOM0VIukBZIWSZqdymYDD6Tm+9NvgDNL5RGxFXgcOBE4CBgfES+U6cfMzGqsGsnioxHxUeAjwOWSTiQ7jfRSar8JaJD0ml7lpXYT+ynvQdI8SZ2SOru6uqoQev0pfb1Pkr/eZ2Z1Y8jJIiJ+mP53A98DZgLrgL1SJxOAjen6Q7681G5dP+W9x3VzRLRGRGtTU9NQQ687/nqfmdWrISULSX8iaW6u6BDgaWAZcHwqm5F+A9xdKk9HGocD95JdCH9Z0qQy/YwZ/nqfmdWroV7g3gScKen1ZEcDzwJtwDeBj0o6FHgjcAlARPxAUoek68juhro4In4FIOlcYLGknwHjgC8NMTYzM6sSjdTXO7e2tkZnZ2etw6iq/r7VMFLn01A0NjaycePGqgyroaGBDRs2VGVYZiOZpJUR0TrQ/vw9C6tbGzdurFqS9HfVzYZmzD7BbWZmlXOyMDOzQj4NVQcqOUVS6mYsXrsws9pzsqgDpQTgC9xmVq+cLOrIHnvswZYtW8qWj0Vx9QRYtHf1hmVmg+ZkUUc2b97Mnnvu2SNh7LHHHmzevLmGUdWOrtlU1buhdnjdpZlVzMmizpQSgySfejKzuuG7oczMrJCThZmZFfJpKKtr1XryuqGhoSrDMRurnCysblVyzcbXdsx2Dp+GqoHGxkYk9fsHFHbT2NhY4ykxs7HCRxY1UK0X5PnleGa2s/jIwszMCjlZmJlZIZ+GqoFqvcbCr7Aws51l0MlC0huBa4EfAfsDv4yID0taBJyc63RxRNyT+rmU7POrDcCKiLgrlU8HLgSeASYCl0TEq4ONrd5V6zUWfoWFme0sQzmyaAT+PSLuBJD0mKRlABFxcu+OJR0LzIyIMyTtCjwm6V7g18CtwGkR8YKkm4D3AEuHEJuZmVXRoJNFRDzUq2gXYAuApIXAb4FxwJKI+A1wJvBA6nerpMeBE4HVwPiIeCEN537gXEZ5sqjGnUx+0MzMdpaqXLOQ9FZgeUQ8IekOYG1EbJH0PmAJMJfs9NLjud42pbIu4KUy5eXGMw+YBzB58uRqhF4TftjMzEaaId8NJWkmMBP4AEBErI6I0ju224FTUvM6YK9crxNSWV/lO4iImyOiNSJam5qahhq6mZlVaEjJQtJsYBZwETBJ0vGSPp7r5BDg6dR8N3B86u81wOHAvcAa4GVJk1J3M4BlQ4nLzMyqayh3Qx0N3A50Ah3AHsBngVclfZrs6KCF7C4nIuIHkjokXUd2N9TFEfGrNKxzgcWSfkZ2neNLg58kMzOrtqFc4F4J7DnAfj7eR/nDZNc1zMysDvkJbhuRpk2b1uOFi9OmTatxRGbV19bWxtSpUxk3bhxTp06lra2tZrE4WdiIM23aNB599FHmzJkDwJw5c3j00UedMGxUaWtr45xzzmH16tX8/ve/Z/Xq1Zxzzjk1Sxgaqbdntra2RmdnZ63DGDa+dba8Sp9Pcd3ZSJdf1s877zy+8pWvbPs9lOVb0sqIaB1ofz6yqAMD+Z7FWFdaSbq6uoiIbX9dXV3b2jtR2GgSEXz5y1+u+XLtZFEH8hu9oj/LzJ07t9/fZiNZfsew3M5iLXYenSxsxGlpaeGuu+7irLPOYv369Zx11lncddddtLS01Do0s6rI7xi2t7f3+F9qv7N3Hn3Nwkak0kXukpaWFlatWlXDiMyqq78jh1pcs/D3LGxEcmKw0S4iyiaMWu3g+zSUmVkNNTY2lr2Zpa8ji766lURjY+OwxekjCzOzGtrw/m6y96dWQ3eVhrMjJwszsxrSNZuqNqyGhgY2LKra4HrwaSizUc7P69S3/m6Vv+2222hubgagubmZ2267rd/uN2zYMGxxOllYoXp6P40NTH/nva2+tbW1sXDhQpYsWQLAkiVLWLhwYe3Wv4E8EFZPf0cffXTY8LvtttsC2OHvtttuq3VoVoH8PFuwYEGP31bfmpubo729PSJi2/xqb2+P5ubmIQ0X6IxBbHP9nIX1K78Heuihh/LUU09t+z1Sl53hVE+3OsL2ePIxlCvb2eqtnurNcL4Dze+GsmEVETz55JNeofuRX8EvvPDCsuU7M5ZKXhlRC/nxzpgxo2z5WBcRNDc3097e3mPvvr29nebm5po8we1kYX2qx/fTjAQRwWc+85maJtZKNia1TvwRwX333VfzOOrVwoULmTt3Lh0dHWzdupWOjg7mzp3LwoULaxKPb521PkXuCdL29nZOOeWUbf9L7W3HPeJyCXQ4Xznf2NjIxo0bB9xfuTgbGhqG7Y6aSusJvGwBnH322QDMnz+fxx9/nMMOO4zFixdvK9/Z6uaahaTTgL8k+3Z3RMQ1/XXvaxZVtmjvKg/v19UdXp0Y7Ia5nKptmOt03lWrroYzgY1FI/rdUJJ2Bz4PNEfEbyV9Q9KpEfGdWsc2VoyUB4NqrR6ftq3XeVe9uhq+p5KtcnWRLIDjgZ9FxG/T7/uB2YCTxU5S7gjTX6Uro8xed63rqW7nXa+6Gsj1rTG1TI0Q9ZIsJgIv5X5vSmU9SJoHzAOYPHnyzolsDPMKW5l6rCfHZNVWL3dDrQP2yv2ekMp6iIibI6I1Ilqbmpp2WnBmZmNdvSSLB4ADJb02/Z4BLKthPGZmllMXp6Ei4jeS/hfwT5K6gFW+uG1mVj/qIlkARMQ9wD21jsPMzHZUL6ehzMysjjlZmJlZIScLMzMr5GRhZmaF6ubdUAOV7pr6WZUGty+wvkrDqhbHVJl6jAnqMy7HVJnRHtOBETHgB9VGbLKoJkmdg3mx1nByTJWpx5igPuNyTJVxTOX5NJSZmRVysjAzs0JOFpmbax1AGY6pMvUYE9RnXI6pMo6pDF+zMDOzQj6yMDOzQk4WZmZWyMmiH5LeJ2ntMAx3qqR2Seen3xdIml/Qz4clzRnk+Doljeun/T6lWKpJ0nJJO3zEqkx3VavnatRtJfEU1Wnqpke9Vmu+V1qvvfrZNk2VjDd19z8kTSlTPujpkHSqpBsHEns/w6xkvO+StFnSFEmzJT1TbpoGMe4e01FaHvoax3CtYwNV6bwvKyL8188fsHaYhrsIOD/3WwXd99t+KP0CU4DvDsM0VhxzNeu5GnVbFE8l01auXqsU26CWhfw0VRj/LcDJ1a7joSzLg5wPm0vTAXwXmFLtcfdq3mEcw7WODVedlfurm1eUDwdJfwNcD3wO2B84Gvg+2Vf49gYejoivSDoI+ERq1wLcFBEPD2J8fwxcDjwGTAV+AlwMXEf2JcAjgIsi4plcPxOAf0o/z5e0F3ATsAZ4HbAB+BLZtz4ejohFkm4nW/i+BbQCnRFxdRreZUAz8ATZR6QagT+S9PZUD/cDvwOOA+ZHRCfZp2qnSFqUhvlW4GzgX4BjgaeBjwEfBp4CDgFuiYj7UyytKdb/Jvs8roC1wPGS7gXuTtMzP9VLN/A8cHkl9VymXj+S4v+nPup2OvABSYcCfwX8oaQ24FfAHODxFGukuvojSd3Ae4B/THW7j6RlfdQtwNsk3Qf8M9kyJqAL+BPgG8A7gR8B0yU9CVyWYn4v8B1J7wQmA49KeonsbQTnkG1oXgFOA14FxkvaJcV8eqrfPSX9PtXHI2ke/R7YB7gv9bseeBL4SL6OyyxvHyWb17cDf52m4Qng/wIzgXMkfRt4AXgb2fLxXJqOSyW9n2z5+Z2k7wFXpWG1ks3z35N9CmGDpO+TfQXzXEnPAMuB84BnyNa9RuDn9Fx+fwe8PyJ+Tk4f03EOWWL4HTAeeB/ZJ5r/IHVzXBr/f0p6gGx5OA84FHg7cBJwCtk6txy4IDU/BUxK8X0zTftNwHmSPgmcAJws6SvAWcCewJWSDqfvdewx4FNk24j9gbsiYnlufuTXvU+nv0fS9CxMdfJlSR9O9dsNvBQRH8tt9z4JHAQcBpyZ6qrfbU1EfIS+1DrL7YQs+l3gjNS8ALi5lF3TAjkxzayjUvlRwB25/tcOYFz/AHwhLZxvAPYj22genNq/ozRscntmwMlkG1/STP5QbpgXpP/nA4ti+17Ks8C49Pd8Kp8KPJbr99bU33dTP4uA61O7vwaW5Ib33V7T8grQkIY/HWgD/iq1ex3ZBkO5WL5LtkA+X6pn0p5pquf/TsM5KsX57krruVy9lvrpp26fAM5I4+9M8VwPfCg33z+Y4jk/Df+O3PSsLajbF4FLUvO3gP9K8V0ILE39n5TqJR/bq2QbsXHA35KtqH+VYvtwrl4vTXH8DXBNqtcHyTbmX0rTsyV1fxLwW+DENA+WpWkquyyTW95y8/qyVAePAUeSLbu3kCWJUh3fATyRmj8HXAGsKg0vlX0z1fEtwN+RbXAvJVt+FqX6fTBN23tTP6U6vpIyy28/y0W56TgrNT8B/Cg19z6y+BTwsfR7GtkO4qrcsvNfZBv3aWRJ8xe5efIQWUKYAjya+nlvmq9TUv+/TM19rmNk8/vy1DyebGfhNX2se3cA70jtDs7Nx1nAil7buum55lmp+bPA2yrd1vT1N1auWTye/k8G9kt7iAuAH5PtMWwF3ilpIdmKPdgPfP8L2VHL98hW8K2pfE36/zTb90z7Mi11B0BEfLGP7tZERHdEdOfGc3i+39x4855K/7vo+d3z3l6MiI1pHA+nuNakmF4kOzLbt9d4VqdYJpNtbFrI9vaeItuj+wXZHvdZZPVQaT33Va/5cZer29J83yPF89YUU2m+35riObNXPKXp7K9uX841ryTb+H2P7Mhl91Re2hvOx7YZ+E0a9tNkG4U1ZPX7MNvr9Y5UHukvvyxFilWp23lkG6hfkW3M9knTVOmy/CKwhKyO9ydLAuXqeAPbl5kX2b4zVPI02V5sqZ6eSstvF9kGMW8N2bz4SW5cT1K8/PZnHXCMpCvJEnd/13UeB4iIVWRHg/npeBI4IrXbm2zjvQD4U7LrvBPINuR7SLqG7IhUuf67cv/7WsemAX+StkUXAY+SHbnAjuteM1k9Qc86mQbsLumyNJxn6Tm/i9b1Src1wNi5wF16mOQR4CcRcUNE3AB8lWwhuQzYHBGLyfYKB+tY4IaIOJZsZXp3Kj8o/T+UbM+tP48AbwRQ5u/66K7cAzKPk50iKjmoTDfl+usmLeySpvfRXT6uSWQbpvW9uu1Rz2QrwG3Av6XuryHbWP4H2SmaSvVVr9B/3Zbi2Zji+d9ke8Kl+f7VFM/dZHuTvfvL6123+Y3f/sDdKb6NpHoC/oisXg8lOzIop9T9I2R7kaV6PaePOEplPyXbyHaRHWV0kyXFT5OdBrmbypflINUx2XxZT1bH3an9yZL2BP6Q7Aix5DmyI72SQ8jmQWn6o2D5fT2omCUAAANzSURBVCTXLWRHb0XLb3/2AF4f2amUZ3uNS5IOJtvIl8pKHqXndLyOLHFDNj9+l7YXy8iOUh8mO13VENkpyn/rFcc+ZWLrvY49QnaKs7Qtup0s4feODbI6PTQ15+vkEWBdbhj/Rpbo8tPdn0q3NUAdfVZ1OEh6M3Ag8PfpXOBS4GPpvOEfAC9HxP+R9A3gekmvTeUHSjoV+GNgb0lnR0RbBaNsBD4haQ1Zhv9n4P3Aaek84pHAfElTyU4ZtEjqIDtvOk3Sm8gODW9KRzn7AN+UdADwFqAhnQc9JxfjhBTjBRHxRUm3pvPzq4DXku09HEh2TWafNM7luXG2ki38r0j6BPBYKttb0sUR8Yk0bZcAiyUdQnYo/M6ICEl/m+oJssP7fchWjP3IDnmnA3eR7eX+I9mh81vIzlFXWs/l6rWkXN2eBRxAdp1jD2C3FM8zqe7WA29Kcb2LbO9xPNmG5FqyjdZrJd3UR93+lmyDfLKku8hO9/yZpMmp+Xdky9E0skRyHXCzpKtSPB/IzfctZHumnWRHOf9Jdn1oEtmG4YJUp01ke8q/T//XpPm7lixBvEJ26ud+4PXAX5Ltkfau478hO8deWt4OT9P/F8AHUsz7kCWOX6T59lGyZeRkoFPSO8iW341kCfczZInke2Tn228hSz5HkZ1au4Dt17UOINuTfyXNj7cAr5P0H2lYvZffshu8dL49v94cTrY9O07Sv5MdtYyX9C6yjfQNaZxvIDvyPUDSjyOiMyKekLRE0i1kp/S6gUck/SnZutMl6QvAn5Gtb58lu0a2W5r/vyTb8f4+2TK0N9kpzt3pYx0jW88/JumK1P2aiOhO61PvdW8BsERSC1kSDICIWCHpTyVdT3YabDfgstx274I0TaVtzXco2NaUq+ttdZ7OVdkwkbQ2IqbsxPG9KSK+n5q/CHwxIu7bWePfmeq5bus5tgEOd1RMx0iWdtA2RsT6tEPyxYg4bWfHMaqPLGot7dXsLel9EfHPhT1Ux/slnUK2h/Pfo3VFq+e6refYBmK0TMcocCBwjaT/Ijuqv7gWQfjIwuqKyjy4FBG37PxIMvUWz0Ckc+PTexU/XKasrqepr+mIQdzenhvmn5Od6sv7VkS8MNhhjnZOFmZmVmis3A1lZmZD4GRhZmaFnCzMzKyQk4WZmRVysjAzs0L/H0pLOaJJmh2oAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# based on the number of loci per sample\n",
    "s_loc = pd.read_csv(base_dir_out+\"res.num_locs_sample\")\n",
    "\n",
    "for label in s_loc.columns:\n",
    "    # now we shall remove any outliers from the data\n",
    "    q25,q50,q75 = s_loc[label].quantile([0.25,0.5,0.75])\n",
    "    iqr = q75-q25\n",
    "    thw = q75+1.5*iqr\n",
    "    tlw = q25-1.5*iqr\n",
    "    ahw = s_loc[s_loc[label]<=thw][label].max()\n",
    "    alw = s_loc[s_loc[label]>=tlw][label].min()\n",
    "    s_loc = s_loc[(s_loc[label]<=ahw)&(s_loc[label]>=alw)].reset_index(drop=True)\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "ax.boxplot(s_loc.T,labels=s_loc.T.index.tolist())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "starting number of real loci: 31218\n",
      "starting number of splicing loci: 25878\n",
      "starting number of intronic loci: 14519\n",
      "starting number of intergenic loci: 29016\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39411\n",
      "number of splicing transcripts: 20351\n",
      "number of intronic transcripts: 4217\n",
      "number of intergenic transcripts: 29423\n",
      "number of real genes: 20912\n",
      "number of solicing genes: 13163\n",
      "number of intronic genes: 2951\n",
      "number of intergenic genes: 29013\n",
      "targeting 7389 real only loci\n",
      "targeting 3436 splicing only loci\n",
      "targeting 286 intronic only loci\n",
      "targeting 4023 real and splicing loci\n",
      "targeting 285 real and intronic loci\n",
      "targeting 481 splicing and intronic loci\n",
      "targeting 233 real and splicing and intronic loci\n",
      "targeting 1026 intergenic loci\n",
      "available real only loci: 16127\n",
      "available splicing only loci: 8015\n",
      "available intronic only loci: 1550\n",
      "available real and splicing loci: 4120\n",
      "available real and intronic loci: 373\n",
      "available splicing and intronic loci: 736\n",
      "available real and splicing and intronic loci: 292\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 7389\n",
      "selecting splicing only loci: 3436\n",
      "selecting intronic only loci: 286\n",
      "selecting real and splicing loci: 4023\n",
      "selecting real and intronic loci: 285\n",
      "selecting splicing and intronic loci: 481\n",
      "selecting real and splicing and intronic loci: 233\n",
      "selecting intergenic loci: 1026\n",
      "final number of real transcripts: 25281\n",
      "final number of splicing transcripts: 13266\n",
      "final number of intronic transcripts: 1983\n",
      "final number of intergenic transcripts: 1038\n",
      "final number of real loci: 11930\n",
      "final number of splicing loci: 8173\n",
      "final number of intronic loci: 1285\n",
      "final number of intergenic loci: 1026\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39156\n",
      "number of splicing transcripts: 20328\n",
      "number of intronic transcripts: 4304\n",
      "number of intergenic transcripts: 29400\n",
      "number of real genes: 20745\n",
      "number of solicing genes: 13287\n",
      "number of intronic genes: 3029\n",
      "number of intergenic genes: 29013\n",
      "targeting 7366 real only loci\n",
      "targeting 3177 splicing only loci\n",
      "targeting 289 intronic only loci\n",
      "targeting 4665 real and splicing loci\n",
      "targeting 231 real and intronic loci\n",
      "targeting 387 splicing and intronic loci\n",
      "targeting 251 real and splicing and intronic loci\n",
      "targeting 1113 intergenic loci\n",
      "available real only loci: 15958\n",
      "available splicing only loci: 8099\n",
      "available intronic only loci: 1589\n",
      "available real and splicing loci: 4123\n",
      "available real and intronic loci: 375\n",
      "available splicing and intronic loci: 776\n",
      "available real and splicing and intronic loci: 289\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 7366\n",
      "selecting splicing only loci: 3177\n",
      "selecting intronic only loci: 289\n",
      "selecting real and splicing loci: 4123\n",
      "selecting real and intronic loci: 231\n",
      "selecting splicing and intronic loci: 387\n",
      "selecting real and splicing and intronic loci: 251\n",
      "selecting intergenic loci: 1113\n",
      "final number of real transcripts: 25379\n",
      "final number of splicing transcripts: 12927\n",
      "final number of intronic transcripts: 1751\n",
      "final number of intergenic transcripts: 1131\n",
      "final number of real loci: 11971\n",
      "final number of splicing loci: 7938\n",
      "final number of intronic loci: 1158\n",
      "final number of intergenic loci: 1113\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39436\n",
      "number of splicing transcripts: 20555\n",
      "number of intronic transcripts: 4283\n",
      "number of intergenic transcripts: 29390\n",
      "number of real genes: 20782\n",
      "number of solicing genes: 13367\n",
      "number of intronic genes: 2942\n",
      "number of intergenic genes: 29013\n",
      "targeting 7718 real only loci\n",
      "targeting 3167 splicing only loci\n",
      "targeting 310 intronic only loci\n",
      "targeting 5074 real and splicing loci\n",
      "targeting 226 real and intronic loci\n",
      "targeting 353 splicing and intronic loci\n",
      "targeting 206 real and splicing and intronic loci\n",
      "targeting 1307 intergenic loci\n",
      "available real only loci: 15933\n",
      "available splicing only loci: 8163\n",
      "available intronic only loci: 1520\n",
      "available real and splicing loci: 4172\n",
      "available real and intronic loci: 390\n",
      "available splicing and intronic loci: 745\n",
      "available real and splicing and intronic loci: 287\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 7718\n",
      "selecting splicing only loci: 3167\n",
      "selecting intronic only loci: 310\n",
      "selecting real and splicing loci: 4172\n",
      "selecting real and intronic loci: 226\n",
      "selecting splicing and intronic loci: 353\n",
      "selecting real and splicing and intronic loci: 206\n",
      "selecting intergenic loci: 1307\n",
      "final number of real transcripts: 26086\n",
      "final number of splicing transcripts: 12881\n",
      "final number of intronic transcripts: 1695\n",
      "final number of intergenic transcripts: 1323\n",
      "final number of real loci: 12322\n",
      "final number of splicing loci: 7898\n",
      "final number of intronic loci: 1095\n",
      "final number of intergenic loci: 1307\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39501\n",
      "number of splicing transcripts: 20434\n",
      "number of intronic transcripts: 4231\n",
      "number of intergenic transcripts: 29430\n",
      "number of real genes: 20813\n",
      "number of solicing genes: 13249\n",
      "number of intronic genes: 2970\n",
      "number of intergenic genes: 29013\n",
      "targeting 6194 real only loci\n",
      "targeting 3646 splicing only loci\n",
      "targeting 290 intronic only loci\n",
      "targeting 4576 real and splicing loci\n",
      "targeting 216 real and intronic loci\n",
      "targeting 541 splicing and intronic loci\n",
      "targeting 304 real and splicing and intronic loci\n",
      "targeting 1267 intergenic loci\n",
      "available real only loci: 16041\n",
      "available splicing only loci: 8072\n",
      "available intronic only loci: 1540\n",
      "available real and splicing loci: 4120\n",
      "available real and intronic loci: 373\n",
      "available splicing and intronic loci: 778\n",
      "available real and splicing and intronic loci: 279\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 6194\n",
      "selecting splicing only loci: 3646\n",
      "selecting intronic only loci: 290\n",
      "selecting real and splicing loci: 4120\n",
      "selecting real and intronic loci: 216\n",
      "selecting splicing and intronic loci: 541\n",
      "selecting real and splicing and intronic loci: 279\n",
      "selecting intergenic loci: 1267\n",
      "final number of real transcripts: 23679\n",
      "final number of splicing transcripts: 13887\n",
      "final number of intronic transcripts: 2034\n",
      "final number of intergenic transcripts: 1297\n",
      "final number of real loci: 10809\n",
      "final number of splicing loci: 8586\n",
      "final number of intronic loci: 1326\n",
      "final number of intergenic loci: 1267\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39082\n",
      "number of splicing transcripts: 20647\n",
      "number of intronic transcripts: 4192\n",
      "number of intergenic transcripts: 29387\n",
      "number of real genes: 20818\n",
      "number of solicing genes: 13449\n",
      "number of intronic genes: 2921\n",
      "number of intergenic genes: 29013\n",
      "targeting 7606 real only loci\n",
      "targeting 3167 splicing only loci\n",
      "targeting 265 intronic only loci\n",
      "targeting 5219 real and splicing loci\n",
      "targeting 228 real and intronic loci\n",
      "targeting 347 splicing and intronic loci\n",
      "targeting 256 real and splicing and intronic loci\n",
      "targeting 1259 intergenic loci\n",
      "available real only loci: 15885\n",
      "available splicing only loci: 8135\n",
      "available intronic only loci: 1508\n",
      "available real and splicing loci: 4264\n",
      "available real and intronic loci: 363\n",
      "available splicing and intronic loci: 744\n",
      "available real and splicing and intronic loci: 306\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 7606\n",
      "selecting splicing only loci: 3167\n",
      "selecting intronic only loci: 265\n",
      "selecting real and splicing loci: 4264\n",
      "selecting real and intronic loci: 228\n",
      "selecting splicing and intronic loci: 347\n",
      "selecting real and splicing and intronic loci: 256\n",
      "selecting intergenic loci: 1259\n",
      "final number of real transcripts: 25985\n",
      "final number of splicing transcripts: 13000\n",
      "final number of intronic transcripts: 1637\n",
      "final number of intergenic transcripts: 1271\n",
      "final number of real loci: 12354\n",
      "final number of splicing loci: 8034\n",
      "final number of intronic loci: 1096\n",
      "final number of intergenic loci: 1259\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39245\n",
      "number of splicing transcripts: 20621\n",
      "number of intronic transcripts: 4236\n",
      "number of intergenic transcripts: 29405\n",
      "number of real genes: 20764\n",
      "number of solicing genes: 13359\n",
      "number of intronic genes: 2985\n",
      "number of intergenic genes: 29013\n",
      "targeting 7913 real only loci\n",
      "targeting 2540 splicing only loci\n",
      "targeting 192 intronic only loci\n",
      "targeting 2835 real and splicing loci\n",
      "targeting 139 real and intronic loci\n",
      "targeting 132 splicing and intronic loci\n",
      "targeting 96 real and splicing and intronic loci\n",
      "targeting 941 intergenic loci\n",
      "available real only loci: 15945\n",
      "available splicing only loci: 8153\n",
      "available intronic only loci: 1522\n",
      "available real and splicing loci: 4130\n",
      "available real and intronic loci: 387\n",
      "available splicing and intronic loci: 774\n",
      "available real and splicing and intronic loci: 302\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 7913\n",
      "selecting splicing only loci: 2540\n",
      "selecting intronic only loci: 192\n",
      "selecting real and splicing loci: 2835\n",
      "selecting real and intronic loci: 139\n",
      "selecting splicing and intronic loci: 132\n",
      "selecting real and splicing and intronic loci: 96\n",
      "selecting intergenic loci: 941\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 21785\n",
      "final number of splicing transcripts: 9021\n",
      "final number of intronic transcripts: 797\n",
      "final number of intergenic transcripts: 955\n",
      "final number of real loci: 10983\n",
      "final number of splicing loci: 5603\n",
      "final number of intronic loci: 559\n",
      "final number of intergenic loci: 941\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39547\n",
      "number of splicing transcripts: 20481\n",
      "number of intronic transcripts: 4146\n",
      "number of intergenic transcripts: 29396\n",
      "number of real genes: 20798\n",
      "number of solicing genes: 13292\n",
      "number of intronic genes: 2981\n",
      "number of intergenic genes: 29013\n",
      "targeting 8722 real only loci\n",
      "targeting 2493 splicing only loci\n",
      "targeting 157 intronic only loci\n",
      "targeting 2867 real and splicing loci\n",
      "targeting 110 real and intronic loci\n",
      "targeting 81 splicing and intronic loci\n",
      "targeting 56 real and splicing and intronic loci\n",
      "targeting 918 intergenic loci\n",
      "available real only loci: 15973\n",
      "available splicing only loci: 8097\n",
      "available intronic only loci: 1552\n",
      "available real and splicing loci: 4161\n",
      "available real and intronic loci: 395\n",
      "available splicing and intronic loci: 765\n",
      "available real and splicing and intronic loci: 269\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 8722\n",
      "selecting splicing only loci: 2493\n",
      "selecting intronic only loci: 157\n",
      "selecting real and splicing loci: 2867\n",
      "selecting real and intronic loci: 110\n",
      "selecting splicing and intronic loci: 81\n",
      "selecting real and splicing and intronic loci: 56\n",
      "selecting intergenic loci: 918\n",
      "final number of real transcripts: 22898\n",
      "final number of splicing transcripts: 8793\n",
      "final number of intronic transcripts: 542\n",
      "final number of intergenic transcripts: 939\n",
      "final number of real loci: 11755\n",
      "final number of splicing loci: 5497\n",
      "final number of intronic loci: 404\n",
      "final number of intergenic loci: 918\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39477\n",
      "number of splicing transcripts: 20644\n",
      "number of intronic transcripts: 4166\n",
      "number of intergenic transcripts: 29394\n",
      "number of real genes: 20770\n",
      "number of solicing genes: 13406\n",
      "number of intronic genes: 2923\n",
      "number of intergenic genes: 29013\n",
      "targeting 8808 real only loci\n",
      "targeting 1951 splicing only loci\n",
      "targeting 164 intronic only loci\n",
      "targeting 3441 real and splicing loci\n",
      "targeting 167 real and intronic loci\n",
      "targeting 100 splicing and intronic loci\n",
      "targeting 98 real and splicing and intronic loci\n",
      "targeting 977 intergenic loci\n",
      "available real only loci: 15911\n",
      "available splicing only loci: 8164\n",
      "available intronic only loci: 1530\n",
      "available real and splicing loci: 4214\n",
      "available real and intronic loci: 365\n",
      "available splicing and intronic loci: 748\n",
      "available real and splicing and intronic loci: 280\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 8808\n",
      "selecting splicing only loci: 1951\n",
      "selecting intronic only loci: 164\n",
      "selecting real and splicing loci: 3441\n",
      "selecting real and intronic loci: 167\n",
      "selecting splicing and intronic loci: 100\n",
      "selecting real and splicing and intronic loci: 98\n",
      "selecting intergenic loci: 977\n",
      "final number of real transcripts: 25063\n",
      "final number of splicing transcripts: 9110\n",
      "final number of intronic transcripts: 821\n",
      "final number of intergenic transcripts: 983\n",
      "final number of real loci: 12514\n",
      "final number of splicing loci: 5590\n",
      "final number of intronic loci: 529\n",
      "final number of intergenic loci: 977\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39318\n",
      "number of splicing transcripts: 20497\n",
      "number of intronic transcripts: 4029\n",
      "number of intergenic transcripts: 29420\n",
      "number of real genes: 20792\n",
      "number of solicing genes: 13371\n",
      "number of intronic genes: 2911\n",
      "number of intergenic genes: 29013\n",
      "targeting 8847 real only loci\n",
      "targeting 4356 splicing only loci\n",
      "targeting 356 intronic only loci\n",
      "targeting 3269 real and splicing loci\n",
      "targeting 281 real and intronic loci\n",
      "targeting 476 splicing and intronic loci\n",
      "targeting 211 real and splicing and intronic loci\n",
      "targeting 1523 intergenic loci\n",
      "available real only loci: 15947\n",
      "available splicing only loci: 8163\n",
      "available intronic only loci: 1515\n",
      "available real and splicing loci: 4190\n",
      "available real and intronic loci: 378\n",
      "available splicing and intronic loci: 741\n",
      "available real and splicing and intronic loci: 277\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 8847\n",
      "selecting splicing only loci: 4356\n",
      "selecting intronic only loci: 356\n",
      "selecting real and splicing loci: 3269\n",
      "selecting real and intronic loci: 281\n",
      "selecting splicing and intronic loci: 476\n",
      "selecting real and splicing and intronic loci: 211\n",
      "selecting intergenic loci: 1523\n",
      "final number of real transcripts: 25031\n",
      "final number of splicing transcripts: 13063\n",
      "final number of intronic transcripts: 1972\n",
      "final number of intergenic transcripts: 1544\n",
      "final number of real loci: 12608\n",
      "final number of splicing loci: 8312\n",
      "final number of intronic loci: 1324\n",
      "final number of intergenic loci: 1523\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 39471\n",
      "number of splicing transcripts: 20381\n",
      "number of intronic transcripts: 4247\n",
      "number of intergenic transcripts: 29410\n",
      "number of real genes: 20835\n",
      "number of solicing genes: 13223\n",
      "number of intronic genes: 2945\n",
      "number of intergenic genes: 29013\n",
      "targeting 9209 real only loci\n",
      "targeting 2528 splicing only loci\n",
      "targeting 177 intronic only loci\n",
      "targeting 4193 real and splicing loci\n",
      "targeting 153 real and intronic loci\n",
      "targeting 101 splicing and intronic loci\n",
      "targeting 106 real and splicing and intronic loci\n",
      "targeting 1108 intergenic loci\n",
      "available real only loci: 16044\n",
      "available splicing only loci: 8092\n",
      "available intronic only loci: 1559\n",
      "available real and splicing loci: 4129\n",
      "available real and intronic loci: 384\n",
      "available splicing and intronic loci: 724\n",
      "available real and splicing and intronic loci: 278\n",
      "available intergenic only loci: 29013\n",
      "selecting real only loci: 9209\n",
      "selecting splicing only loci: 2528\n",
      "selecting intronic only loci: 177\n",
      "selecting real and splicing loci: 4129\n",
      "selecting real and intronic loci: 153\n",
      "selecting splicing and intronic loci: 101\n",
      "selecting real and splicing and intronic loci: 106\n",
      "selecting intergenic loci: 1108\n",
      "final number of real transcripts: 27961\n",
      "final number of splicing transcripts: 11273\n",
      "final number of intronic transcripts: 851\n",
      "final number of intergenic transcripts: 1115\n",
      "final number of real loci: 13597\n",
      "final number of splicing loci: 6864\n",
      "final number of intronic loci: 537\n",
      "final number of intergenic loci: 1108\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "starting number of real loci: 29592\n",
      "starting number of splicing loci: 24533\n",
      "starting number of intronic loci: 13694\n",
      "starting number of intergenic loci: 29016\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37515\n",
      "number of splicing transcripts: 19584\n",
      "number of intronic transcripts: 3824\n",
      "number of intergenic transcripts: 29372\n",
      "number of real genes: 19777\n",
      "number of solicing genes: 12694\n",
      "number of intronic genes: 2739\n",
      "number of intergenic genes: 29016\n",
      "targeting 8091 real only loci\n",
      "targeting 3500 splicing only loci\n",
      "targeting 291 intronic only loci\n",
      "targeting 4380 real and splicing loci\n",
      "targeting 194 real and intronic loci\n",
      "targeting 400 splicing and intronic loci\n",
      "targeting 232 real and splicing and intronic loci\n",
      "targeting 1231 intergenic loci\n",
      "available real only loci: 15091\n",
      "available splicing only loci: 7658\n",
      "available intronic only loci: 1449\n",
      "available real and splicing loci: 4090\n",
      "available real and intronic loci: 344\n",
      "available splicing and intronic loci: 694\n",
      "available real and splicing and intronic loci: 252\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8091\n",
      "selecting splicing only loci: 3500\n",
      "selecting intronic only loci: 291\n",
      "selecting real and splicing loci: 4090\n",
      "selecting real and intronic loci: 194\n",
      "selecting splicing and intronic loci: 400\n",
      "selecting real and splicing and intronic loci: 232\n",
      "selecting intergenic loci: 1231\n",
      "final number of real transcripts: 26530\n",
      "final number of splicing transcripts: 13198\n",
      "final number of intronic transcripts: 1662\n",
      "final number of intergenic transcripts: 1254\n",
      "final number of real loci: 12607\n",
      "final number of splicing loci: 8222\n",
      "final number of intronic loci: 1117\n",
      "final number of intergenic loci: 1231\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37553\n",
      "number of splicing transcripts: 19492\n",
      "number of intronic transcripts: 4094\n",
      "number of intergenic transcripts: 29371\n",
      "number of real genes: 19757\n",
      "number of solicing genes: 12722\n",
      "number of intronic genes: 2830\n",
      "number of intergenic genes: 29016\n",
      "targeting 6144 real only loci\n",
      "targeting 3143 splicing only loci\n",
      "targeting 228 intronic only loci\n",
      "targeting 2820 real and splicing loci\n",
      "targeting 160 real and intronic loci\n",
      "targeting 348 splicing and intronic loci\n",
      "targeting 144 real and splicing and intronic loci\n",
      "targeting 886 intergenic loci\n",
      "available real only loci: 15023\n",
      "available splicing only loci: 7661\n",
      "available intronic only loci: 1471\n",
      "available real and splicing loci: 4069\n",
      "available real and intronic loci: 367\n",
      "available splicing and intronic loci: 694\n",
      "available real and splicing and intronic loci: 298\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 6144\n",
      "selecting splicing only loci: 3143\n",
      "selecting intronic only loci: 228\n",
      "selecting real and splicing loci: 2820\n",
      "selecting real and intronic loci: 160\n",
      "selecting splicing and intronic loci: 348\n",
      "selecting real and splicing and intronic loci: 144\n",
      "selecting intergenic loci: 886\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 19195\n",
      "final number of splicing transcripts: 10167\n",
      "final number of intronic transcripts: 1447\n",
      "final number of intergenic transcripts: 899\n",
      "final number of real loci: 9268\n",
      "final number of splicing loci: 6455\n",
      "final number of intronic loci: 880\n",
      "final number of intergenic loci: 886\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37735\n",
      "number of splicing transcripts: 19295\n",
      "number of intronic transcripts: 4042\n",
      "number of intergenic transcripts: 29367\n",
      "number of real genes: 19742\n",
      "number of solicing genes: 12597\n",
      "number of intronic genes: 2830\n",
      "number of intergenic genes: 29016\n",
      "targeting 7901 real only loci\n",
      "targeting 2622 splicing only loci\n",
      "targeting 179 intronic only loci\n",
      "targeting 3058 real and splicing loci\n",
      "targeting 154 real and intronic loci\n",
      "targeting 146 splicing and intronic loci\n",
      "targeting 94 real and splicing and intronic loci\n",
      "targeting 920 intergenic loci\n",
      "available real only loci: 15124\n",
      "available splicing only loci: 7649\n",
      "available intronic only loci: 1489\n",
      "available real and splicing loci: 3979\n",
      "available real and intronic loci: 372\n",
      "available splicing and intronic loci: 702\n",
      "available real and splicing and intronic loci: 267\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 7901\n",
      "selecting splicing only loci: 2622\n",
      "selecting intronic only loci: 179\n",
      "selecting real and splicing loci: 3058\n",
      "selecting real and intronic loci: 154\n",
      "selecting splicing and intronic loci: 146\n",
      "selecting real and splicing and intronic loci: 94\n",
      "selecting intergenic loci: 920\n",
      "final number of real transcripts: 22407\n",
      "final number of splicing transcripts: 9535\n",
      "final number of intronic transcripts: 864\n",
      "final number of intergenic transcripts: 932\n",
      "final number of real loci: 11207\n",
      "final number of splicing loci: 5920\n",
      "final number of intronic loci: 573\n",
      "final number of intergenic loci: 920\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37753\n",
      "number of splicing transcripts: 19660\n",
      "number of intronic transcripts: 3936\n",
      "number of intergenic transcripts: 29376\n",
      "number of real genes: 19873\n",
      "number of solicing genes: 12571\n",
      "number of intronic genes: 2785\n",
      "number of intergenic genes: 29016\n",
      "targeting 8488 real only loci\n",
      "targeting 4617 splicing only loci\n",
      "targeting 347 intronic only loci\n",
      "targeting 3613 real and splicing loci\n",
      "targeting 238 real and intronic loci\n",
      "targeting 607 splicing and intronic loci\n",
      "targeting 244 real and splicing and intronic loci\n",
      "targeting 1604 intergenic loci\n",
      "available real only loci: 15204\n",
      "available splicing only loci: 7557\n",
      "available intronic only loci: 1457\n",
      "available real and splicing loci: 4035\n",
      "available real and intronic loci: 349\n",
      "available splicing and intronic loci: 694\n",
      "available real and splicing and intronic loci: 285\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8488\n",
      "selecting splicing only loci: 4617\n",
      "selecting intronic only loci: 347\n",
      "selecting real and splicing loci: 3613\n",
      "selecting real and intronic loci: 238\n",
      "selecting splicing and intronic loci: 607\n",
      "selecting real and splicing and intronic loci: 244\n",
      "selecting intergenic loci: 1604\n",
      "final number of real transcripts: 25463\n",
      "final number of splicing transcripts: 14572\n",
      "final number of intronic transcripts: 2197\n",
      "final number of intergenic transcripts: 1625\n",
      "final number of real loci: 12583\n",
      "final number of splicing loci: 9081\n",
      "final number of intronic loci: 1436\n",
      "final number of intergenic loci: 1604\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37431\n",
      "number of splicing transcripts: 19650\n",
      "number of intronic transcripts: 4072\n",
      "number of intergenic transcripts: 29398\n",
      "number of real genes: 19796\n",
      "number of solicing genes: 12704\n",
      "number of intronic genes: 2851\n",
      "number of intergenic genes: 29016\n",
      "targeting 6925 real only loci\n",
      "targeting 3394 splicing only loci\n",
      "targeting 258 intronic only loci\n",
      "targeting 4431 real and splicing loci\n",
      "targeting 187 real and intronic loci\n",
      "targeting 349 splicing and intronic loci\n",
      "targeting 185 real and splicing and intronic loci\n",
      "targeting 1145 intergenic loci\n",
      "available real only loci: 15051\n",
      "available splicing only loci: 7588\n",
      "available intronic only loci: 1458\n",
      "available real and splicing loci: 4086\n",
      "available real and intronic loci: 363\n",
      "available splicing and intronic loci: 734\n",
      "available real and splicing and intronic loci: 296\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 6925\n",
      "selecting splicing only loci: 3394\n",
      "selecting intronic only loci: 258\n",
      "selecting real and splicing loci: 4086\n",
      "selecting real and intronic loci: 187\n",
      "selecting splicing and intronic loci: 349\n",
      "selecting real and splicing and intronic loci: 185\n",
      "selecting intergenic loci: 1145\n",
      "final number of real transcripts: 24284\n",
      "final number of splicing transcripts: 12878\n",
      "final number of intronic transcripts: 1516\n",
      "final number of intergenic transcripts: 1168\n",
      "final number of real loci: 11383\n",
      "final number of splicing loci: 8014\n",
      "final number of intronic loci: 979\n",
      "final number of intergenic loci: 1145\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37439\n",
      "number of splicing transcripts: 19590\n",
      "number of intronic transcripts: 3932\n",
      "number of intergenic transcripts: 29395\n",
      "number of real genes: 19801\n",
      "number of solicing genes: 12641\n",
      "number of intronic genes: 2797\n",
      "number of intergenic genes: 29016\n",
      "targeting 8642 real only loci\n",
      "targeting 3386 splicing only loci\n",
      "targeting 321 intronic only loci\n",
      "targeting 4410 real and splicing loci\n",
      "targeting 230 real and intronic loci\n",
      "targeting 324 splicing and intronic loci\n",
      "targeting 202 real and splicing and intronic loci\n",
      "targeting 1455 intergenic loci\n",
      "available real only loci: 15139\n",
      "available splicing only loci: 7591\n",
      "available intronic only loci: 1442\n",
      "available real and splicing loci: 4052\n",
      "available real and intronic loci: 357\n",
      "available splicing and intronic loci: 745\n",
      "available real and splicing and intronic loci: 253\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8642\n",
      "selecting splicing only loci: 3386\n",
      "selecting intronic only loci: 321\n",
      "selecting real and splicing loci: 4052\n",
      "selecting real and intronic loci: 230\n",
      "selecting splicing and intronic loci: 324\n",
      "selecting real and splicing and intronic loci: 202\n",
      "selecting intergenic loci: 1455\n",
      "final number of real transcripts: 26781\n",
      "final number of splicing transcripts: 12935\n",
      "final number of intronic transcripts: 1634\n",
      "final number of intergenic transcripts: 1467\n",
      "final number of real loci: 13126\n",
      "final number of splicing loci: 7964\n",
      "final number of intronic loci: 1077\n",
      "final number of intergenic loci: 1455\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37686\n",
      "number of splicing transcripts: 19260\n",
      "number of intronic transcripts: 4025\n",
      "number of intergenic transcripts: 29385\n",
      "number of real genes: 19834\n",
      "number of solicing genes: 12561\n",
      "number of intronic genes: 2825\n",
      "number of intergenic genes: 29016\n",
      "targeting 9628 real only loci\n",
      "targeting 2621 splicing only loci\n",
      "targeting 148 intronic only loci\n",
      "targeting 3195 real and splicing loci\n",
      "targeting 131 real and intronic loci\n",
      "targeting 84 splicing and intronic loci\n",
      "targeting 61 real and splicing and intronic loci\n",
      "targeting 1127 intergenic loci\n",
      "available real only loci: 15159\n",
      "available splicing only loci: 7568\n",
      "available intronic only loci: 1488\n",
      "available real and splicing loci: 4022\n",
      "available real and intronic loci: 366\n",
      "available splicing and intronic loci: 684\n",
      "available real and splicing and intronic loci: 287\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 9628\n",
      "selecting splicing only loci: 2621\n",
      "selecting intronic only loci: 148\n",
      "selecting real and splicing loci: 3195\n",
      "selecting real and intronic loci: 131\n",
      "selecting splicing and intronic loci: 84\n",
      "selecting real and splicing and intronic loci: 61\n",
      "selecting intergenic loci: 1127\n",
      "final number of real transcripts: 25239\n",
      "final number of splicing transcripts: 9517\n",
      "final number of intronic transcripts: 613\n",
      "final number of intergenic transcripts: 1141\n",
      "final number of real loci: 13015\n",
      "final number of splicing loci: 5961\n",
      "final number of intronic loci: 424\n",
      "final number of intergenic loci: 1127\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37782\n",
      "number of splicing transcripts: 19595\n",
      "number of intronic transcripts: 3966\n",
      "number of intergenic transcripts: 29350\n",
      "number of real genes: 19806\n",
      "number of solicing genes: 12633\n",
      "number of intronic genes: 2817\n",
      "number of intergenic genes: 29016\n",
      "targeting 8176 real only loci\n",
      "targeting 2764 splicing only loci\n",
      "targeting 283 intronic only loci\n",
      "targeting 4435 real and splicing loci\n",
      "targeting 260 real and intronic loci\n",
      "targeting 255 splicing and intronic loci\n",
      "targeting 254 real and splicing and intronic loci\n",
      "targeting 1236 intergenic loci\n",
      "available real only loci: 15119\n",
      "available splicing only loci: 7580\n",
      "available intronic only loci: 1462\n",
      "available real and splicing loci: 4062\n",
      "available real and intronic loci: 364\n",
      "available splicing and intronic loci: 730\n",
      "available real and splicing and intronic loci: 261\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8176\n",
      "selecting splicing only loci: 2764\n",
      "selecting intronic only loci: 283\n",
      "selecting real and splicing loci: 4062\n",
      "selecting real and intronic loci: 260\n",
      "selecting splicing and intronic loci: 255\n",
      "selecting real and splicing and intronic loci: 254\n",
      "selecting intergenic loci: 1236\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 26882\n",
      "final number of splicing transcripts: 12011\n",
      "final number of intronic transcripts: 1537\n",
      "final number of intergenic transcripts: 1252\n",
      "final number of real loci: 12752\n",
      "final number of splicing loci: 7335\n",
      "final number of intronic loci: 1052\n",
      "final number of intergenic loci: 1236\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37640\n",
      "number of splicing transcripts: 19604\n",
      "number of intronic transcripts: 4018\n",
      "number of intergenic transcripts: 29390\n",
      "number of real genes: 19761\n",
      "number of solicing genes: 12654\n",
      "number of intronic genes: 2830\n",
      "number of intergenic genes: 29016\n",
      "targeting 7053 real only loci\n",
      "targeting 2980 splicing only loci\n",
      "targeting 318 intronic only loci\n",
      "targeting 4828 real and splicing loci\n",
      "targeting 297 real and intronic loci\n",
      "targeting 457 splicing and intronic loci\n",
      "targeting 385 real and splicing and intronic loci\n",
      "targeting 1157 intergenic loci\n",
      "available real only loci: 15081\n",
      "available splicing only loci: 7641\n",
      "available intronic only loci: 1472\n",
      "available real and splicing loci: 4027\n",
      "available real and intronic loci: 372\n",
      "available splicing and intronic loci: 705\n",
      "available real and splicing and intronic loci: 281\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 7053\n",
      "selecting splicing only loci: 2980\n",
      "selecting intronic only loci: 318\n",
      "selecting real and splicing loci: 4027\n",
      "selecting real and intronic loci: 297\n",
      "selecting splicing and intronic loci: 457\n",
      "selecting real and splicing and intronic loci: 281\n",
      "selecting intergenic loci: 1157\n",
      "final number of real transcripts: 25335\n",
      "final number of splicing transcripts: 12731\n",
      "final number of intronic transcripts: 2050\n",
      "final number of intergenic transcripts: 1170\n",
      "final number of real loci: 11658\n",
      "final number of splicing loci: 7745\n",
      "final number of intronic loci: 1353\n",
      "final number of intergenic loci: 1157\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 37540\n",
      "number of splicing transcripts: 19435\n",
      "number of intronic transcripts: 3956\n",
      "number of intergenic transcripts: 29393\n",
      "number of real genes: 19817\n",
      "number of solicing genes: 12678\n",
      "number of intronic genes: 2735\n",
      "number of intergenic genes: 29016\n",
      "targeting 8695 real only loci\n",
      "targeting 2773 splicing only loci\n",
      "targeting 363 intronic only loci\n",
      "targeting 5150 real and splicing loci\n",
      "targeting 337 real and intronic loci\n",
      "targeting 320 splicing and intronic loci\n",
      "targeting 298 real and splicing and intronic loci\n",
      "targeting 1457 intergenic loci\n",
      "available real only loci: 15120\n",
      "available splicing only loci: 7627\n",
      "available intronic only loci: 1462\n",
      "available real and splicing loci: 4096\n",
      "available real and intronic loci: 318\n",
      "available splicing and intronic loci: 672\n",
      "available real and splicing and intronic loci: 283\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8695\n",
      "selecting splicing only loci: 2773\n",
      "selecting intronic only loci: 363\n",
      "selecting real and splicing loci: 4096\n",
      "selecting real and intronic loci: 318\n",
      "selecting splicing and intronic loci: 320\n",
      "selecting real and splicing and intronic loci: 283\n",
      "selecting intergenic loci: 1457\n",
      "final number of real transcripts: 27767\n",
      "final number of splicing transcripts: 12162\n",
      "final number of intronic transcripts: 1952\n",
      "final number of intergenic transcripts: 1471\n",
      "final number of real loci: 13392\n",
      "final number of splicing loci: 7472\n",
      "final number of intronic loci: 1284\n",
      "final number of intergenic loci: 1457\n",
      "\n",
      "=================\n",
      "Tissue #2\n",
      "=================\n",
      "\n",
      "starting number of real loci: 25849\n",
      "starting number of splicing loci: 21232\n",
      "starting number of intronic loci: 11935\n",
      "starting number of intergenic loci: 29016\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 32851\n",
      "number of splicing transcripts: 16948\n",
      "number of intronic transcripts: 3502\n",
      "number of intergenic transcripts: 29329\n",
      "number of real genes: 17305\n",
      "number of solicing genes: 10966\n",
      "number of intronic genes: 2409\n",
      "number of intergenic genes: 29016\n",
      "targeting 7486 real only loci\n",
      "targeting 2692 splicing only loci\n",
      "targeting 176 intronic only loci\n",
      "targeting 3649 real and splicing loci\n",
      "targeting 154 real and intronic loci\n",
      "targeting 160 splicing and intronic loci\n",
      "targeting 120 real and splicing and intronic loci\n",
      "targeting 1021 intergenic loci\n",
      "available real only loci: 13329\n",
      "available splicing only loci: 6677\n",
      "available intronic only loci: 1222\n",
      "available real and splicing loci: 3424\n",
      "available real and intronic loci: 322\n",
      "available splicing and intronic loci: 635\n",
      "available real and splicing and intronic loci: 230\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 7486\n",
      "selecting splicing only loci: 2692\n",
      "selecting intronic only loci: 176\n",
      "selecting real and splicing loci: 3424\n",
      "selecting real and intronic loci: 154\n",
      "selecting splicing and intronic loci: 160\n",
      "selecting real and splicing and intronic loci: 120\n",
      "selecting intergenic loci: 1021\n",
      "final number of real transcripts: 23014\n",
      "final number of splicing transcripts: 10342\n",
      "final number of intronic transcripts: 974\n",
      "final number of intergenic transcripts: 1032\n",
      "final number of real loci: 11184\n",
      "final number of splicing loci: 6396\n",
      "final number of intronic loci: 610\n",
      "final number of intergenic loci: 1021\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 33002\n",
      "number of splicing transcripts: 16964\n",
      "number of intronic transcripts: 3456\n",
      "number of intergenic transcripts: 29347\n",
      "number of real genes: 17419\n",
      "number of solicing genes: 10926\n",
      "number of intronic genes: 2423\n",
      "number of intergenic genes: 29016\n",
      "targeting 8504 real only loci\n",
      "targeting 3047 splicing only loci\n",
      "targeting 254 intronic only loci\n",
      "targeting 4656 real and splicing loci\n",
      "targeting 257 real and intronic loci\n",
      "targeting 282 splicing and intronic loci\n",
      "targeting 208 real and splicing and intronic loci\n",
      "targeting 1214 intergenic loci\n",
      "available real only loci: 13369\n",
      "available splicing only loci: 6551\n",
      "available intronic only loci: 1217\n",
      "available real and splicing loci: 3491\n",
      "available real and intronic loci: 322\n",
      "available splicing and intronic loci: 647\n",
      "available real and splicing and intronic loci: 237\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8504\n",
      "selecting splicing only loci: 3047\n",
      "selecting intronic only loci: 254\n",
      "selecting real and splicing loci: 3491\n",
      "selecting real and intronic loci: 257\n",
      "selecting splicing and intronic loci: 282\n",
      "selecting real and splicing and intronic loci: 208\n",
      "selecting intergenic loci: 1214\n",
      "final number of real transcripts: 25255\n",
      "final number of splicing transcripts: 11405\n",
      "final number of intronic transcripts: 1519\n",
      "final number of intergenic transcripts: 1223\n",
      "final number of real loci: 12460\n",
      "final number of splicing loci: 7028\n",
      "final number of intronic loci: 1001\n",
      "final number of intergenic loci: 1214\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 32650\n",
      "number of splicing transcripts: 16935\n",
      "number of intronic transcripts: 3397\n",
      "number of intergenic transcripts: 29347\n",
      "number of real genes: 17395\n",
      "number of solicing genes: 10873\n",
      "number of intronic genes: 2413\n",
      "number of intergenic genes: 29016\n",
      "targeting 7330 real only loci\n",
      "targeting 2768 splicing only loci\n",
      "targeting 202 intronic only loci\n",
      "targeting 3457 real and splicing loci\n",
      "targeting 213 real and intronic loci\n",
      "targeting 287 splicing and intronic loci\n",
      "targeting 189 real and splicing and intronic loci\n",
      "targeting 829 intergenic loci\n",
      "available real only loci: 13418\n",
      "available splicing only loci: 6589\n",
      "available intronic only loci: 1231\n",
      "available real and splicing loci: 3416\n",
      "available real and intronic loci: 314\n",
      "available splicing and intronic loci: 621\n",
      "available real and splicing and intronic loci: 247\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 7330\n",
      "selecting splicing only loci: 2768\n",
      "selecting intronic only loci: 202\n",
      "selecting real and splicing loci: 3416\n",
      "selecting real and intronic loci: 213\n",
      "selecting splicing and intronic loci: 287\n",
      "selecting real and splicing and intronic loci: 189\n",
      "selecting intergenic loci: 829\n",
      "final number of real transcripts: 22944\n",
      "final number of splicing transcripts: 10917\n",
      "final number of intronic transcripts: 1314\n",
      "final number of intergenic transcripts: 845\n",
      "final number of real loci: 11148\n",
      "final number of splicing loci: 6660\n",
      "final number of intronic loci: 891\n",
      "final number of intergenic loci: 829\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 33090\n",
      "number of splicing transcripts: 17053\n",
      "number of intronic transcripts: 3326\n",
      "number of intergenic transcripts: 29356\n",
      "number of real genes: 17376\n",
      "number of solicing genes: 10934\n",
      "number of intronic genes: 2398\n",
      "number of intergenic genes: 29016\n",
      "targeting 8072 real only loci\n",
      "targeting 3465 splicing only loci\n",
      "targeting 261 intronic only loci\n",
      "targeting 5231 real and splicing loci\n",
      "targeting 217 real and intronic loci\n",
      "targeting 289 splicing and intronic loci\n",
      "targeting 224 real and splicing and intronic loci\n",
      "targeting 1232 intergenic loci\n",
      "available real only loci: 13359\n",
      "available splicing only loci: 6641\n",
      "available intronic only loci: 1201\n",
      "available real and splicing loci: 3435\n",
      "available real and intronic loci: 339\n",
      "available splicing and intronic loci: 615\n",
      "available real and splicing and intronic loci: 243\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8072\n",
      "selecting splicing only loci: 3465\n",
      "selecting intronic only loci: 261\n",
      "selecting real and splicing loci: 3435\n",
      "selecting real and intronic loci: 217\n",
      "selecting splicing and intronic loci: 289\n",
      "selecting real and splicing and intronic loci: 224\n",
      "selecting intergenic loci: 1232\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 24650\n",
      "final number of splicing transcripts: 11987\n",
      "final number of intronic transcripts: 1458\n",
      "final number of intergenic transcripts: 1245\n",
      "final number of real loci: 11948\n",
      "final number of splicing loci: 7413\n",
      "final number of intronic loci: 991\n",
      "final number of intergenic loci: 1232\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 33022\n",
      "number of splicing transcripts: 16928\n",
      "number of intronic transcripts: 3529\n",
      "number of intergenic transcripts: 29342\n",
      "number of real genes: 17388\n",
      "number of solicing genes: 10939\n",
      "number of intronic genes: 2440\n",
      "number of intergenic genes: 29015\n",
      "targeting 8775 real only loci\n",
      "targeting 2703 splicing only loci\n",
      "targeting 181 intronic only loci\n",
      "targeting 3716 real and splicing loci\n",
      "targeting 156 real and intronic loci\n",
      "targeting 161 splicing and intronic loci\n",
      "targeting 133 real and splicing and intronic loci\n",
      "targeting 1060 intergenic loci\n",
      "available real only loci: 13306\n",
      "available splicing only loci: 6603\n",
      "available intronic only loci: 1232\n",
      "available real and splicing loci: 3488\n",
      "available real and intronic loci: 360\n",
      "available splicing and intronic loci: 614\n",
      "available real and splicing and intronic loci: 234\n",
      "available intergenic only loci: 29015\n",
      "selecting real only loci: 8775\n",
      "selecting splicing only loci: 2703\n",
      "selecting intronic only loci: 181\n",
      "selecting real and splicing loci: 3488\n",
      "selecting real and intronic loci: 156\n",
      "selecting splicing and intronic loci: 161\n",
      "selecting real and splicing and intronic loci: 133\n",
      "selecting intergenic loci: 1060\n",
      "final number of real transcripts: 25402\n",
      "final number of splicing transcripts: 10576\n",
      "final number of intronic transcripts: 997\n",
      "final number of intergenic transcripts: 1071\n",
      "final number of real loci: 12552\n",
      "final number of splicing loci: 6485\n",
      "final number of intronic loci: 631\n",
      "final number of intergenic loci: 1060\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 33139\n",
      "number of splicing transcripts: 16938\n",
      "number of intronic transcripts: 3490\n",
      "number of intergenic transcripts: 29367\n",
      "number of real genes: 17361\n",
      "number of solicing genes: 11019\n",
      "number of intronic genes: 2435\n",
      "number of intergenic genes: 29016\n",
      "targeting 6677 real only loci\n",
      "targeting 3869 splicing only loci\n",
      "targeting 250 intronic only loci\n",
      "targeting 2900 real and splicing loci\n",
      "targeting 134 real and intronic loci\n",
      "targeting 329 splicing and intronic loci\n",
      "targeting 127 real and splicing and intronic loci\n",
      "targeting 901 intergenic loci\n",
      "available real only loci: 13259\n",
      "available splicing only loci: 6600\n",
      "available intronic only loci: 1237\n",
      "available real and splicing loci: 3542\n",
      "available real and intronic loci: 321\n",
      "available splicing and intronic loci: 638\n",
      "available real and splicing and intronic loci: 239\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 6677\n",
      "selecting splicing only loci: 3869\n",
      "selecting intronic only loci: 250\n",
      "selecting real and splicing loci: 2900\n",
      "selecting real and intronic loci: 134\n",
      "selecting splicing and intronic loci: 329\n",
      "selecting real and splicing and intronic loci: 127\n",
      "selecting intergenic loci: 901\n",
      "final number of real transcripts: 20129\n",
      "final number of splicing transcripts: 11250\n",
      "final number of intronic transcripts: 1221\n",
      "final number of intergenic transcripts: 908\n",
      "final number of real loci: 9838\n",
      "final number of splicing loci: 7225\n",
      "final number of intronic loci: 840\n",
      "final number of intergenic loci: 901\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 32749\n",
      "number of splicing transcripts: 16656\n",
      "number of intronic transcripts: 3442\n",
      "number of intergenic transcripts: 29345\n",
      "number of real genes: 17418\n",
      "number of solicing genes: 10819\n",
      "number of intronic genes: 2429\n",
      "number of intergenic genes: 29015\n",
      "targeting 8340 real only loci\n",
      "targeting 1979 splicing only loci\n",
      "targeting 178 intronic only loci\n",
      "targeting 3182 real and splicing loci\n",
      "targeting 161 real and intronic loci\n",
      "targeting 100 splicing and intronic loci\n",
      "targeting 78 real and splicing and intronic loci\n",
      "targeting 1034 intergenic loci\n",
      "available real only loci: 13470\n",
      "available splicing only loci: 6524\n",
      "available intronic only loci: 1240\n",
      "available real and splicing loci: 3418\n",
      "available real and intronic loci: 312\n",
      "available splicing and intronic loci: 659\n",
      "available real and splicing and intronic loci: 218\n",
      "available intergenic only loci: 29015\n",
      "selecting real only loci: 8340\n",
      "selecting splicing only loci: 1979\n",
      "selecting intronic only loci: 178\n",
      "selecting real and splicing loci: 3182\n",
      "selecting real and intronic loci: 161\n",
      "selecting splicing and intronic loci: 100\n",
      "selecting real and splicing and intronic loci: 78\n",
      "selecting intergenic loci: 1034\n",
      "final number of real transcripts: 23309\n",
      "final number of splicing transcripts: 8683\n",
      "final number of intronic transcripts: 777\n",
      "final number of intergenic transcripts: 1036\n",
      "final number of real loci: 11761\n",
      "final number of splicing loci: 5339\n",
      "final number of intronic loci: 517\n",
      "final number of intergenic loci: 1034\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 32863\n",
      "number of splicing transcripts: 16906\n",
      "number of intronic transcripts: 3473\n",
      "number of intergenic transcripts: 29348\n",
      "number of real genes: 17389\n",
      "number of solicing genes: 10949\n",
      "number of intronic genes: 2448\n",
      "number of intergenic genes: 29016\n",
      "targeting 9039 real only loci\n",
      "targeting 3543 splicing only loci\n",
      "targeting 231 intronic only loci\n",
      "targeting 3885 real and splicing loci\n",
      "targeting 215 real and intronic loci\n",
      "targeting 236 splicing and intronic loci\n",
      "targeting 146 real and splicing and intronic loci\n",
      "targeting 1231 intergenic loci\n",
      "available real only loci: 13284\n",
      "available splicing only loci: 6596\n",
      "available intronic only loci: 1248\n",
      "available real and splicing loci: 3512\n",
      "available real and intronic loci: 359\n",
      "available splicing and intronic loci: 607\n",
      "available real and splicing and intronic loci: 234\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 9039\n",
      "selecting splicing only loci: 3543\n",
      "selecting intronic only loci: 231\n",
      "selecting real and splicing loci: 3512\n",
      "selecting real and intronic loci: 215\n",
      "selecting splicing and intronic loci: 236\n",
      "selecting real and splicing and intronic loci: 146\n",
      "selecting intergenic loci: 1231\n",
      "final number of real transcripts: 25792\n",
      "final number of splicing transcripts: 11804\n",
      "final number of intronic transcripts: 1238\n",
      "final number of intergenic transcripts: 1242\n",
      "final number of real loci: 12912\n",
      "final number of splicing loci: 7437\n",
      "final number of intronic loci: 828\n",
      "final number of intergenic loci: 1231\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 32702\n",
      "number of splicing transcripts: 16915\n",
      "number of intronic transcripts: 3518\n",
      "number of intergenic transcripts: 29340\n",
      "number of real genes: 17359\n",
      "number of solicing genes: 10944\n",
      "number of intronic genes: 2478\n",
      "number of intergenic genes: 29016\n",
      "targeting 8695 real only loci\n",
      "targeting 3475 splicing only loci\n",
      "targeting 248 intronic only loci\n",
      "targeting 3787 real and splicing loci\n",
      "targeting 201 real and intronic loci\n",
      "targeting 266 splicing and intronic loci\n",
      "targeting 170 real and splicing and intronic loci\n",
      "targeting 1297 intergenic loci\n",
      "available real only loci: 13331\n",
      "available splicing only loci: 6550\n",
      "available intronic only loci: 1239\n",
      "available real and splicing loci: 3474\n",
      "available real and intronic loci: 319\n",
      "available splicing and intronic loci: 685\n",
      "available real and splicing and intronic loci: 235\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8695\n",
      "selecting splicing only loci: 3475\n",
      "selecting intronic only loci: 248\n",
      "selecting real and splicing loci: 3474\n",
      "selecting real and intronic loci: 201\n",
      "selecting splicing and intronic loci: 266\n",
      "selecting real and splicing and intronic loci: 170\n",
      "selecting intergenic loci: 1297\n",
      "final number of real transcripts: 25039\n",
      "final number of splicing transcripts: 11808\n",
      "final number of intronic transcripts: 1350\n",
      "final number of intergenic transcripts: 1305\n",
      "final number of real loci: 12540\n",
      "final number of splicing loci: 7385\n",
      "final number of intronic loci: 885\n",
      "final number of intergenic loci: 1297\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of real transcripts: 33124\n",
      "number of splicing transcripts: 17087\n",
      "number of intronic transcripts: 3567\n",
      "number of intergenic transcripts: 29339\n",
      "number of real genes: 17312\n",
      "number of solicing genes: 10958\n",
      "number of intronic genes: 2393\n",
      "number of intergenic genes: 29016\n",
      "targeting 8294 real only loci\n",
      "targeting 2961 splicing only loci\n",
      "targeting 311 intronic only loci\n",
      "targeting 5455 real and splicing loci\n",
      "targeting 347 real and intronic loci\n",
      "targeting 351 splicing and intronic loci\n",
      "targeting 314 real and splicing and intronic loci\n",
      "targeting 1371 intergenic loci\n",
      "available real only loci: 13330\n",
      "available splicing only loci: 6681\n",
      "available intronic only loci: 1220\n",
      "available real and splicing loci: 3435\n",
      "available real and intronic loci: 331\n",
      "available splicing and intronic loci: 626\n",
      "available real and splicing and intronic loci: 216\n",
      "available intergenic only loci: 29016\n",
      "selecting real only loci: 8294\n",
      "selecting splicing only loci: 2961\n",
      "selecting intronic only loci: 311\n",
      "selecting real and splicing loci: 3435\n",
      "selecting real and intronic loci: 331\n",
      "selecting splicing and intronic loci: 351\n",
      "selecting real and splicing and intronic loci: 216\n",
      "selecting intergenic loci: 1371\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 25416\n",
      "final number of splicing transcripts: 11442\n",
      "final number of intronic transcripts: 1936\n",
      "final number of intergenic transcripts: 1376\n",
      "final number of real loci: 12276\n",
      "final number of splicing loci: 6963\n",
      "final number of intronic loci: 1209\n",
      "final number of intergenic loci: 1371\n"
     ]
    }
   ],
   "source": [
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    realDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.real_t\"+str(tissue_num))\n",
    "    spliceDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.splice_t\"+str(tissue_num))\n",
    "    intDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.int_t\"+str(tissue_num))\n",
    "    polDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.pol_t\"+str(tissue_num))\n",
    "    \n",
    "    print(\"starting number of real loci: \"+str(len(set(realDF[\"lid\"]))))\n",
    "    print(\"starting number of splicing loci: \"+str(len(set(spliceDF[\"lid\"]))))\n",
    "    print(\"starting number of intronic loci: \"+str(len(set(intDF[\"lid\"]))))\n",
    "    print(\"starting number of intergenic loci: \"+str(len(set(polDF[\"lid\"]))))\n",
    "    \n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        # ==================================\n",
    "        # get numbers\n",
    "        # ==================================\n",
    "        realn,splicen,intn,real_splicen,real_intn,splice_intn,real_splice_intn,all_realn,poln = s_loc.sample()[[\"real\",\\\n",
    "                                                                                                                \"splicing\",\\\n",
    "                                                                                                                \"intronic\",\\\n",
    "                                                                                                                \"real_splicing\",\\\n",
    "                                                                                                                \"real_intronic\",\\\n",
    "                                                                                                                \"splicing_intronic\",\\\n",
    "                                                                                                                \"real_splicing_intronic\",\\\n",
    "                                                                                                                \"all_real\",\\\n",
    "                                                                                                                \"intergenic\"]].iloc[0].tolist()\n",
    "        \n",
    "        # ==================================\n",
    "        # Now to generate sample from tissue\n",
    "        # for intergenic set\n",
    "        # ==================================\n",
    "        \n",
    "        # now let's split this into samples the old way\n",
    "        # now let's split this into samples the old way\n",
    "        realDF_g = realDF.groupby(\"lid\").agg({\"tid_real\":{\"tids_real\":lambda x:list(x),\n",
    "                                                                \"count_real\":\"count\"},\n",
    "                                              \"real_tpms\":\"first\"}).reset_index()\n",
    "        realDF_g.columns = [\"lid\",\"tids_real\",\"count_real\",\"real_tpms\"]\n",
    "        realDF_g[\"real_tpms\"] = realDF_g[\"real_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        spliceDF_g = spliceDF.groupby(\"lid\").agg({\"tid_splicing\":{\"tids_splicing\":lambda x:list(x),\n",
    "                                                                \"count_splicing\":\"count\"},\n",
    "                                              \"splicing_tpms\":\"first\"}).reset_index()\n",
    "        spliceDF_g.columns = [\"lid\",\"tids_splicing\",\"count_splicing\",\"splicing_tpms\"]\n",
    "        spliceDF_g[\"splicing_tpms\"] = spliceDF_g[\"splicing_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        intDF_g = intDF.groupby(\"lid\").agg({\"tid_intronic\":{\"tids_intronic\":lambda x:list(x),\n",
    "                                                                \"count_intronic\":\"count\"},\n",
    "                                              \"intronic_tpms\":\"first\"}).reset_index()\n",
    "        intDF_g.columns = [\"lid\",\"tids_intronic\",\"count_intronic\",\"intronic_tpms\"]\n",
    "        intDF_g[\"intronic_tpms\"] = intDF_g[\"intronic_tpms\"].str.split(\";\")\n",
    "\n",
    "        # now let's split this into samples the old way\n",
    "        polDF_g = polDF.groupby(\"lid\").agg({\"tid_intergenic\":{\"tids_intergenic\":lambda x:list(x),\n",
    "                                                                \"count_intergenic\":\"count\"},\n",
    "                                              \"intergenic_tpms\":\"first\"}).reset_index()\n",
    "        polDF_g.columns = [\"lid\",\"tids_intergenic\",\"count_intergenic\",\"intergenic_tpms\"]\n",
    "        polDF_g[\"intergenic_tpms\"] = polDF_g[\"intergenic_tpms\"].str.split(\";\")\n",
    "\n",
    "        all_g_lid = realDF_g.merge(spliceDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid = all_g_lid.merge(intDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid = all_g_lid.merge(polDF_g,how=\"outer\",on=\"lid\")\n",
    "        all_g_lid[\"count_real\"] = all_g_lid[\"count_real\"].fillna(0)\n",
    "        all_g_lid[\"count_splicing\"] = all_g_lid[\"count_splicing\"].fillna(0)\n",
    "        all_g_lid[\"count_intronic\"] = all_g_lid[\"count_intronic\"].fillna(0)\n",
    "        all_g_lid[\"count_intergenic\"] = all_g_lid[\"count_intergenic\"].fillna(0)\n",
    "\n",
    "        all_g_lid[\"count_real\"]=all_g_lid[\"count_real\"].astype(int)\n",
    "        all_g_lid[\"count_splicing\"]=all_g_lid[\"count_splicing\"].astype(int)\n",
    "        all_g_lid[\"count_intronic\"]=all_g_lid[\"count_intronic\"].astype(int)\n",
    "        all_g_lid[\"count_intergenic\"]=all_g_lid[\"count_intergenic\"].astype(int)\n",
    "\n",
    "        all_g_lid.fillna(\"\",inplace=True)\n",
    "\n",
    "        # need a function to get the samples from this now\n",
    "        lids_intergenic=[]\n",
    "        tids_intergenic=[]\n",
    "        tpms_intergenic=[]\n",
    "\n",
    "        lids_real=[]\n",
    "        tids_real=[]\n",
    "        tpms_real=[]\n",
    "\n",
    "        lids_splicing=[]\n",
    "        tids_splicing=[]\n",
    "        tpms_splicing=[]\n",
    "\n",
    "        lids_intronic=[]\n",
    "        tids_intronic=[]\n",
    "        tpms_intronic=[]\n",
    "\n",
    "        def get_samples(row):\n",
    "            global lids_intergenic\n",
    "            global tids_intergenic\n",
    "            global tpms_intergenic\n",
    "            if(row[\"count_intergenic\"]>0):\n",
    "                all_intergenic_tids = np.array(row[\"tids_intergenic\"])\n",
    "                all_intergenic_tpms = row[\"intergenic_tpms\"]\n",
    "                # randomly choose a sample\n",
    "                rand_sample_idx = random.randint(0,len(all_intergenic_tpms)-1)\n",
    "                tmp_tpms = all_intergenic_tpms[rand_sample_idx].split(\"-\")\n",
    "                tid_ids = []\n",
    "                res_tpms = []\n",
    "                for i in range(len(tmp_tpms)):\n",
    "                    if not tmp_tpms[i]==\"0\":\n",
    "                        tid_ids.append(i)\n",
    "                        res_tpms.append(float(tmp_tpms[i]))\n",
    "                res_tids = all_intergenic_tids[tid_ids]\n",
    "                lids_intergenic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                tids_intergenic.extend(res_tids)\n",
    "                tpms_intergenic.extend(res_tpms)\n",
    "            else:\n",
    "                all_real_tpms = row[\"real_tpms\"]\n",
    "                all_splicing_tpms = row[\"splicing_tpms\"]\n",
    "                all_intronic_tpms = row[\"intronic_tpms\"]\n",
    "\n",
    "                all_real_tids = np.array(row[\"tids_real\"])\n",
    "                all_splicing_tids = np.array(row[\"tids_splicing\"])\n",
    "                all_intronic_tids = np.array(row[\"tids_intronic\"])\n",
    "\n",
    "                num_observations = max(len(all_real_tpms),len(all_splicing_tpms),len(all_intronic_tpms))\n",
    "                rand_sample_idx = random.randint(0,num_observations-1)\n",
    "\n",
    "                if row[\"count_real\"]>0:\n",
    "                    assert len(all_real_tpms)==num_observations\n",
    "                    tmp_tpms = all_real_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_real_tids[tid_ids]\n",
    "                    lids_real.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_real.extend(res_tids)\n",
    "                    tpms_real.extend(res_tpms)\n",
    "                if row[\"count_splicing\"]>0:\n",
    "                    assert len(all_splicing_tpms)==num_observations\n",
    "                    tmp_tpms = all_splicing_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_splicing_tids[tid_ids]\n",
    "                    lids_splicing.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_splicing.extend(res_tids)\n",
    "                    tpms_splicing.extend(res_tpms)\n",
    "                if row[\"count_intronic\"]>0:\n",
    "                    assert len(all_intronic_tpms)==num_observations\n",
    "                    tmp_tpms = all_intronic_tpms[rand_sample_idx].split(\"-\")\n",
    "                    tid_ids = []\n",
    "                    res_tpms = []\n",
    "                    for i in range(len(tmp_tpms)):\n",
    "                        if not tmp_tpms[i]==\"0\":\n",
    "                            tid_ids.append(i)\n",
    "                            res_tpms.append(float(tmp_tpms[i]))\n",
    "                    res_tids = all_intronic_tids[tid_ids]\n",
    "                    lids_intronic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "                    tids_intronic.extend(res_tids)\n",
    "                    tpms_intronic.extend(res_tpms)\n",
    "\n",
    "        all_g_lid.apply(lambda row: get_samples(row),axis=1)\n",
    "        \n",
    "        print(\"number of real transcripts: \"+str(len(set(tids_real))))\n",
    "        print(\"number of splicing transcripts: \"+str(len(set(tids_splicing))))\n",
    "        print(\"number of intronic transcripts: \"+str(len(set(tids_intronic))))\n",
    "        print(\"number of intergenic transcripts: \"+str(len(set(tids_intergenic))))\n",
    "        print(\"number of real genes: \"+str(len(set(lids_real))))\n",
    "        print(\"number of solicing genes: \"+str(len(set(lids_splicing))))\n",
    "        print(\"number of intronic genes: \"+str(len(set(lids_intronic))))\n",
    "        print(\"number of intergenic genes: \"+str(len(set(lids_intergenic))))\n",
    "\n",
    "        print(\"targeting \"+str(realn)+\" real only loci\")\n",
    "        print(\"targeting \"+str(splicen)+\" splicing only loci\")\n",
    "        print(\"targeting \"+str(intn)+\" intronic only loci\")\n",
    "        print(\"targeting \"+str(real_splicen)+\" real and splicing loci\")\n",
    "        print(\"targeting \"+str(real_intn)+\" real and intronic loci\")\n",
    "        print(\"targeting \"+str(splice_intn)+\" splicing and intronic loci\")\n",
    "        print(\"targeting \"+str(real_splice_intn)+\" real and splicing and intronic loci\")\n",
    "        print(\"targeting \"+str(poln)+\" intergenic loci\")\n",
    "\n",
    "        real_sample_locs = set(lids_real)\n",
    "        splicing_sample_locs = set(lids_splicing)\n",
    "        intronic_sample_locs = set(lids_intronic)\n",
    "        intergenic_sample_locs = list(set(lids_intergenic))\n",
    "\n",
    "        real_only_sample_locs = list(real_sample_locs-(splicing_sample_locs.union(intronic_sample_locs)))\n",
    "        splicing_only_sample_locs = list(splicing_sample_locs-(real_sample_locs.union(intronic_sample_locs)))\n",
    "        intronic_only_sample_locs = list(intronic_sample_locs-(real_sample_locs.union(splicing_sample_locs)))\n",
    "\n",
    "        real_splicing_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs)-intronic_sample_locs)\n",
    "        real_intronic_sample_locs = list(real_sample_locs.intersection(intronic_sample_locs)-splicing_sample_locs)\n",
    "        splicing_intronic_sample_locs = list(splicing_sample_locs.intersection(intronic_sample_locs)-real_sample_locs)\n",
    "\n",
    "        real_splicing_intronic_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs.intersection(intronic_sample_locs)))\n",
    "\n",
    "        print(\"available real only loci: \"+str(len(real_only_sample_locs)))\n",
    "        print(\"available splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "        print(\"available intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "        print(\"available real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "        print(\"available real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "        print(\"available splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "        print(\"available real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "        print(\"available intergenic only loci: \"+str(len(intergenic_sample_locs)))\n",
    "\n",
    "        random.shuffle(real_only_sample_locs)\n",
    "        min_realn = min(len(real_only_sample_locs),realn)\n",
    "        real_only_sample_locs = real_only_sample_locs[:min_realn]\n",
    "        random.shuffle(splicing_only_sample_locs)\n",
    "        min_splicen = min(len(splicing_only_sample_locs),splicen)\n",
    "        splicing_only_sample_locs = splicing_only_sample_locs[:min_splicen]\n",
    "        random.shuffle(intronic_only_sample_locs)\n",
    "        min_intn = min(len(intronic_only_sample_locs),intn)\n",
    "        intronic_only_sample_locs = intronic_only_sample_locs[:min_intn]\n",
    "        random.shuffle(intergenic_sample_locs)\n",
    "        min_poln = min(len(intergenic_sample_locs),poln)\n",
    "        intergenic_sample_locs = intergenic_sample_locs[:min_poln]\n",
    "\n",
    "        random.shuffle(real_splicing_sample_locs)\n",
    "        min_real_splicen = min(len(real_splicing_sample_locs),real_splicen)\n",
    "        real_splicing_sample_locs = real_splicing_sample_locs[:min_real_splicen]\n",
    "        random.shuffle(real_intronic_sample_locs)\n",
    "        min_real_intn = min(len(real_intronic_sample_locs),real_intn)\n",
    "        real_intronic_sample_locs = real_intronic_sample_locs[:min_real_intn]\n",
    "        random.shuffle(splicing_intronic_sample_locs)\n",
    "        min_splice_intn = min(len(splicing_intronic_sample_locs),splice_intn)\n",
    "        splicing_intronic_sample_locs = splicing_intronic_sample_locs[:min_splice_intn]\n",
    "\n",
    "        random.shuffle(real_splicing_intronic_sample_locs)\n",
    "        min_real_splice_intn = min(len(real_splicing_intronic_sample_locs),real_splice_intn)\n",
    "        real_splicing_intronic_sample_locs = real_splicing_intronic_sample_locs[:min_real_splice_intn]\n",
    "\n",
    "        print(\"selecting real only loci: \"+str(len(real_only_sample_locs)))\n",
    "        print(\"selecting splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "        print(\"selecting intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "        print(\"selecting real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "        print(\"selecting real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "        print(\"selecting splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "        print(\"selecting real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "        print(\"selecting intergenic loci: \"+str(len(intergenic_sample_locs)))\n",
    "        \n",
    "        real_sub = pd.concat([pd.DataFrame(lids_real,columns=[\"lid\"]),pd.DataFrame(tids_real,columns=[\"tid\"]),pd.DataFrame(tpms_real,columns=[\"tpm\"])],axis=1)\n",
    "        splicing_sub = pd.concat([pd.DataFrame(lids_splicing,columns=[\"lid\"]),pd.DataFrame(tids_splicing,columns=[\"tid\"]),pd.DataFrame(tpms_splicing,columns=[\"tpm\"])],axis=1)\n",
    "        intronic_sub = pd.concat([pd.DataFrame(lids_intronic,columns=[\"lid\"]),pd.DataFrame(tids_intronic,columns=[\"tid\"]),pd.DataFrame(tpms_intronic,columns=[\"tpm\"])],axis=1)\n",
    "        intergenic_sub = pd.concat([pd.DataFrame(lids_intergenic,columns=[\"lid\"]),pd.DataFrame(tids_intergenic,columns=[\"tid\"]),pd.DataFrame(tpms_intergenic,columns=[\"tpm\"])],axis=1)\n",
    "        \n",
    "        real_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_only_sample_locs)].reset_index(drop=True)\n",
    "        splicing_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_only_sample_locs)].reset_index(drop=True)\n",
    "        intronic_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(intronic_only_sample_locs)].reset_index(drop=True)\n",
    "        intergenic_sub2 = intergenic_sub[intergenic_sub[\"lid\"].isin(intergenic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_splicing_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "        splicing_real_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_real_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        splicing_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_splicing_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        splicing_real_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "        intronic_real_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "        real_sub2 = pd.concat([real_only_sample_df,\\\n",
    "                               real_splicing_only_sample_df,\\\n",
    "                               real_intronic_only_sample_df,\\\n",
    "                               real_splicing_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "        splicing_sub2 = pd.concat([splicing_only_sample_df,\\\n",
    "                                   splicing_real_only_sample_df,\\\n",
    "                                   splicing_intronic_only_sample_df,\\\n",
    "                                   splicing_real_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "        intronic_sub2 = pd.concat([intronic_only_sample_df,\\\n",
    "                                   intronic_real_only_sample_df,\\\n",
    "                                   intronic_splicing_only_sample_df,\\\n",
    "                                   intronic_real_splicing_only_sample_df],axis=0).reset_index(drop=True)\n",
    "\n",
    "        print(\"final number of real transcripts: \"+str(len(set(real_sub2[\"tid\"]))))\n",
    "        print(\"final number of splicing transcripts: \"+str(len(set(splicing_sub2[\"tid\"]))))\n",
    "        print(\"final number of intronic transcripts: \"+str(len(set(intronic_sub2[\"tid\"]))))\n",
    "        print(\"final number of intergenic transcripts: \"+str(len(set(intergenic_sub2[\"tid\"]))))\n",
    "        print(\"final number of real loci: \"+str(len(set(real_sub2[\"lid\"]))))\n",
    "        print(\"final number of splicing loci: \"+str(len(set(splicing_sub2[\"lid\"]))))\n",
    "        print(\"final number of intronic loci: \"+str(len(set(intronic_sub2[\"lid\"]))))\n",
    "        print(\"final number of intergenic loci: \"+str(len(set(intergenic_sub2[\"lid\"]))))\n",
    "\n",
    "        ttx_real = list(real_sub2[\"tid\"])\n",
    "        tpm_real = list(real_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_splice = list(splicing_sub2[\"tid\"])\n",
    "        tpm_splice = list(splicing_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_int = list(intronic_sub2[\"tid\"])\n",
    "        tpm_int = list(intronic_sub2[\"tpm\"])\n",
    "\n",
    "        ttx_intergenic = list(intergenic_sub2[\"tid\"])\n",
    "        tpm_intergenic = list(intergenic_sub2[\"tpm\"])\n",
    "\n",
    "        # now we can save the results\n",
    "        res_real = pd.concat([pd.DataFrame(ttx_real,columns=[\"tid_real\"]),pd.DataFrame(tpm_real,columns=[\"tpm\"])],axis=1)\n",
    "        realDF_f = realDF[[\"lid\",\"tid_real\"]].merge(res_real,on=\"tid_real\",how=\"inner\")\n",
    "        realDF_f = real_gtf.merge(realDF_f[[\"tid_real\",\"tpm\"]],on=\"tid_real\",how=\"inner\")\n",
    "        realDF_f[\"attributes\"] = realDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "                                    \";sim_tpm=\"+realDF_f[\"tpm\"].astype(str)\n",
    "        realDF_f[gff3cols].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        realDF_f[realDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "\n",
    "        res_splicing = pd.concat([pd.DataFrame(ttx_splice,columns=[\"tid_splicing\"]),pd.DataFrame(tpm_splice,columns=[\"tpm\"])],axis=1)\n",
    "        splicingDF_f = spliceDF[[\"lid\",\"tid_splicing\"]].merge(res_splicing,on=\"tid_splicing\",how=\"inner\")\n",
    "        splicingDF_f = splice_gtf.merge(splicingDF_f[[\"tid_splicing\",\"tpm\"]],on=\"tid_splicing\",how=\"inner\")\n",
    "        splicingDF_f[\"attributes\"] = splicingDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "            \";sim_tpm=\"+splicingDF_f[\"tpm\"].astype(str)\n",
    "        splicingDF_f[gff3cols].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        splicingDF_f[splicingDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "        splicingDF_f.head()\n",
    "\n",
    "        res_intronic = pd.concat([pd.DataFrame(ttx_int,columns=[\"tid_intronic\"]),pd.DataFrame(tpm_int,columns=[\"tpm\"])],axis=1)\n",
    "        intronicDF_f = intDF[[\"lid\",\"tid_intronic\"]].merge(res_intronic,on=\"tid_intronic\",how=\"inner\")\n",
    "        intronicDF_f = int_gtf.merge(intronicDF_f[[\"tid_intronic\",\"tpm\"]],on=\"tid_intronic\",how=\"inner\")\n",
    "        intronicDF_f[\"attributes\"] = intronicDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "            \";sim_tpm=\"+intronicDF_f[\"tpm\"].astype(str)\n",
    "        intronicDF_f[gff3cols].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        intronicDF_f[intronicDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n",
    "\n",
    "        res_intergenic = pd.concat([pd.DataFrame(ttx_intergenic,columns=[\"tid_intergenic\"]),pd.DataFrame(tpm_intergenic,columns=[\"tpm\"])],axis=1)\n",
    "        intergenicDF_f = polDF[[\"lid\",\"tid_intergenic\"]].merge(res_intergenic,on=\"tid_intergenic\",how=\"inner\")\n",
    "        intergenicDF_f = pol_gtf.merge(intergenicDF_f[[\"tid_intergenic\",\"tpm\"]],on=\"tid_intergenic\",how=\"inner\")\n",
    "        intergenicDF_f[\"attributes\"] = intergenicDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "                                    \";sim_tpm=\"+intergenicDF_f[\"tpm\"].astype(str)\n",
    "        intergenicDF_f[gff3cols].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",index=False,header=False,quoting=csv.QUOTE_NONE)\n",
    "        intergenicDF_f[intergenicDF_f[\"type\"]==\"transcript\"][[\"tpm\"]].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",index=False,header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lid</th>\n",
       "      <th>tid_real</th>\n",
       "      <th>real_tpms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1</td>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>0;0;0;0;0;0.68;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100.2</td>\n",
       "      <td>0-0-0;0-0-0.42;0-0-3.04;0.44-0-1.11;0-2.05-0;0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>CHS.10000.12</td>\n",
       "      <td>0-0-9.23-0-0-0-0-0;0-2.73-3.86-0-0-0-0-0;0-0-6...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         lid      tid_real                                          real_tpms\n",
       "0      CHS.1       CHS.1.1  0;0;0;0;0;0.68;0;0;0;0;0;0;0;0;0;0;0;0;0;0;1.0...\n",
       "1    CHS.100     CHS.100.2  0-0-0;0-0-0.42;0-0-3.04;0.44-0-1.11;0-2.05-0;0...\n",
       "2    CHS.100     CHS.100.4                                                NaN\n",
       "3    CHS.100     CHS.100.5                                                NaN\n",
       "4  CHS.10000  CHS.10000.12  0-0-9.23-0-0-0-0-0;0-2.73-3.86-0-0-0-0-0;0-0-6..."
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# as seen in test.ipynb we currently have duplicate transcripts in the file\n",
    "# this needs to be fixed!!!\n",
    "realDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.real_t2\")\n",
    "spliceDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.splice_t2\")\n",
    "intDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.int_t2\")\n",
    "polDF = pd.read_csv(out_dir+\"stage2_tid_lid_exp.pol_t2\")\n",
    "realDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lid</th>\n",
       "      <th>tid_real</th>\n",
       "      <th>real_tpms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39868</th>\n",
       "      <td>CHS.3</td>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>9.40;8.90;19.33;16.20;10.71;9.26;12.01;18.20;2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         lid tid_real                                          real_tpms\n",
       "39868  CHS.3  CHS.3.2  9.40;8.90;19.33;16.20;10.71;9.26;12.01;18.20;2..."
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "realDF[realDF[\"tid_real\"]==\"CHS.3.2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "realn,splicen,intn,real_splicen,real_intn,splice_intn,real_splice_intn,all_realn,poln = s_loc.sample()[[\"real\",\\\n",
    "                                                                                                        \"splicing\",\\\n",
    "                                                                                                        \"intronic\",\\\n",
    "                                                                                                        \"real_splicing\",\\\n",
    "                                                                                                        \"real_intronic\",\\\n",
    "                                                                                                        \"splicing_intronic\",\\\n",
    "                                                                                                        \"real_splicing_intronic\",\\\n",
    "                                                                                                        \"all_real\",\\\n",
    "                                                                                                        \"intergenic\"]].iloc[0].tolist()\n",
    "\n",
    "# ==================================\n",
    "# Now to generate sample from tissue\n",
    "# for intergenic set\n",
    "# ==================================\n",
    "\n",
    "# now let's split this into samples the old way\n",
    "# now let's split this into samples the old way\n",
    "realDF_g = realDF.groupby(\"lid\").agg({\"tid_real\":{\"tids_real\":lambda x:list(x),\n",
    "                                                        \"count_real\":\"count\"},\n",
    "                                      \"real_tpms\":\"first\"}).reset_index()\n",
    "realDF_g.columns = [\"lid\",\"tids_real\",\"count_real\",\"real_tpms\"]\n",
    "realDF_g[\"real_tpms\"] = realDF_g[\"real_tpms\"].str.split(\";\")\n",
    "\n",
    "# now let's split this into samples the old way\n",
    "spliceDF_g = spliceDF.groupby(\"lid\").agg({\"tid_splicing\":{\"tids_splicing\":lambda x:list(x),\n",
    "                                                        \"count_splicing\":\"count\"},\n",
    "                                      \"splicing_tpms\":\"first\"}).reset_index()\n",
    "spliceDF_g.columns = [\"lid\",\"tids_splicing\",\"count_splicing\",\"splicing_tpms\"]\n",
    "spliceDF_g[\"splicing_tpms\"] = spliceDF_g[\"splicing_tpms\"].str.split(\";\")\n",
    "\n",
    "# now let's split this into samples the old way\n",
    "intDF_g = intDF.groupby(\"lid\").agg({\"tid_intronic\":{\"tids_intronic\":lambda x:list(x),\n",
    "                                                        \"count_intronic\":\"count\"},\n",
    "                                      \"intronic_tpms\":\"first\"}).reset_index()\n",
    "intDF_g.columns = [\"lid\",\"tids_intronic\",\"count_intronic\",\"intronic_tpms\"]\n",
    "intDF_g[\"intronic_tpms\"] = intDF_g[\"intronic_tpms\"].str.split(\";\")\n",
    "\n",
    "# now let's split this into samples the old way\n",
    "polDF_g = polDF.groupby(\"lid\").agg({\"tid_intergenic\":{\"tids_intergenic\":lambda x:list(x),\n",
    "                                                        \"count_intergenic\":\"count\"},\n",
    "                                      \"intergenic_tpms\":\"first\"}).reset_index()\n",
    "polDF_g.columns = [\"lid\",\"tids_intergenic\",\"count_intergenic\",\"intergenic_tpms\"]\n",
    "polDF_g[\"intergenic_tpms\"] = polDF_g[\"intergenic_tpms\"].str.split(\";\")\n",
    "\n",
    "all_g_lid = realDF_g.merge(spliceDF_g,how=\"outer\",on=\"lid\")\n",
    "all_g_lid = all_g_lid.merge(intDF_g,how=\"outer\",on=\"lid\")\n",
    "all_g_lid = all_g_lid.merge(polDF_g,how=\"outer\",on=\"lid\")\n",
    "all_g_lid[\"count_real\"] = all_g_lid[\"count_real\"].fillna(0)\n",
    "all_g_lid[\"count_splicing\"] = all_g_lid[\"count_splicing\"].fillna(0)\n",
    "all_g_lid[\"count_intronic\"] = all_g_lid[\"count_intronic\"].fillna(0)\n",
    "all_g_lid[\"count_intergenic\"] = all_g_lid[\"count_intergenic\"].fillna(0)\n",
    "\n",
    "all_g_lid[\"count_real\"]=all_g_lid[\"count_real\"].astype(int)\n",
    "all_g_lid[\"count_splicing\"]=all_g_lid[\"count_splicing\"].astype(int)\n",
    "all_g_lid[\"count_intronic\"]=all_g_lid[\"count_intronic\"].astype(int)\n",
    "all_g_lid[\"count_intergenic\"]=all_g_lid[\"count_intergenic\"].astype(int)\n",
    "\n",
    "all_g_lid.fillna(\"\",inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lid</th>\n",
       "      <th>tids_real</th>\n",
       "      <th>count_real</th>\n",
       "      <th>real_tpms</th>\n",
       "      <th>tids_splicing</th>\n",
       "      <th>count_splicing</th>\n",
       "      <th>splicing_tpms</th>\n",
       "      <th>tids_intronic</th>\n",
       "      <th>count_intronic</th>\n",
       "      <th>intronic_tpms</th>\n",
       "      <th>tids_intergenic</th>\n",
       "      <th>count_intergenic</th>\n",
       "      <th>intergenic_tpms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1</td>\n",
       "      <td>[CHS.1.1]</td>\n",
       "      <td>1</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0.68, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[ALL_00359026, ALL_02662430, ALL_02081871, ALL...</td>\n",
       "      <td>14</td>\n",
       "      <td>[0-0-1.28-0-0-0-0-0-0-0-0-0-0-0, 0-0-0.85-0-0-...</td>\n",
       "      <td>[ALL_00407336]</td>\n",
       "      <td>1</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.100</td>\n",
       "      <td>[CHS.100.2, CHS.100.4, CHS.100.5]</td>\n",
       "      <td>3</td>\n",
       "      <td>[0-0-0, 0-0-0.42, 0-0-3.04, 0.44-0-1.11, 0-2.0...</td>\n",
       "      <td>[ALL_01178735, ALL_00210202]</td>\n",
       "      <td>2</td>\n",
       "      <td>[12.17-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>[CHS.10000.12, CHS.10000.26, CHS.10000.18, CHS...</td>\n",
       "      <td>8</td>\n",
       "      <td>[0-0-9.23-0-0-0-0-0, 0-2.73-3.86-0-0-0-0-0, 0-...</td>\n",
       "      <td>[ALL_04211943, ALL_04212040, ALL_04323137, ALL...</td>\n",
       "      <td>5</td>\n",
       "      <td>[0-0-0-0-0, 0-0-0-0-0, 0-0-0-0-0, 0-0-0-0-0, 0...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10001</td>\n",
       "      <td>[CHS.10001.1]</td>\n",
       "      <td>1</td>\n",
       "      <td>[0, 6.59]</td>\n",
       "      <td>[ALL_04811006, ALL_04288012, ALL_04894318]</td>\n",
       "      <td>3</td>\n",
       "      <td>[0-1.04-0.96, 4.43-0-0]</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10002</td>\n",
       "      <td>[CHS.10002.4, CHS.10002.3, CHS.10002.2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[0-0-1.65, 0-0-2.47, 0-0-2.13, 0-0-0, 0-0-2.16...</td>\n",
       "      <td>[ALL_04288028, ALL_04811019, ALL_05116577, ALL...</td>\n",
       "      <td>26</td>\n",
       "      <td>[0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0...</td>\n",
       "      <td>[ALL_05228546, ALL_04995954, ALL_04860726, ALL...</td>\n",
       "      <td>30</td>\n",
       "      <td>[0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         lid                                          tids_real  count_real  \\\n",
       "0      CHS.1                                          [CHS.1.1]           1   \n",
       "1    CHS.100                  [CHS.100.2, CHS.100.4, CHS.100.5]           3   \n",
       "2  CHS.10000  [CHS.10000.12, CHS.10000.26, CHS.10000.18, CHS...           8   \n",
       "3  CHS.10001                                      [CHS.10001.1]           1   \n",
       "4  CHS.10002            [CHS.10002.4, CHS.10002.3, CHS.10002.2]           3   \n",
       "\n",
       "                                           real_tpms  \\\n",
       "0  [0, 0, 0, 0, 0, 0.68, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "1  [0-0-0, 0-0-0.42, 0-0-3.04, 0.44-0-1.11, 0-2.0...   \n",
       "2  [0-0-9.23-0-0-0-0-0, 0-2.73-3.86-0-0-0-0-0, 0-...   \n",
       "3                                          [0, 6.59]   \n",
       "4  [0-0-1.65, 0-0-2.47, 0-0-2.13, 0-0-0, 0-0-2.16...   \n",
       "\n",
       "                                       tids_splicing  count_splicing  \\\n",
       "0  [ALL_00359026, ALL_02662430, ALL_02081871, ALL...              14   \n",
       "1                       [ALL_01178735, ALL_00210202]               2   \n",
       "2  [ALL_04211943, ALL_04212040, ALL_04323137, ALL...               5   \n",
       "3         [ALL_04811006, ALL_04288012, ALL_04894318]               3   \n",
       "4  [ALL_04288028, ALL_04811019, ALL_05116577, ALL...              26   \n",
       "\n",
       "                                       splicing_tpms  \\\n",
       "0  [0-0-1.28-0-0-0-0-0-0-0-0-0-0-0, 0-0-0.85-0-0-...   \n",
       "1  [12.17-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0-0, 0...   \n",
       "2  [0-0-0-0-0, 0-0-0-0-0, 0-0-0-0-0, 0-0-0-0-0, 0...   \n",
       "3                            [0-1.04-0.96, 4.43-0-0]   \n",
       "4  [0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0...   \n",
       "\n",
       "                                       tids_intronic  count_intronic  \\\n",
       "0                                     [ALL_00407336]               1   \n",
       "1                                                                  0   \n",
       "2                                                                  0   \n",
       "3                                                                  0   \n",
       "4  [ALL_05228546, ALL_04995954, ALL_04860726, ALL...              30   \n",
       "\n",
       "                                       intronic_tpms tids_intergenic  \\\n",
       "0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...                   \n",
       "1                                                                      \n",
       "2                                                                      \n",
       "3                                                                      \n",
       "4  [0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0...                   \n",
       "\n",
       "   count_intergenic intergenic_tpms  \n",
       "0                 0                  \n",
       "1                 0                  \n",
       "2                 0                  \n",
       "3                 0                  \n",
       "4                 0                  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_g_lid.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lid</th>\n",
       "      <th>tids_real</th>\n",
       "      <th>count_real</th>\n",
       "      <th>real_tpms</th>\n",
       "      <th>tids_splicing</th>\n",
       "      <th>count_splicing</th>\n",
       "      <th>splicing_tpms</th>\n",
       "      <th>tids_intronic</th>\n",
       "      <th>count_intronic</th>\n",
       "      <th>intronic_tpms</th>\n",
       "      <th>tids_intergenic</th>\n",
       "      <th>count_intergenic</th>\n",
       "      <th>intergenic_tpms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9716</th>\n",
       "      <td>CHS.3</td>\n",
       "      <td>[CHS.3.2]</td>\n",
       "      <td>1</td>\n",
       "      <td>[9.40, 8.90, 19.33, 16.20, 10.71, 9.26, 12.01,...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        lid  tids_real  count_real  \\\n",
       "9716  CHS.3  [CHS.3.2]           1   \n",
       "\n",
       "                                              real_tpms tids_splicing  \\\n",
       "9716  [9.40, 8.90, 19.33, 16.20, 10.71, 9.26, 12.01,...                 \n",
       "\n",
       "      count_splicing splicing_tpms tids_intronic  count_intronic  \\\n",
       "9716               0                                           0   \n",
       "\n",
       "     intronic_tpms tids_intergenic  count_intergenic intergenic_tpms  \n",
       "9716                                               0                  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_g_lid[all_g_lid[\"lid\"]==\"CHS.3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        None\n",
       "1        None\n",
       "2        None\n",
       "3        None\n",
       "4        None\n",
       "         ... \n",
       "54860    None\n",
       "54861    None\n",
       "54862    None\n",
       "54863    None\n",
       "54864    None\n",
       "Length: 54865, dtype: object"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need a function to get the samples from this now\n",
    "lids_intergenic=[]\n",
    "tids_intergenic=[]\n",
    "tpms_intergenic=[]\n",
    "\n",
    "lids_real=[]\n",
    "tids_real=[]\n",
    "tpms_real=[]\n",
    "\n",
    "lids_splicing=[]\n",
    "tids_splicing=[]\n",
    "tpms_splicing=[]\n",
    "\n",
    "lids_intronic=[]\n",
    "tids_intronic=[]\n",
    "tpms_intronic=[]\n",
    "\n",
    "def get_samples(row):\n",
    "    global lids_intergenic\n",
    "    global tids_intergenic\n",
    "    global tpms_intergenic\n",
    "    if(row[\"count_intergenic\"]>0):\n",
    "        all_intergenic_tids = np.array(row[\"tids_intergenic\"])\n",
    "        all_intergenic_tpms = row[\"intergenic_tpms\"]\n",
    "        # randomly choose a sample\n",
    "        rand_sample_idx = random.randint(0,len(all_intergenic_tpms)-1)\n",
    "        tmp_tpms = all_intergenic_tpms[rand_sample_idx].split(\"-\")\n",
    "        tid_ids = []\n",
    "        res_tpms = []\n",
    "        for i in range(len(tmp_tpms)):\n",
    "            if not tmp_tpms[i]==\"0\":\n",
    "                tid_ids.append(i)\n",
    "                res_tpms.append(float(tmp_tpms[i]))\n",
    "        res_tids = all_intergenic_tids[tid_ids]\n",
    "        lids_intergenic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "        tids_intergenic.extend(res_tids)\n",
    "        tpms_intergenic.extend(res_tpms)\n",
    "    else:\n",
    "        all_real_tpms = row[\"real_tpms\"]\n",
    "        all_splicing_tpms = row[\"splicing_tpms\"]\n",
    "        all_intronic_tpms = row[\"intronic_tpms\"]\n",
    "\n",
    "        all_real_tids = np.array(row[\"tids_real\"])\n",
    "        all_splicing_tids = np.array(row[\"tids_splicing\"])\n",
    "        all_intronic_tids = np.array(row[\"tids_intronic\"])\n",
    "\n",
    "        num_observations = max(len(all_real_tpms),len(all_splicing_tpms),len(all_intronic_tpms))\n",
    "        rand_sample_idx = random.randint(0,num_observations-1)\n",
    "\n",
    "        if row[\"count_real\"]>0:\n",
    "            assert len(all_real_tpms)==num_observations\n",
    "            tmp_tpms = all_real_tpms[rand_sample_idx].split(\"-\")\n",
    "            tid_ids = []\n",
    "            res_tpms = []\n",
    "            for i in range(len(tmp_tpms)):\n",
    "                if not tmp_tpms[i]==\"0\":\n",
    "                    tid_ids.append(i)\n",
    "                    res_tpms.append(float(tmp_tpms[i]))\n",
    "            res_tids = all_real_tids[tid_ids]\n",
    "            lids_real.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "            tids_real.extend(res_tids)\n",
    "            tpms_real.extend(res_tpms)\n",
    "        if row[\"count_splicing\"]>0:\n",
    "            assert len(all_splicing_tpms)==num_observations\n",
    "            tmp_tpms = all_splicing_tpms[rand_sample_idx].split(\"-\")\n",
    "            tid_ids = []\n",
    "            res_tpms = []\n",
    "            for i in range(len(tmp_tpms)):\n",
    "                if not tmp_tpms[i]==\"0\":\n",
    "                    tid_ids.append(i)\n",
    "                    res_tpms.append(float(tmp_tpms[i]))\n",
    "            res_tids = all_splicing_tids[tid_ids]\n",
    "            lids_splicing.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "            tids_splicing.extend(res_tids)\n",
    "            tpms_splicing.extend(res_tpms)\n",
    "        if row[\"count_intronic\"]>0:\n",
    "            assert len(all_intronic_tpms)==num_observations\n",
    "            tmp_tpms = all_intronic_tpms[rand_sample_idx].split(\"-\")\n",
    "            tid_ids = []\n",
    "            res_tpms = []\n",
    "            for i in range(len(tmp_tpms)):\n",
    "                if not tmp_tpms[i]==\"0\":\n",
    "                    tid_ids.append(i)\n",
    "                    res_tpms.append(float(tmp_tpms[i]))\n",
    "            res_tids = all_intronic_tids[tid_ids]\n",
    "            lids_intronic.extend([row[\"lid\"] for i in range(len(res_tids))])\n",
    "            tids_intronic.extend(res_tids)\n",
    "            tpms_intronic.extend(res_tpms)\n",
    "\n",
    "all_g_lid.apply(lambda row: get_samples(row),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33134\n",
      "33134\n"
     ]
    }
   ],
   "source": [
    "print(len(tids_real))\n",
    "print(len(set(tids_real)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of real transcripts: 33134\n",
      "number of splicing transcripts: 16730\n",
      "number of intronic transcripts: 3361\n",
      "number of intergenic transcripts: 29363\n",
      "number of real genes: 17458\n",
      "number of solicing genes: 10866\n",
      "number of intronic genes: 2376\n",
      "number of intergenic genes: 29016\n",
      "targeting 8925 real only loci\n",
      "targeting 2731 splicing only loci\n",
      "targeting 219 intronic only loci\n",
      "targeting 4439 real and splicing loci\n",
      "targeting 236 real and intronic loci\n",
      "targeting 204 splicing and intronic loci\n",
      "targeting 146 real and splicing and intronic loci\n",
      "targeting 1171 intergenic loci\n",
      "available real only loci: 13416\n",
      "available splicing only loci: 6603\n",
      "available intronic only loci: 1201\n",
      "available real and splicing loci: 3445\n",
      "available real and intronic loci: 357\n",
      "available splicing and intronic loci: 578\n",
      "available real and splicing and intronic loci: 240\n",
      "available intergenic only loci: 29016\n"
     ]
    }
   ],
   "source": [
    "print(\"number of real transcripts: \"+str(len(set(tids_real))))\n",
    "print(\"number of splicing transcripts: \"+str(len(set(tids_splicing))))\n",
    "print(\"number of intronic transcripts: \"+str(len(set(tids_intronic))))\n",
    "print(\"number of intergenic transcripts: \"+str(len(set(tids_intergenic))))\n",
    "print(\"number of real genes: \"+str(len(set(lids_real))))\n",
    "print(\"number of solicing genes: \"+str(len(set(lids_splicing))))\n",
    "print(\"number of intronic genes: \"+str(len(set(lids_intronic))))\n",
    "print(\"number of intergenic genes: \"+str(len(set(lids_intergenic))))\n",
    "\n",
    "print(\"targeting \"+str(realn)+\" real only loci\")\n",
    "print(\"targeting \"+str(splicen)+\" splicing only loci\")\n",
    "print(\"targeting \"+str(intn)+\" intronic only loci\")\n",
    "print(\"targeting \"+str(real_splicen)+\" real and splicing loci\")\n",
    "print(\"targeting \"+str(real_intn)+\" real and intronic loci\")\n",
    "print(\"targeting \"+str(splice_intn)+\" splicing and intronic loci\")\n",
    "print(\"targeting \"+str(real_splice_intn)+\" real and splicing and intronic loci\")\n",
    "print(\"targeting \"+str(poln)+\" intergenic loci\")\n",
    "\n",
    "real_sample_locs = set(lids_real)\n",
    "splicing_sample_locs = set(lids_splicing)\n",
    "intronic_sample_locs = set(lids_intronic)\n",
    "intergenic_sample_locs = list(set(lids_intergenic))\n",
    "\n",
    "real_only_sample_locs = list(real_sample_locs-(splicing_sample_locs.union(intronic_sample_locs)))\n",
    "splicing_only_sample_locs = list(splicing_sample_locs-(real_sample_locs.union(intronic_sample_locs)))\n",
    "intronic_only_sample_locs = list(intronic_sample_locs-(real_sample_locs.union(splicing_sample_locs)))\n",
    "\n",
    "real_splicing_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs)-intronic_sample_locs)\n",
    "real_intronic_sample_locs = list(real_sample_locs.intersection(intronic_sample_locs)-splicing_sample_locs)\n",
    "splicing_intronic_sample_locs = list(splicing_sample_locs.intersection(intronic_sample_locs)-real_sample_locs)\n",
    "\n",
    "real_splicing_intronic_sample_locs = list(real_sample_locs.intersection(splicing_sample_locs.intersection(intronic_sample_locs)))\n",
    "\n",
    "print(\"available real only loci: \"+str(len(real_only_sample_locs)))\n",
    "print(\"available splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "print(\"available intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "print(\"available real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "print(\"available real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "print(\"available splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "print(\"available real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "print(\"available intergenic only loci: \"+str(len(intergenic_sample_locs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n",
      "False\n",
      "False\n",
      "True\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "print(\"CHS.3\" in real_sample_locs)\n",
    "print(\"CHS.3\" in splicing_sample_locs)\n",
    "print(\"CHS.3\" in intronic_sample_locs)\n",
    "print(\"CHS.3\" in intergenic_sample_locs)\n",
    "print(\"CHS.3\" in real_only_sample_locs)\n",
    "print(\"CHS.3\" in splicing_only_sample_locs)\n",
    "print(\"CHS.3\" in intronic_only_sample_locs)\n",
    "print(\"CHS.3\" in real_splicing_sample_locs)\n",
    "print(\"CHS.3\" in real_intronic_sample_locs)\n",
    "print(\"CHS.3\" in splicing_intronic_sample_locs)\n",
    "print(\"CHS.3\" in real_splicing_intronic_sample_locs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13416\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "6603\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "1201\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "3445\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "357\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "578\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "240\n"
     ]
    }
   ],
   "source": [
    "for x in [real_only_sample_locs,\n",
    "          splicing_only_sample_locs,\n",
    "          intronic_only_sample_locs,\n",
    "          real_splicing_sample_locs,\n",
    "          real_intronic_sample_locs,\n",
    "          splicing_intronic_sample_locs,\n",
    "          real_splicing_intronic_sample_locs]:\n",
    "    for y in [real_only_sample_locs,\n",
    "              splicing_only_sample_locs,\n",
    "              intronic_only_sample_locs,\n",
    "              real_splicing_sample_locs,\n",
    "              real_intronic_sample_locs,\n",
    "              splicing_intronic_sample_locs,\n",
    "              real_splicing_intronic_sample_locs]:\n",
    "        print(len(set(x).intersection(set(y))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selecting real only loci: 8925\n",
      "selecting splicing only loci: 2731\n",
      "selecting intronic only loci: 219\n",
      "selecting real and splicing loci: 3445\n",
      "selecting real and intronic loci: 236\n",
      "selecting splicing and intronic loci: 204\n",
      "selecting real and splicing and intronic loci: 146\n",
      "selecting intergenic loci: 1171\n"
     ]
    }
   ],
   "source": [
    "random.shuffle(real_only_sample_locs)\n",
    "min_realn = min(len(real_only_sample_locs),realn)\n",
    "real_only_sample_locs = real_only_sample_locs[:min_realn]\n",
    "random.shuffle(splicing_only_sample_locs)\n",
    "min_splicen = min(len(splicing_only_sample_locs),splicen)\n",
    "splicing_only_sample_locs = splicing_only_sample_locs[:min_splicen]\n",
    "random.shuffle(intronic_only_sample_locs)\n",
    "min_intn = min(len(intronic_only_sample_locs),intn)\n",
    "intronic_only_sample_locs = intronic_only_sample_locs[:min_intn]\n",
    "random.shuffle(intergenic_sample_locs)\n",
    "min_poln = min(len(intergenic_sample_locs),poln)\n",
    "intergenic_sample_locs = intergenic_sample_locs[:min_poln]\n",
    "\n",
    "random.shuffle(real_splicing_sample_locs)\n",
    "min_real_splicen = min(len(real_splicing_sample_locs),real_splicen)\n",
    "real_splicing_sample_locs = real_splicing_sample_locs[:min_real_splicen]\n",
    "random.shuffle(real_intronic_sample_locs)\n",
    "min_real_intn = min(len(real_intronic_sample_locs),real_intn)\n",
    "real_intronic_sample_locs = real_intronic_sample_locs[:min_real_intn]\n",
    "random.shuffle(splicing_intronic_sample_locs)\n",
    "min_splice_intn = min(len(splicing_intronic_sample_locs),splice_intn)\n",
    "splicing_intronic_sample_locs = splicing_intronic_sample_locs[:min_splice_intn]\n",
    "\n",
    "random.shuffle(real_splicing_intronic_sample_locs)\n",
    "min_real_splice_intn = min(len(real_splicing_intronic_sample_locs),real_splice_intn)\n",
    "real_splicing_intronic_sample_locs = real_splicing_intronic_sample_locs[:min_real_splice_intn]\n",
    "\n",
    "print(\"selecting real only loci: \"+str(len(real_only_sample_locs)))\n",
    "print(\"selecting splicing only loci: \"+str(len(splicing_only_sample_locs)))\n",
    "print(\"selecting intronic only loci: \"+str(len(intronic_only_sample_locs)))\n",
    "print(\"selecting real and splicing loci: \"+str(len(real_splicing_sample_locs)))\n",
    "print(\"selecting real and intronic loci: \"+str(len(real_intronic_sample_locs)))\n",
    "print(\"selecting splicing and intronic loci: \"+str(len(splicing_intronic_sample_locs)))\n",
    "print(\"selecting real and splicing and intronic loci: \"+str(len(real_splicing_intronic_sample_locs)))\n",
    "print(\"selecting intergenic loci: \"+str(len(intergenic_sample_locs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final number of real transcripts: 25641\n",
      "final number of splicing transcripts: 10597\n",
      "final number of intronic transcripts: 1211\n",
      "final number of intergenic transcripts: 1185\n",
      "final number of real loci: 12752\n",
      "final number of splicing loci: 6526\n",
      "final number of intronic loci: 805\n",
      "final number of intergenic loci: 1171\n"
     ]
    }
   ],
   "source": [
    "real_sub = pd.concat([pd.DataFrame(lids_real,columns=[\"lid\"]),pd.DataFrame(tids_real,columns=[\"tid\"]),pd.DataFrame(tpms_real,columns=[\"tpm\"])],axis=1)\n",
    "splicing_sub = pd.concat([pd.DataFrame(lids_splicing,columns=[\"lid\"]),pd.DataFrame(tids_splicing,columns=[\"tid\"]),pd.DataFrame(tpms_splicing,columns=[\"tpm\"])],axis=1)\n",
    "intronic_sub = pd.concat([pd.DataFrame(lids_intronic,columns=[\"lid\"]),pd.DataFrame(tids_intronic,columns=[\"tid\"]),pd.DataFrame(tpms_intronic,columns=[\"tpm\"])],axis=1)\n",
    "intergenic_sub = pd.concat([pd.DataFrame(lids_intergenic,columns=[\"lid\"]),pd.DataFrame(tids_intergenic,columns=[\"tid\"]),pd.DataFrame(tpms_intergenic,columns=[\"tpm\"])],axis=1)\n",
    "\n",
    "real_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_only_sample_locs)].reset_index(drop=True)\n",
    "splicing_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_only_sample_locs)].reset_index(drop=True)\n",
    "intronic_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(intronic_only_sample_locs)].reset_index(drop=True)\n",
    "intergenic_sub2 = intergenic_sub[intergenic_sub[\"lid\"].isin(intergenic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "real_splicing_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "splicing_real_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "real_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "intronic_real_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "splicing_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "intronic_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "real_splicing_intronic_only_sample_df = real_sub[real_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "splicing_real_intronic_only_sample_df = splicing_sub[splicing_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "intronic_real_splicing_only_sample_df = intronic_sub[intronic_sub[\"lid\"].isin(real_splicing_intronic_sample_locs)].reset_index(drop=True)\n",
    "\n",
    "real_sub2 = pd.concat([real_only_sample_df,\\\n",
    "                       real_splicing_only_sample_df,\\\n",
    "                       real_intronic_only_sample_df,\\\n",
    "                       real_splicing_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "splicing_sub2 = pd.concat([splicing_only_sample_df,\\\n",
    "                           splicing_real_only_sample_df,\\\n",
    "                           splicing_intronic_only_sample_df,\\\n",
    "                           splicing_real_intronic_only_sample_df],axis=0).reset_index(drop=True)\n",
    "intronic_sub2 = pd.concat([intronic_only_sample_df,\\\n",
    "                           intronic_real_only_sample_df,\\\n",
    "                           intronic_splicing_only_sample_df,\\\n",
    "                           intronic_real_splicing_only_sample_df],axis=0).reset_index(drop=True)\n",
    "\n",
    "print(\"final number of real transcripts: \"+str(len(set(real_sub2[\"tid\"]))))\n",
    "print(\"final number of splicing transcripts: \"+str(len(set(splicing_sub2[\"tid\"]))))\n",
    "print(\"final number of intronic transcripts: \"+str(len(set(intronic_sub2[\"tid\"]))))\n",
    "print(\"final number of intergenic transcripts: \"+str(len(set(intergenic_sub2[\"tid\"]))))\n",
    "print(\"final number of real loci: \"+str(len(set(real_sub2[\"lid\"]))))\n",
    "print(\"final number of splicing loci: \"+str(len(set(splicing_sub2[\"lid\"]))))\n",
    "print(\"final number of intronic loci: \"+str(len(set(intronic_sub2[\"lid\"]))))\n",
    "print(\"final number of intergenic loci: \"+str(len(set(intergenic_sub2[\"lid\"]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lid</th>\n",
       "      <th>tid</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lid, tid, tpm]\n",
       "Index: []"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "real_sub2[real_sub2.duplicated(\"tid\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "ttx_real = list(real_sub2[\"tid\"])\n",
    "tpm_real = list(real_sub2[\"tpm\"])\n",
    "\n",
    "ttx_splice = list(splicing_sub2[\"tid\"])\n",
    "tpm_splice = list(splicing_sub2[\"tpm\"])\n",
    "\n",
    "ttx_int = list(intronic_sub2[\"tid\"])\n",
    "tpm_int = list(intronic_sub2[\"tpm\"])\n",
    "\n",
    "ttx_intergenic = list(intergenic_sub2[\"tid\"])\n",
    "tpm_intergenic = list(intergenic_sub2[\"tpm\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>tid_real</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29848</td>\n",
       "      <td>33409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...</td>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>5.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>29848</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...</td>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>5.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976</td>\n",
       "      <td>33409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...</td>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>5.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1048011</td>\n",
       "      <td>1049473</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...</td>\n",
       "      <td>CHS.48.17</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>1048011</td>\n",
       "      <td>1049059</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...</td>\n",
       "      <td>CHS.48.17</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source        type    start      end score strand phase  \\\n",
       "0  chr1  StringTie  transcript    29848    33409     .      +     .   \n",
       "1  chr1  StringTie        exon    29848    30667     .      +     .   \n",
       "2  chr1  StringTie        exon    30976    33409     .      +     .   \n",
       "3  chr1  StringTie  transcript  1048011  1049473     .      +     .   \n",
       "4  chr1  StringTie        exon  1048011  1049059     .      +     .   \n",
       "\n",
       "                                          attributes   tid_real   tpm  \n",
       "0  transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...    CHS.3.2  5.24  \n",
       "1  transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...    CHS.3.2  5.24  \n",
       "2  transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...    CHS.3.2  5.24  \n",
       "3  transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...  CHS.48.17  2.40  \n",
       "4  transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...  CHS.48.17  2.40  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_real = pd.concat([pd.DataFrame(ttx_real,columns=[\"tid_real\"]),pd.DataFrame(tpm_real,columns=[\"tpm\"])],axis=1)\n",
    "realDF_f = realDF[[\"lid\",\"tid_real\"]].merge(res_real,on=\"tid_real\",how=\"inner\")\n",
    "realDF_f = real_gtf.merge(realDF_f[[\"tid_real\",\"tpm\"]],on=\"tid_real\",how=\"inner\")\n",
    "realDF_f[\"attributes\"] = realDF_f[\"attributes\"].str.rstrip(\";\")+\\\n",
    "                            \";sim_tpm=\"+realDF_f[\"tpm\"].astype(str)\n",
    "realDF_f.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>tid_real</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29848</td>\n",
       "      <td>33409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...</td>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>5.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1048011</td>\n",
       "      <td>1049473</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...</td>\n",
       "      <td>CHS.48.17</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1059691</td>\n",
       "      <td>1066419</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.49.6\"; gene_id \"CHS.49\"; ol...</td>\n",
       "      <td>CHS.49.6</td>\n",
       "      <td>1.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11079</td>\n",
       "      <td>29924</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...</td>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>5.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>134846</td>\n",
       "      <td>143135</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.8.10\"; gene_id \"CHS.8\"; old...</td>\n",
       "      <td>CHS.8.10</td>\n",
       "      <td>2.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid     source        type    start      end score strand phase  \\\n",
       "0   chr1  StringTie  transcript    29848    33409     .      +     .   \n",
       "3   chr1  StringTie  transcript  1048011  1049473     .      +     .   \n",
       "6   chr1  StringTie  transcript  1059691  1066419     .      +     .   \n",
       "10  chr1  StringTie  transcript    11079    29924     .      -     .   \n",
       "22  chr1  StringTie  transcript   134846   143135     .      -     .   \n",
       "\n",
       "                                           attributes   tid_real   tpm  \n",
       "0   transcript_id \"CHS.3.2\"; gene_id \"CHS.3\"; old_...    CHS.3.2  5.24  \n",
       "3   transcript_id \"CHS.48.17\"; gene_id \"CHS.48\"; o...  CHS.48.17  2.40  \n",
       "6   transcript_id \"CHS.49.6\"; gene_id \"CHS.49\"; ol...   CHS.49.6  1.28  \n",
       "10  transcript_id \"CHS.2.1\"; gene_id \"CHS.2\"; old_...    CHS.2.1  5.73  \n",
       "22  transcript_id \"CHS.8.10\"; gene_id \"CHS.8\"; old...   CHS.8.10  2.02  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = realDF_f[realDF_f[\"type\"]==\"transcript\"]\n",
    "t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>tid_real</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqid, source, type, start, end, score, strand, phase, attributes, tid_real, tpm]\n",
       "Index: []"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t[t.duplicated(\"tid_real\")]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
