{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "base_dir_data = \"/ccb/salz8-1/avaraby/tx_noise/data/\"\n",
    "base_dir_out = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/GTEx_aggs/\"\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/sim_samples/\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "readlen = 101\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we need to implement a method for converting TPM information into\n",
    "# coverage information for polyester\n",
    "# we can do this by following the RSEM implementation of the TPM to number of reads conversion\n",
    "\n",
    "# first load the distribution of the number of reads per sample\n",
    "readlen_stats = pd.read_csv(\"/ccb/salz8-1/avaraby/tx_noise/data/readlen.stats\",usecols=[\"readlen\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEBCAYAAABojF4hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAANPUlEQVR4nO3df6zd9V3H8ecLyRAIne24COpsAXUxHT/EA4wfjiWgqGTqMOpMZKmaNKJEQSJDQwaE/QC2VcXosClb1cXpAhKmGDPSRJICS7glhbgtGdGBTgm92CYtQWocb/+4h+RSTntP23Pvad88H//0fH9wvu9L4Hm//dxzzk1VIUk6+h0z7QEkSZNh0CWpCYMuSU0YdElqwqBLUhMGXZKamHrQk5yaZFOSJ8c49/QkDyS5OcnfJPmZ5ZhRko4GUw86cCnwEJAxzr0J2FpVdwJ3AZ9eysEk6Wgy9aBX1f3AnoX7kqxN8pdJfi/JfUnOGB56EZgZPp4Bti3jqJJ0RDt22gPsxybgxqp6PMn7mL8T/wCwAXgwyQbgAuCO6Y0oSUeWIzXoZwM/keS9wPHAy8P9m4FNVfWFJDPAs0nOqKqdU5pTko4YR2rQnwb+rqqeSXIc83fnAO8EXhg+3gW8xhGwbCRJR4JM+8O5klwGfAj4SeAzzC+vrAFuBL7JfMQ/X1Vbk1wKXA88BZwObKuqe6cxtyQdaaYedEnSZLhcIUlNTG0N/eSTT641a9ZM6/KSdFTatm3bS1U1M+rY1IK+Zs0aZmdnp3V5SToqJXl+f8dccpGkJgy6JDUx1pJLklOBjwLnVNX5I46vA94D/CtwHvAnVfX4BOeUJC1i3DX01z9A69z9HP9e4PqqejXJhcy/df+sCcwnSRrTWEsuoz5Aa5/jH6uqVxc858ujzkuyPslsktm5ubmDHlaStH8TXUNPEuB3gN8ddbyqNlbVoKoGMzMjX3UjSTpEEwv6MOafBDZX1ROTel5J0ngOOehJThx+4iFJvgP4Y+Dvq+qfkvz8pAaUJI1n3Fe5XAZcA5yW5BbmP0BrHfM/+PwN5u/Mfw44e/5GnTOBB5ZgXumgDf+bXHJ+LpKmbWofzjUYDMp3iupIlcRA64iUZFtVDUYd841FktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqYtGgJzk1yaYkTx7gnCuSbE9y3WTHkySNa5w79EuBh4CMOphkBfBdwPYJziVJOkiLBr2q7gf2HOD47uE5kqQpWtY19CTrk8wmmZ2bm1vOS0tSe8sa9KraWFWDqhrMzMws56Ulqb1DCnqSE5NYZEk6gozzKpfLgGuA05LckuR4YB1wx4Jzfhs4G7gyyS8t0aySpANIVU3lwoPBoGZnZ6dybWkxSZjW/xvSgSTZVlWDUcd8Y5EkNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRw77QGkg7Vq1Sp27dq15NdJsqTPv3LlSnbu3Lmk19Bbi0HXUWfXrl1U1bTHOGxL/Q1Dbz0uuUhSEwZdkpoYa8klyanAR4Fzqur8EcePAT4OvAysBu6rqq9MclBJ0oGNu4Z+KfAQcO5+jv8isKKqbk6yCvhKkh+uqm9PYkhJ0uLGWnKpqvuBPQc45SrgieG5O4FXgbWHPZ0kaWyTWkM/hTcGf/dw3xskWZ9kNsns3NzchC4tSYLJBX0HcNKC7RXDfW9QVRuralBVg5mZmQldWpIEhxH0JCcmeb3KDwMXDfevAr4T+OrhjydJGtdYQU9yGXANcFqSW5IcD6wD7hie8kVgT5JbgU8CH/IHopK0vMZ6lUtVPQo8us/uP11w/DXgwxOcS5J0kHxjkSQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktTEseOclOQK4GpgB1BVdfs+x9cAtwNfBdYCG6rq6YlOKg3VrSvgtrdPe4zDVreumPYIambRoCc5AbgXWFtVe5M8kOTyqtqy4LQ/Av6iqh5MchbweeCcpRlZb3W5fTdVNe0xDlsS6rZpT6FOxllyuQh4vqr2DrcfA67a55wfBP59+PjfgLOTnLzvEyVZn2Q2yezc3NyhzixJGmGcoJ8C7FmwvXu4b6GtwHuGjy8Y/vmmv09W1caqGlTVYGZm5mBnlSQdwDhB3wGctGB7xXDfQjcC70hyA7Aa+G/gWxOZUJI0lnF+KPoEsDrJccNll0uAP0uyCvi/qtoNfA/wqap6Jcm7gC9X1f8u3diSpH0tGvRhpK8F7kkyBzxTVVuS3A3sBO4ELgZ+OskssAq4bimHliS9Wab1aoHBYFCzs7NTubaObkn6vMqlwdeh5ZVkW1UNRh3zjUWS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJamJY6c9gHQokkx7hMO2cuXKaY+gZgy6jjpVteTXSLIs15EmySUXSWpirDv0JFcAVwM7gKqq2/c5fjrwKeBJ4Fzgr6vqSxOeVZJ0AIsGPckJwL3A2qram+SBJJdX1ZYFp90EbK2qP0zyI8AXAYMuSctonCWXi4Dnq2rvcPsx4Kp9znkRmBk+ngG2jXqiJOuTzCaZnZubO5R5JUn7MU7QTwH2LNjePdy30AbgwiQbgI8Anxv1RFW1saoGVTWYmZkZdYok6RCNs4a+AzhpwfaK4b6FNgObquoLSWaAZ5OcUVU7JzOmJGkx49yhPwGsTnLccPsS4OEkq5KsGO57J/DC8PEu4LUxn1uSNCGL3qFX1StJrgXuSTIHPFNVW5LcDewE7gRuAK5PcjFwOvAHVfXSUg4uSXqjsV62WFWPAI/ss++mBY+3AlsnO5ok6WC4LCJJTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCaOHeekJFcAVwM7gKqq2/c5fh9w5oJdZwPnVdVzE5pTkrSIRYOe5ATgXmBtVe1N8kCSy6tqy4LTvlxVfzs8fwWw2ZhL0vIaZ8nlIuD5qto73H4MuGrhCa/HfOjXgc9OZjxJ0rjGCfopwJ4F27uH+94kyTHAlcDD+zm+Pslsktm5ubmDnVWSdADjBH0HcNKC7RXDfaP8LPAPVVWjDlbVxqoaVNVgZmbm4CaVJB3QOEF/Alid5Ljh9iXAw0lWDdfLF1oHbJ7ceJKkcS36Q9GqeiXJtcA9SeaAZ6pqS5K7gZ3AnQBJzgW+UVUvL+nEkqSRxnrZYlU9Ajyyz76b9tneDmyf3GiSpIPhG4skqQmDLklNGHRJamKsNXTpaJZkWf65/bxaV1o2Bl3tGVq9VbjkIklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpiUzrTRfDj+J9fioXlxZ3MvDStIeQRlhdVSN/Q9DUgi4dyZLMVtVg2nNIB8MlF0lqwqBLUhMGXRpt47QHkA6Wa+iS1IR36JLUhEGXpCYMujRCkquSfDPJmiSXJnkqyfumPZd0IAZdGqGqHmb4xreq2go8M92JpMX5K+jURpJfBT4BfAb4PuBHgceBHcDbge1V9VdJzgA2DI+dBXy6qrYnOQ64D3gR+C9gxX6u82PArwFfA94F3AysBD4HPAe8AFwMfHz4jUFaFr7KRa0k+Wfg7qr6xyQfBs6sqvWZ/43PXwfeC7wNOKWqnkpyHvD7VfULSX4LeHdVXZvkGObj/N6qei7JZmAz8CjwLeCCqvrPJOuAc6rqhuHjK6vql5OcD3ykqt6/jF++3uK8Q1dHXx/++f3AaUluHm7/C3Aq83fgH0zyU8zfhb/+uRhrgWcBquq1JM+NeO6TgVXANfPfI1gFfHvB8W8M/5wDTprEFyONy6Cro9f/2vk0sLeq7gRI8gHm77pvB3ZV1ceS/BBw4fD8rwHvHp57DLBmxHO/xHys/7yqdiV5B3DRiGtLy86gq40kPw6sBq5Lchfz6+F3J7mN+WWW/6mqB5M8AHxiuGb+NmB1ksuBTcBnk9wD7AReBn4zyZeAs4FrgCeBDwJ3JfkP5v8WsCHJdwPvB1Ym+QHgV15/3qraslz/DvTW5hq6JDXhyxYlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJv4f0jQZ/vP/vaAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "for label in readlen_stats.columns:\n",
    "    # now we shall remove any outliers from the data\n",
    "    q25,q50,q75 = readlen_stats[label].quantile([0.25,0.5,0.75])\n",
    "    iqr = q75-q25\n",
    "    thw = q75+0.5*iqr\n",
    "    tlw = q25-0.5*iqr\n",
    "    ahw = readlen_stats[readlen_stats[label]<=thw][label].max()\n",
    "    alw = readlen_stats[readlen_stats[label]>=tlw][label].min()\n",
    "    readlen_stats = readlen_stats[(readlen_stats[label]<=ahw)&(readlen_stats[label]>=alw)].reset_index(drop=True)\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "ax.boxplot(readlen_stats.T,labels=readlen_stats.T.index.tolist())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 73823384.99041033\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 85105738.84843396\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 116548929.3445094\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 89638996.16990443\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 82413511.93373978\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 82644467.15347973\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 111124938.56628963\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 97192405.64986117\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 93652790.65603974\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 62826802.96979342\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 98234773.85894907\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 87373858.3069808\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 95882169.57037285\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 83849317.81291062\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 88425942.6090655\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 107268352.88645823\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 104160323.92976527\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 87385025.13205875\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 88790789.90902606\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 85146716.66188717\n",
      "\n",
      "=================\n",
      "Tissue #2\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 100376215.75724527\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 103930613.56949122\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 90557531.37434255\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 89821325.37974657\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 95776232.38691962\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 90334049.38907082\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 97937402.75873277\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 95453863.53458613\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 92372051.18228665\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "number of reads in sample is: 90065768.04709198\n"
     ]
    }
   ],
   "source": [
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        total_nreads = np.random.normal(readlen_stats[\"readlen\"].mean(),readlen_stats[\"readlen\"].std())\n",
    "        print(\"number of reads in sample is: \"+str(total_nreads))\n",
    "\n",
    "        real = pd.read_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        real[\"tid\"] = real[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        real[\"gid\"] = real[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        realt = real[real[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        reale = real[real[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        reale[\"elen\"] = reale[\"end\"]-reale[\"start\"]\n",
    "        reale = reale[[\"tid\",\"elen\"]]\n",
    "        reale = reale.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(realt[\"tid\"])==set(reale[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        reale = realt.merge(reale,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(reale),\"number of tpms different from the number of transcripts\"\n",
    "        reale[\"tpm\"] = tpms[\"tpm\"]\n",
    "\n",
    "        splicing = pd.read_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        splicing[\"tid\"] = splicing[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        splicing[\"gid\"] = splicing[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        splicingt = splicing[splicing[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        splicinge = splicing[splicing[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        splicinge[\"elen\"] = splicinge[\"end\"]-splicinge[\"start\"]\n",
    "        splicinge = splicinge[[\"tid\",\"elen\"]]\n",
    "        splicinge = splicinge.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(splicingt[\"tid\"])==set(splicinge[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        splicinge = splicingt.merge(splicinge,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(splicinge),\"number of tpms different from the number of transcripts\"\n",
    "        splicinge[\"tpm\"] = tpms[\"tpm\"]\n",
    "\n",
    "        intronic = pd.read_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        intronic[\"tid\"] = intronic[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        intronic[\"gid\"] = intronic[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        intronict = intronic[intronic[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        intronice = intronic[intronic[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        intronice[\"elen\"] = intronice[\"end\"]-intronice[\"start\"]\n",
    "        intronice = intronice[[\"tid\",\"elen\"]]\n",
    "        intronice = intronice.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(intronict[\"tid\"])==set(intronice[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        intronice = intronict.merge(intronice,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(intronice),\"number of tpms different from the number of transcripts\"\n",
    "        intronice[\"tpm\"] = tpms[\"tpm\"]\n",
    "\n",
    "        pol = pd.read_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        pol[\"tid\"] = pol[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        pol[\"gid\"] = pol[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        polt = pol[pol[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        pole = pol[pol[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        pole[\"elen\"] = pole[\"end\"]-pole[\"start\"]\n",
    "        pole = pole[[\"tid\",\"elen\"]]\n",
    "        pole = pole.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(polt[\"tid\"])==set(pole[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        pole = polt.merge(pole,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(pole),\"number of tpms different from the number of transcripts\"\n",
    "        pole[\"tpm\"] = tpms[\"tpm\"]\n",
    "        \n",
    "        joined = pd.concat([reale[[\"tid\",\"elen\",\"tpm\"]],splicinge[[\"tid\",\"elen\",\"tpm\"]],intronice[[\"tid\",\"elen\",\"tpm\"]],pole[[\"tid\",\"elen\",\"tpm\"]]],axis=0).reset_index(drop=True)\n",
    "        joined[\"theta\"] = joined[\"elen\"]*joined[\"tpm\"]\n",
    "        denom = joined[\"theta\"].sum()\n",
    "        joined[\"cor\"] = joined[\"theta\"]/denom\n",
    "        # now that we have all these values, we \n",
    "        joined[\"cov\"] = (joined[\"cor\"]*total_nreads*readlen)/joined[\"elen\"]\n",
    "        # now we can merge the data to comply with the original ordering\n",
    "        # and proceed to write it out\n",
    "        realt.merge(joined[[\"tid\",\"cov\"]],how=\"left\",on=\"tid\")[[\"cov\"]].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",index=False,header=False)\n",
    "        splicingt.merge(joined[[\"tid\",\"cov\"]],how=\"left\",on=\"tid\")[[\"cov\"]].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",index=False,header=False)\n",
    "        intronict.merge(joined[[\"tid\",\"cov\"]],how=\"left\",on=\"tid\")[[\"cov\"]].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",index=False,header=False)\n",
    "        polt.merge(joined[[\"tid\",\"cov\"]],how=\"left\",on=\"tid\")[[\"cov\"]].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2\n",
      "=================\n",
      "\n",
      "++++++\n",
      ">Sample #0\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #1\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #2\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #3\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #4\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #5\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #6\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #7\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #8\n",
      "++++++\n",
      "\n",
      "++++++\n",
      ">Sample #9\n",
      "++++++\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# now we can also compute TPMs for each sample\n",
    "for tissue_num in range(num_tissues):\n",
    "    print(\"\\n=================\\nTissue #\"+str(tissue_num)+\"\\n=================\\n\")\n",
    "    for sample_num in range(num_samples):\n",
    "        print(\"++++++\\n>Sample #\"+str(sample_num)+\"\\n++++++\\n\")\n",
    "        real = pd.read_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        real[\"tid\"] = real[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        real[\"gid\"] = real[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        realt = real[real[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        reale = real[real[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        reale[\"elen\"] = reale[\"end\"]-reale[\"start\"]\n",
    "        reale = reale[[\"tid\",\"elen\"]]\n",
    "        reale = reale.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(realt[\"tid\"])==set(reale[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        reale = realt.merge(reale,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(reale),\"number of tpms different from the number of transcripts\"\n",
    "        reale[\"tpm\"] = tpms[\"tpm\"]\n",
    "        covs = pd.read_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",names=[\"cov\"])\n",
    "        assert len(covs)==len(reale),\"number of covs different from the number of transcripts\"\n",
    "        reale[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "        splicing = pd.read_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        splicing[\"tid\"] = splicing[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        splicing[\"gid\"] = splicing[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        splicingt = splicing[splicing[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        splicinge = splicing[splicing[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        splicinge[\"elen\"] = splicinge[\"end\"]-splicinge[\"start\"]\n",
    "        splicinge = splicinge[[\"tid\",\"elen\"]]\n",
    "        splicinge = splicinge.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(splicingt[\"tid\"])==set(splicinge[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        splicinge = splicingt.merge(splicinge,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(splicinge),\"number of tpms different from the number of transcripts\"\n",
    "        splicinge[\"tpm\"] = tpms[\"tpm\"]\n",
    "        covs = pd.read_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",names=[\"cov\"])\n",
    "        assert len(covs)==len(splicinge),\"number of covs different from the number of transcripts\"\n",
    "        splicinge[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "        intronic = pd.read_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        intronic[\"tid\"] = intronic[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        intronic[\"gid\"] = intronic[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        intronict = intronic[intronic[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        intronice = intronic[intronic[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        intronice[\"elen\"] = intronice[\"end\"]-intronice[\"start\"]\n",
    "        intronice = intronice[[\"tid\",\"elen\"]]\n",
    "        intronice = intronice.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(intronict[\"tid\"])==set(intronice[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        intronice = intronict.merge(intronice,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(intronice),\"number of tpms different from the number of transcripts\"\n",
    "        intronice[\"tpm\"] = tpms[\"tpm\"]\n",
    "        covs = pd.read_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",names=[\"cov\"])\n",
    "        assert len(covs)==len(intronice),\"number of covs different from the number of transcripts\"\n",
    "        intronice[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "        pol = pd.read_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".gtf\",sep=\"\\t\",names=gff3cols)\n",
    "        pol[\"tid\"] = pol[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        pol[\"gid\"] = pol[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "        polt = pol[pol[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "        pole = pol[pol[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "        pole[\"elen\"] = pole[\"end\"]-pole[\"start\"]\n",
    "        pole = pole[[\"tid\",\"elen\"]]\n",
    "        pole = pole.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "        assert set(polt[\"tid\"])==set(pole[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "        pole = polt.merge(pole,how=\"left\",on=\"tid\")\n",
    "        tpms = pd.read_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".exp\",names=[\"tpm\"])\n",
    "        assert len(tpms)==len(pole),\"number of tpms different from the number of transcripts\"\n",
    "        pole[\"tpm\"] = tpms[\"tpm\"]\n",
    "        covs = pd.read_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".cov\",names=[\"cov\"])\n",
    "        assert len(covs)==len(pole),\"number of covs different from the number of transcripts\"\n",
    "        pole[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "        joined = pd.concat([reale[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                            splicinge[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                            intronice[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                            pole[[\"tid\",\"elen\",\"tpm\",\"cov\"]]],axis=0).reset_index(drop=True)\n",
    "        joined[\"new_tpm\"]=(joined[\"cov\"]*1000000)/joined[\"cov\"].sum()\n",
    "        realt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        splicingt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        intronict.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        polt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tid  elen    tpm         cov\n",
       "0    CHS.3.2  3252   0.25    1.655020\n",
       "1  CHS.39.13  2540   8.12   53.755053\n",
       "2  CHS.39.22   461   2.40   15.888193\n",
       "3   CHS.46.9   872  53.02  350.996662\n",
       "4   CHS.12.1  3788   7.50   49.650603"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's now take one sample\n",
    "# convert the coverages back to TPM\n",
    "# scatterplot original TPM vs the obtained TPM\n",
    "# there might be some differences, but should follow the same pattern largely\n",
    "real = pd.read_csv(out_dir+\"real.t0_s0.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "real[\"tid\"] = real[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "real[\"gid\"] = real[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "realt = real[real[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "reale = real[real[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "reale[\"elen\"] = reale[\"end\"]-reale[\"start\"]\n",
    "reale = reale[[\"tid\",\"elen\"]]\n",
    "reale = reale.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "assert set(realt[\"tid\"])==set(reale[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "reale = realt.merge(reale,how=\"left\",on=\"tid\")\n",
    "tpms = pd.read_csv(out_dir+\"real.t0_s0.exp\",names=[\"tpm\"])\n",
    "assert len(tpms)==len(reale),\"number of tpms different from the number of transcripts\"\n",
    "reale[\"tpm\"] = tpms[\"tpm\"]\n",
    "covs = pd.read_csv(out_dir+\"real.t0_s0.cov\",names=[\"cov\"])\n",
    "assert len(covs)==len(reale),\"number of covs different from the number of transcripts\"\n",
    "reale[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "splicing = pd.read_csv(out_dir+\"splicing.t0_s0.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "splicing[\"tid\"] = splicing[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "splicing[\"gid\"] = splicing[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "splicingt = splicing[splicing[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "splicinge = splicing[splicing[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "splicinge[\"elen\"] = splicinge[\"end\"]-splicinge[\"start\"]\n",
    "splicinge = splicinge[[\"tid\",\"elen\"]]\n",
    "splicinge = splicinge.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "assert set(splicingt[\"tid\"])==set(splicinge[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "splicinge = splicingt.merge(splicinge,how=\"left\",on=\"tid\")\n",
    "tpms = pd.read_csv(out_dir+\"splicing.t0_s0.exp\",names=[\"tpm\"])\n",
    "assert len(tpms)==len(splicinge),\"number of tpms different from the number of transcripts\"\n",
    "splicinge[\"tpm\"] = tpms[\"tpm\"]\n",
    "covs = pd.read_csv(out_dir+\"splicing.t0_s0.cov\",names=[\"cov\"])\n",
    "assert len(covs)==len(splicinge),\"number of covs different from the number of transcripts\"\n",
    "splicinge[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "intronic = pd.read_csv(out_dir+\"intronic.t0_s0.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "intronic[\"tid\"] = intronic[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "intronic[\"gid\"] = intronic[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "intronict = intronic[intronic[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "intronice = intronic[intronic[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "intronice[\"elen\"] = intronice[\"end\"]-intronice[\"start\"]\n",
    "intronice = intronice[[\"tid\",\"elen\"]]\n",
    "intronice = intronice.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "assert set(intronict[\"tid\"])==set(intronice[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "intronice = intronict.merge(intronice,how=\"left\",on=\"tid\")\n",
    "tpms = pd.read_csv(out_dir+\"intronic.t0_s0.exp\",names=[\"tpm\"])\n",
    "assert len(tpms)==len(intronice),\"number of tpms different from the number of transcripts\"\n",
    "intronice[\"tpm\"] = tpms[\"tpm\"]\n",
    "covs = pd.read_csv(out_dir+\"intronic.t0_s0.cov\",names=[\"cov\"])\n",
    "assert len(covs)==len(intronice),\"number of covs different from the number of transcripts\"\n",
    "intronice[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "pol = pd.read_csv(out_dir+\"intergenic.t0_s0.gtf\",sep=\"\\t\",names=gff3cols)\n",
    "pol[\"tid\"] = pol[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "pol[\"gid\"] = pol[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "polt = pol[pol[\"type\"]==\"transcript\"][[\"tid\",\"gid\"]].reset_index(drop=True) # intended for order\n",
    "pole = pol[pol[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "pole[\"elen\"] = pole[\"end\"]-pole[\"start\"]\n",
    "pole = pole[[\"tid\",\"elen\"]]\n",
    "pole = pole.groupby(\"tid\").agg({\"elen\":\"sum\"}).reset_index()\n",
    "assert set(polt[\"tid\"])==set(pole[\"tid\"]),\"number of transcripts is not the same as number of groupped exons\"\n",
    "pole = polt.merge(pole,how=\"left\",on=\"tid\")\n",
    "tpms = pd.read_csv(out_dir+\"intergenic.t0_s0.exp\",names=[\"tpm\"])\n",
    "assert len(tpms)==len(pole),\"number of tpms different from the number of transcripts\"\n",
    "pole[\"tpm\"] = tpms[\"tpm\"]\n",
    "covs = pd.read_csv(out_dir+\"intergenic.t0_s0.cov\",names=[\"cov\"])\n",
    "assert len(covs)==len(pole),\"number of covs different from the number of transcripts\"\n",
    "pole[\"cov\"] = covs[\"cov\"]\n",
    "\n",
    "joined = pd.concat([reale[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                    splicinge[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                    intronice[[\"tid\",\"elen\",\"tpm\",\"cov\"]],\\\n",
    "                    pole[[\"tid\",\"elen\",\"tpm\",\"cov\"]]],axis=0).reset_index(drop=True)\n",
    "joined[\"new_tpm\"]=(joined[\"cov\"]*1000000)/joined[\"cov\"].sum()\n",
    "realt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"real.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        splicingt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"splicing.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        intronict.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"intronic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)\n",
    "        polt.merge(joined[[\"tid\",\"new_tpm\"]],how=\"left\",on=\"tid\")[[\"new_tpm\"]].to_csv(out_dir+\"intergenic.t\"+str(tissue_num)+\"_s\"+str(sample_num)+\".tpm\",index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "      <th>nr</th>\n",
       "      <th>rpk</th>\n",
       "      <th>new_tpm</th>\n",
       "      <th>ratio</th>\n",
       "      <th>strg_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "      <td>53.288370</td>\n",
       "      <td>8.193169</td>\n",
       "      <td>0.598069</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>0.598069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "      <td>1351.859742</td>\n",
       "      <td>266.114122</td>\n",
       "      <td>19.425272</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>19.425272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "      <td>72.519376</td>\n",
       "      <td>78.654420</td>\n",
       "      <td>5.741460</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>5.741460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "      <td>3030.387023</td>\n",
       "      <td>1737.607238</td>\n",
       "      <td>126.838415</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>126.838415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "      <td>1862.143404</td>\n",
       "      <td>245.795064</td>\n",
       "      <td>17.942062</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>17.942062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tid  elen    tpm         cov           nr          rpk     new_tpm  \\\n",
       "0    CHS.3.2  3252   0.25    1.655020    53.288370     8.193169    0.598069   \n",
       "1  CHS.39.13  2540   8.12   53.755053  1351.859742   266.114122   19.425272   \n",
       "2  CHS.39.22   461   2.40   15.888193    72.519376    78.654420    5.741460   \n",
       "3   CHS.46.9   872  53.02  350.996662  3030.387023  1737.607238  126.838415   \n",
       "4   CHS.12.1  3788   7.50   49.650603  1862.143404   245.795064   17.942062   \n",
       "\n",
       "      ratio    strg_tpm  \n",
       "0  2.392275    0.598069  \n",
       "1  2.392275   19.425272  \n",
       "2  2.392275    5.741460  \n",
       "3  2.392275  126.838415  \n",
       "4  2.392275   17.942062  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined[\"nr\"] = (joined[\"cov\"]*joined[\"elen\"])/readlen\n",
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "      <th>nr</th>\n",
       "      <th>rpk</th>\n",
       "      <th>new_tpm</th>\n",
       "      <th>ratio</th>\n",
       "      <th>strg_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "      <td>26.644185</td>\n",
       "      <td>8.193169</td>\n",
       "      <td>0.598069</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>0.598069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "      <td>675.929871</td>\n",
       "      <td>266.114122</td>\n",
       "      <td>19.425272</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>19.425272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "      <td>36.259688</td>\n",
       "      <td>78.654420</td>\n",
       "      <td>5.741460</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>5.741460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "      <td>1515.193512</td>\n",
       "      <td>1737.607238</td>\n",
       "      <td>126.838415</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>126.838415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "      <td>931.071702</td>\n",
       "      <td>245.795064</td>\n",
       "      <td>17.942062</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>17.942062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tid  elen    tpm         cov           nr          rpk     new_tpm  \\\n",
       "0    CHS.3.2  3252   0.25    1.655020    26.644185     8.193169    0.598069   \n",
       "1  CHS.39.13  2540   8.12   53.755053   675.929871   266.114122   19.425272   \n",
       "2  CHS.39.22   461   2.40   15.888193    36.259688    78.654420    5.741460   \n",
       "3   CHS.46.9   872  53.02  350.996662  1515.193512  1737.607238  126.838415   \n",
       "4   CHS.12.1  3788   7.50   49.650603   931.071702   245.795064   17.942062   \n",
       "\n",
       "      ratio    strg_tpm  \n",
       "0  2.392275    0.598069  \n",
       "1  2.392275   19.425272  \n",
       "2  2.392275    5.741460  \n",
       "3  2.392275  126.838415  \n",
       "4  2.392275   17.942062  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined[\"rpk\"] = joined[\"nr\"]/(joined[\"elen\"]/1000)\n",
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13.699376741338735"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pmf = joined[\"rpk\"].sum()/1000000\n",
    "pmf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "      <th>nr</th>\n",
       "      <th>rpk</th>\n",
       "      <th>new_tpm</th>\n",
       "      <th>ratio</th>\n",
       "      <th>strg_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "      <td>26.644185</td>\n",
       "      <td>8.193169</td>\n",
       "      <td>0.598069</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>0.598069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "      <td>675.929871</td>\n",
       "      <td>266.114122</td>\n",
       "      <td>19.425272</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>19.425272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "      <td>36.259688</td>\n",
       "      <td>78.654420</td>\n",
       "      <td>5.741460</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>5.741460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "      <td>1515.193512</td>\n",
       "      <td>1737.607238</td>\n",
       "      <td>126.838415</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>126.838415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "      <td>931.071702</td>\n",
       "      <td>245.795064</td>\n",
       "      <td>17.942062</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>17.942062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tid  elen    tpm         cov           nr          rpk     new_tpm  \\\n",
       "0    CHS.3.2  3252   0.25    1.655020    26.644185     8.193169    0.598069   \n",
       "1  CHS.39.13  2540   8.12   53.755053   675.929871   266.114122   19.425272   \n",
       "2  CHS.39.22   461   2.40   15.888193    36.259688    78.654420    5.741460   \n",
       "3   CHS.46.9   872  53.02  350.996662  1515.193512  1737.607238  126.838415   \n",
       "4   CHS.12.1  3788   7.50   49.650603   931.071702   245.795064   17.942062   \n",
       "\n",
       "      ratio    strg_tpm  \n",
       "0  2.392275    0.598069  \n",
       "1  2.392275   19.425272  \n",
       "2  2.392275    5.741460  \n",
       "3  2.392275  126.838415  \n",
       "4  2.392275   17.942062  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined[\"new_tpm\"]=joined[\"rpk\"]/pmf\n",
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "      <th>nr</th>\n",
       "      <th>rpk</th>\n",
       "      <th>new_tpm</th>\n",
       "      <th>ratio</th>\n",
       "      <th>strg_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "      <td>53.288370</td>\n",
       "      <td>16.386338</td>\n",
       "      <td>0.598069</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>0.598069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "      <td>1351.859742</td>\n",
       "      <td>532.228245</td>\n",
       "      <td>19.425272</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>19.425272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "      <td>72.519376</td>\n",
       "      <td>157.308841</td>\n",
       "      <td>5.741460</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>5.741460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "      <td>3030.387023</td>\n",
       "      <td>3475.214476</td>\n",
       "      <td>126.838415</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>126.838415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "      <td>1862.143404</td>\n",
       "      <td>491.590128</td>\n",
       "      <td>17.942062</td>\n",
       "      <td>2.392275</td>\n",
       "      <td>17.942062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tid  elen    tpm         cov           nr          rpk     new_tpm  \\\n",
       "0    CHS.3.2  3252   0.25    1.655020    53.288370    16.386338    0.598069   \n",
       "1  CHS.39.13  2540   8.12   53.755053  1351.859742   532.228245   19.425272   \n",
       "2  CHS.39.22   461   2.40   15.888193    72.519376   157.308841    5.741460   \n",
       "3   CHS.46.9   872  53.02  350.996662  3030.387023  3475.214476  126.838415   \n",
       "4   CHS.12.1  3788   7.50   49.650603  1862.143404   491.590128   17.942062   \n",
       "\n",
       "      ratio    strg_tpm  \n",
       "0  2.392275    0.598069  \n",
       "1  2.392275   19.425272  \n",
       "2  2.392275    5.741460  \n",
       "3  2.392275  126.838415  \n",
       "4  2.392275   17.942062  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined[\"strg_tpm\"]=(joined[\"cov\"]*1000000)/joined[\"cov\"].sum()\n",
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{nan, nan, 2.392274904155898, 2.3922749041558973, 2.3922749041558986, 2.3922749041558977, 2.392274904155899, 2.3922749041558995, 2.392274904155897, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 2.3922749041559, 2.3922749041558964}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>elen</th>\n",
       "      <th>tpm</th>\n",
       "      <th>cov</th>\n",
       "      <th>nr</th>\n",
       "      <th>rpk</th>\n",
       "      <th>new_tpm</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.3.2</td>\n",
       "      <td>3252</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.655020</td>\n",
       "      <td>53.288370</td>\n",
       "      <td>16.386338</td>\n",
       "      <td>0.598069</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.39.13</td>\n",
       "      <td>2540</td>\n",
       "      <td>8.12</td>\n",
       "      <td>53.755053</td>\n",
       "      <td>1351.859742</td>\n",
       "      <td>532.228245</td>\n",
       "      <td>19.425272</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.39.22</td>\n",
       "      <td>461</td>\n",
       "      <td>2.40</td>\n",
       "      <td>15.888193</td>\n",
       "      <td>72.519376</td>\n",
       "      <td>157.308841</td>\n",
       "      <td>5.741460</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.46.9</td>\n",
       "      <td>872</td>\n",
       "      <td>53.02</td>\n",
       "      <td>350.996662</td>\n",
       "      <td>3030.387023</td>\n",
       "      <td>3475.214476</td>\n",
       "      <td>126.838415</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.12.1</td>\n",
       "      <td>3788</td>\n",
       "      <td>7.50</td>\n",
       "      <td>49.650603</td>\n",
       "      <td>1862.143404</td>\n",
       "      <td>491.590128</td>\n",
       "      <td>17.942062</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41563</th>\n",
       "      <td>ALL_30460985</td>\n",
       "      <td>229</td>\n",
       "      <td>1.11</td>\n",
       "      <td>7.348289</td>\n",
       "      <td>16.660973</td>\n",
       "      <td>72.755339</td>\n",
       "      <td>2.655425</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41564</th>\n",
       "      <td>ALL_30466440</td>\n",
       "      <td>254</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2.184627</td>\n",
       "      <td>5.494011</td>\n",
       "      <td>21.629966</td>\n",
       "      <td>0.789451</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41565</th>\n",
       "      <td>ALL_30404134</td>\n",
       "      <td>160</td>\n",
       "      <td>0.41</td>\n",
       "      <td>2.714233</td>\n",
       "      <td>4.299775</td>\n",
       "      <td>26.873594</td>\n",
       "      <td>0.980833</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41566</th>\n",
       "      <td>ALL_30440285</td>\n",
       "      <td>251</td>\n",
       "      <td>0.95</td>\n",
       "      <td>6.289076</td>\n",
       "      <td>15.629289</td>\n",
       "      <td>62.268083</td>\n",
       "      <td>2.272661</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41567</th>\n",
       "      <td>ALL_30462839</td>\n",
       "      <td>315</td>\n",
       "      <td>0.54</td>\n",
       "      <td>3.574843</td>\n",
       "      <td>11.149264</td>\n",
       "      <td>35.394489</td>\n",
       "      <td>1.291828</td>\n",
       "      <td>2.392275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41568 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                tid  elen    tpm         cov           nr          rpk  \\\n",
       "0           CHS.3.2  3252   0.25    1.655020    53.288370    16.386338   \n",
       "1         CHS.39.13  2540   8.12   53.755053  1351.859742   532.228245   \n",
       "2         CHS.39.22   461   2.40   15.888193    72.519376   157.308841   \n",
       "3          CHS.46.9   872  53.02  350.996662  3030.387023  3475.214476   \n",
       "4          CHS.12.1  3788   7.50   49.650603  1862.143404   491.590128   \n",
       "...             ...   ...    ...         ...          ...          ...   \n",
       "41563  ALL_30460985   229   1.11    7.348289    16.660973    72.755339   \n",
       "41564  ALL_30466440   254   0.33    2.184627     5.494011    21.629966   \n",
       "41565  ALL_30404134   160   0.41    2.714233     4.299775    26.873594   \n",
       "41566  ALL_30440285   251   0.95    6.289076    15.629289    62.268083   \n",
       "41567  ALL_30462839   315   0.54    3.574843    11.149264    35.394489   \n",
       "\n",
       "          new_tpm     ratio  \n",
       "0        0.598069  2.392275  \n",
       "1       19.425272  2.392275  \n",
       "2        5.741460  2.392275  \n",
       "3      126.838415  2.392275  \n",
       "4       17.942062  2.392275  \n",
       "...           ...       ...  \n",
       "41563    2.655425  2.392275  \n",
       "41564    0.789451  2.392275  \n",
       "41565    0.980833  2.392275  \n",
       "41566    2.272661  2.392275  \n",
       "41567    1.291828  2.392275  \n",
       "\n",
       "[41568 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined[\"ratio\"] = joined[\"new_tpm\"]/joined[\"tpm\"]\n",
    "print(set(joined[\"ratio\"]))\n",
    "joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD9CAYAAABHnDf0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAATWElEQVR4nO3df6zddX3H8eebUuqFtHYLLb9iW7MZTBpoNbeZTVkgAmJsRrBZZhZk/61JE//YVBClSpvUMIPMhSWMdDPRBLZs/PBixgKpnetiU43FSptl4jZmMTXxVhdotd2Vlff+uOfivZdz2nPO9/s9P77f5yPhj++nPef7Pj2XVz99fz/fzzcyE0lSvVw07AIkSeUz3CWphgx3Saohw12Sashwl6QaMtwlqYYuHnYBAJdffnmuW7du2GVI0lh54YUXfpaZq9r92kiE+7p16zh8+PCwy5CksRIRxzv9mm0ZSaohw12Sashwl6QaMtwlqYYMd0mqoZFYLSNJTTN15AQPPv8SP3n1LFevnODu267ljvdcU9r7lx7uEbEO+Evgp8A/Z+bfln0OSRpnU0dO8Omnj3H29XMAnHj1LJ9++hhAaQHfVVsmIq6MiL+JiO8uGr8lIh6JiF0Rcf+8X/p34DBwrJQqJalGHnz+pTeDfc7Z18/x4PMvlXaObmfuNwDPABvnBiLiUuBRYH1mzkTEUxFxM/CvwC7gLPCPwNbSqpWkGvjJq2d7Gu9HVzP3zHwSOL1oeDNwPDNnWscHmQ3ydwEX5ewjnjr+5RER2yPicEQcPnnyZO+VS9KYunrlRE/j/SiyWmY1CwP/VGvsauBzEfEx4OlOL87MvZk5mZmTq1a13RpBkmrp7tuuZWLpkgVjE0uXcPdt15Z2jiIXVKeB5fOOVwDTmfkN4BuFqpKkGpu7aDqqq2UOAWsjYlmrNbMFeKScsiSp3u54zzWlhvli3a6WuRG4C7gqInZGxERmngF2AA9HxB7gaGbur6xSSVLXupq5Z+YB4ECb8X3AvrKLkiQV4x2qktSjqu8uLYPhLkk9GMTdpWVw4zBJ6sEg7i4tg+EuST0YxN2lZTDcJakHg7i7tAyGuyT1YBB3l5bBC6qS1INB3F1aBsNdknpU9d2lZbAtI0k1ZLhLUg0Z7pJUQ4a7JNWQF1Ql1d447AVTNsNdUq2Ny14wZbMtI6nWxmUvmLI5c5dUO/PbMNnh94zaXjBlM9wl1criNkwno7YXTNkMd0ljb/5M/aIIzmWn+fqsUdwLpmyGu6Sxtnimfr5gD3C1jCSNg3YXTNu5ZuUEB+99/wAqGg2ulpE01rq5MNqENsxihruksdbpwuiSCILZGfsD266rfRtmMdsyksba3bdd+5bVMRNLlzQy0Ocz3CWNtXF5eMagVRbuEfFN4LOZ+a2qziFJMB4Pzxi0SnruEfEB4JdVvLck6cK6mrlHxJXAHmBDZm6aN34LsA2YBjIzd0dEAJPA4QrqlSR1odu2zA3AM8DGuYGIuBR4FFifmTMR8VRE3AysBKaAPyi7WElSd7oK98x8MiJuWjS8GTiemTOt44PAVuAEsIrZ2ftlEfFSZp5c/J4RsR3YDrBmzZr+qpc0tpq4x/ogFem5rwZOzzs+BazOzIeA54A3gHPAa+1enJl7M3MyMydXrVpVoAxJ42Zuy4ATrV0b5/ZYnzpyYtil1UaRcJ8Gls87XtEaIzN/lJm3Z+a9mfmrIgVKqpepIyf4xD+82Mg91gepSLgfAtZGxLLW8Rbg2eIlSaqruRl7p8296r7H+iB1Fe4RcSNwF3BVROyMiInMPAPsAB6OiD3A0czcX2GtksbchTb5qvse64PU7QXVA8CBNuP7gH1lFyWpns43M2/i5l5VcvsBSaXrtBLm6pUTnGgT8EsiGr8XTNkMd0mlmDpygvu+doxf/mph22VuJQy4ydcgueWvpMKmjpzgE0+8+JZgnzO3EuaO91zDA9uu45qVE43ejncQnLlLKmRuaeOFnls61293k6/BMNwl9W3n1DEe//YrnD/WZ7kSZrAMd0k9mTpygl1f/zdePft6169xJczgGe6SujZ15AR3P/Eir7/RzVx91mWXLOHzH7avPmiGu6SuPfj8S10HewB3vm8Ne+64rtqi1JbhLqmjnVPH+Lvv/JhzmSyJuOBFU5gN9S99ZKMz9SEz3CW9xdSRE/zJ339/wVi3wX7n+9YY7CPAcJe0wM6pYzz27Vd6ft3KiaXsun29wT4iDHdJb7r1z/+F/5i+8OOPV04sfXO1zG9cupT7f89QHzWGu6SuQx1m94H5/v0fqLgiFWW4Sw3Wrrd+IX/4O++oqBqVyXCXGqqf3voVyy9xaeOYMNylBpnbirfdtrsXcsXyS/jOfbdWUJWqYLhLDdHvKhiAj3oz0tgx3KWa62cvmDlbfus3efyPN1dQlapmuEs11cuOjYvZghl/hrtUQ70sbZzPTb7qw3CXaqTfvvo1855zqnow3KWauPOvD3Hwv/6n59etWLaEg/e+v4KKNEyGuzTmiqyCWbFsCUd3f7DkijQKDHdpTPVzd+kclzbWXyXhHhEbgE3ACmBlZn6uivNITTT7NKTv8/ob/b3eYG+GrsM9Iq4E9gAbMnPTvPFbgG3ANJCZuTszX4yI08Anga+VXLPUWO++75/433P9LG6Ed62+jH0fv6ncgjSyepm53wA8A2ycG4iIS4FHgfWZORMRT0XEzZm5PzNfjoh7gK8C+0qtWmqYIi0YQ72Zug73zHwyIm5aNLwZOJ6ZM63jg8DWiLg4M5/PzF9ExPJ27xcR24HtAGvWrOm9cqkhfvvTz/J/fUzWJ5ZexAPbrnd5Y0MV7bmvBk7POz7VGlsVEZ8B3gC+0u6FmbkX2AswOTnZ378zpRrrd2kj2FdX8XCfBubPzFcA05n5WMH3lRqrSKgD/IUPpxbFw/0QsDYilrVaM1uAR4qXJTVTvy0YsLeuhXpZLXMjcBdwVUTsBB7KzDMRsQN4OCJOAkczc39FtUq1df39z3Fq5lxfr3WmrnZ6uaB6ADjQZnwfroaR+tLvBl9zDHZ14h2q0pCsu/fZvl9rqOtCDHdpwN5577N97bEOPjxD3TPcpQEp0lcP4EvO1tUDw10agCItGGfr6ofhLlWoyAXTty0JfvD5D5VckZrCcJcqUGQvGPAOUxVnuEslK9KC8UYklcVwl0pSZDtecHmjymW4SyUoMlu/YvklfOe+W0usRjLcpUKKhHoA//1nW8srRprHcJf6UCTUweWNqp7hLvWg6CoYWzAaFMNd6lKR7XgBfmQLRgNkuEsXUDTUXbOuYTDcpfMo0lu/OOA/H3C2ruEw3KU2il4wtQWjYTPcpXmKbMcLroLR6DDcpRZn66oTw12NZ6irjgx3NVbRVTCuWdcoM9zVSM7WVXeGuxrFUFdTGO5qhKKh7lORNG4Md9Wes3U1keGu2nLnRjVZJeEeEbcD7waWAj/MzCeqOI/Uzs6pYzz27VcKvYezdY27rsM9Iq4E9gAbMnPTvPFbgG3ANJCZuRt4ITO/HhFvB74MGO4aCFsw0qxeZu43AM8AG+cGIuJS4FFgfWbORMRTEXFzZu5v/ZYPA18srVqpg6KhvmLZEo7u/mBJ1UjD13W4Z+aTEXHTouHNwPHMnGkdHwS2AvsjYivwMnCi3ftFxHZgO8CaNWt6LFuadf39z3Fq5lyh93C2rjoq2nNfDZyed3wKWB0RdwCfAl4ElgN3Ln5hZu4F9gJMTk4W2atJDWULRuqsaLhPMxvec1YA05k5BUwVfG+praKhDga76q9ouB8C1kbEslZrZgvwSPGypPacrUvd6WW1zI3AXcBVEbETeCgzz0TEDuDhiDgJHJ13MVUqjaEu9aaXC6oHgANtxvcB+8osSppjC0bqj3eoamQ5W5f6Z7hr5BjqUnGGu0aGLRipPIa7RoKzdalchruGqmiof/R9a9hzx3UlVSPVh+GuobAFI1XLcNfA2YKRqme4a2BswUiDY7ircrZgpMEz3FUpWzDScBjuqkTRUH/bkuAHn/9QSdVIzWO4q1TvvPdZim7O72xdKs5wV2lswUijw3BXYV4wlUaP4a5CnK1Lo8lwV18MdWm0Ge7qiS0YaTwY7uqas3VpfBjuuiBDXRo/hrs6sgUjjS/DXW05W5fGm+GuBdw2QKoHw12ALRipbgx32YKRaqiScI+Ii4FPAWszc3sV51Bxztal+qpq5n4Z8Bywo6L3VwGGulR/XYd7RFwJ7AE2ZOameeO3ANuAaSAzc3dmvhYRPy+9WhVmC0Zqhl5m7jcAzwAb5wYi4lLgUWB9Zs5ExFMRcXNm7i+5ThXkbF1qlq7DPTOfjIibFg1vBo5n5kzr+CCwFbhguEfEdmA7wJo1a7otQ31wti41T9Ge+2rg9LzjU8DqiAjgI8C1EfHezPze4hdm5l5gL8Dk5GTRh/eoDUNdaq6i4T4NLJ93vAKYzswEvtD6TwNmC0ZS0XA/BKyNiGWt1swW4JHiZalfztYlQW+rZW4E7gKuioidwEOZeSYidgAPR8RJ4KgXU4fDUJc0Xy8XVA8AB9qM7wP2lVmUumcLRlI7bj8wxpytS+rEcB9DhrqkCzHcx0QZ7Rcw2KWmMNxHVFlhPsdQl5rFcB8RZYf5HENdaibDfYiqCvQ5BrvUXIb7AFUd5nMMdUmG+wAMKtTBYJc0y3CvyCADHQx1SQsZ7iVyhi5pVBjuJRhEqBvmknphuBfgahdJo8pw70OVoW6gSyqD4d6DKkLdMJdUBcO9C24FIGncGO4XUGawG+qSBsVwPw8fhCFpXBnuHbhnuqRxZriXyECXNCoM9zZ6nbUb6pJGjeG+SC/BbqhLGlUXDbuAcWWwSxplhvs83c7aDXZJo85wbzHYJdWJ4c7g916XpKoZ7j1w1i5pXDQ+3G3HSKqjRoe77RhJddXocO+Ws3ZJ46ax4W47RlKdNTLcbcdIqrtGhnu3nLVLGleNC3fbMZKaoPSNwyLiUmAX8Arw08x8ouxz9Mtgl9QUXYV7RFwJ7AE2ZOameeO3ANuAaSAzc3fr+LuZ+URETAEjE+6S1BTdztxvAJ4BNs4NtGbojwLrM3MmIp6KiJuBdwCHWr9tosxii3DWLqlJuuq5Z+aTwOlFw5uB45k50zo+CGwFfgysao2d7fSeEbE9Ig5HxOGTJ0/2VnWPXB0jqWmKXFBdzcLAP9UaexrYFBEfAx7v9OLM3JuZk5k5uWrVqk6/baCctUuqiyIXVKeB5fOOVwDTmXkGuKdQVSWyHSOpiYrM3A8BayNiWet4CzBS/Q+DXVJTdRXuEXEjcBdwVUTsjIiJ1gx9B/BwROwBjmbm/gpr7Yl9dklN1lVbJjMPAAfajO8D9pVd1CA5a5dUR7W8Q9V2jKSmq2W4S1LT1S7cnbVLUs3C3WCXpFm1CXeDXZJ+rTbhLkn6tVqEu7N2SVpo7MPdm5Uk6a3GPty75axdUpOMdbjbjpGk9sY23A12Seqs9GeoQsfH70mSBqSqB2S/5fF7w9gx0lm7pKaqoi3T6fF7C1T9mD2DXVKTVRHunR6/t8AoPmZPkuqiinBv+/i9Cs7TkbN2SU1XRbgP5PF7nQLcYJekCi6oZuaZiJh7/N5JKnz8nkEuSe1VshSyDo/fk6RxNrY3MUmSOjPcJamGDHdJqiHDXZJqKDJz2DXQWlVzvMBbXA78rKRyxunco6rufyZ+vvFVt8+2NjPb3gU6EuFeVEQczszJpp17VNX9z8TPN77q/NkWsy0jSTVkuEtSDdUl3Pc29Nyjqu5/Jn6+8VXnz7ZALXrukqSF6jJzlyTNY7hLUg0Z7pJUQ5XsCtlUrefH7gJeAX6amU8Mt6LREBEXA59i9oaL7cOup2wRcTvwbmAp8MM6fe8RsQHYxOxDd1Zm5ueGXFIlIuKbwGcz81vDrqUsYx3uEXELsI3ZJz1lZu6u4BxXAnuADZm56QLn3gZ8NzOfiIgpoDb/kxd0GfAcsGPYhXSrx+/9hcz8ekS8HfgyI/699/LZMvPFiDgNfBL42lAKrlhEfAD45bDrKNvYhntrlvwosD4zZyLiqYi4uYIHg9wAPANsvNC5gXcw+yQqgImS6xgpPQbEaxHx8yGV2q+uv/d5P3MfBr44+FJ71tNny8yXI+Ie4KuMwXMaevnZjIgAJoHDQym2QuPcc98MHM/MmdbxQaD0RzNl5pMsfOD3+c79Y2Bun4ezZdcyYuYCIuYG5gXEn2bmLuD61l96Y6fH752I2Aq8DJwYWJF96uWzRcRtrdf8goXPRh5lvfxsbgOmhlFk1cZ25g6sZuEP6KnW2DDP/TSwKyKuAB4fUC1DkZlPRsRNi4Y7hV8lj1kcgrbfe0Tcwew1hReZDcA7h1BbUZ1+pldFxGeAN4CvDKGunvX4s3mC2QnZJHBZRLyUmScHVmyFxjncp1k4k1jRGhvauTPzDHDPgGoYRZ3CL4CPANdGxHsz83tDqa64Tt/7FOM/++v02R4bUj1la/uzmZkfj4h1wIeAc8Brgy+tGuPcljkErI2IZa3jLcCzDTj3KOsUEJmZX8jM3x3jYId6f+91/mxwnslgZv4oM2/PzHsz81dDqa4CYztzz8wzEbEDeLi1H/zRCi6mEhE3AncBV0XETuChQZ17DL0ZEK1//m4BHhlyTX2p8/de5892HrX52eyWe8uoL62A+CPgg8BfMRsQZyPiVuD3gZPA61UsT5XOx5/NWYa7JNXQOPfcJUkdGO6SVEOGuyTVkOEuSTVkuEtSDRnuklRDhrsk1ZDhLkk1ZLhLUg0Z7pJUQ/8PA4y5E+3xaVoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(joined[\"tpm\"],joined[\"new_tpm\"])\n",
    "plt.yscale(\"symlog\")\n",
    "plt.xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD3CAYAAADmBxSSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAN4klEQVR4nO3db4xc11nH8e8TnLrrKiaJspYjBXuhVKWJig1ZF1m2aioHUmTaEoMKAswLkCwMFGgIkRVM1ZSluCKRUCQgGFUqon0Vp8QhpkmtqFrAMpBNRcPfBkhipAq6m4Jqp05N2zy8mOvu7GrGnvXd2T/Pfj/SSHPOzD1z7tmZn+6e+y8yE0lSLdcsdwckSYvPcJekggx3SSrIcJekggx3SSpo3XJ3AOCmm27KsbGx5e6GJK0qzz777MuZOdrrtRUR7mNjY0xNTS13NyRpVYmIs/1ec1pGkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgpaEWeotjF2+OQ3n790dN8y9kSSVg633CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpoKGFe0R8JiJ2D6t9SVJ/Qwn3iPhB4CvDaFuSdGXrBnlTRGwGJoBtmbmjq/4OYD8wDWRm3h8RAYwDU0PoryRpAAOFO7AbOAFsv1QRERuAh4HbMvNiRDwaEXuB64HHgPdersGIOAgcBNiyZctVdF2S1M9A0zKZeRw4P696J3A2My825dPAPmAMeDudrff3RMRonzaPZeZ4Zo6PjvZ8iyTpKrWZc9/E3MA/B2zKzAeBJ4HXgG8AX27xGZKkqzDotEwv08B1XeWNTR2Z+RLw7hZtS5JaaLPlfgbYGhHrm/Iu4GT7LkmS2hoo3CNiD3AAuDkijkTESGZeAA4BD0XEBPBcZj49xL5KkgY00LRMZk4Ckz3qTwGnFrtTkqR2vPyAJBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBW0bhiNRsQ2YAewEbg+Mz8wjM+RJPU2cLhHxGZgAtiWmTu66u8A9gPTQGbm/Zn5uYg4D9wD/Nki91mSdAULmZbZDZwA4lJFRGwAHgben5kfBL47IvYCZOYLwL3Az/dqLCIORsRUREzNzMxcZfclSb0MHO6ZeRw4P696J3A2My825dPAvoi4s1nmFeC6Pu0dy8zxzBwfHR1deM8lSX21nXPfxNzAP9fUjUbEfcBrwMdafoYkaYHahvs0c7fMNwLTmfnxlu1KklpoeyjkGWBrRKxvyruAky3blCS1NHC4R8Qe4ABwc0QciYiRzLwAHAIeiogJ4LnMfHpIfZUkDWjgaZnMnAQme9SfAk4tZqckSe14hqokFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFWS4S1JBhrskFbRuGI1GxLuB7wKuBZ7PzEeG8TmSpN4GDveI2AxMANsyc0dX/R3AfmAayMy8H3g2Mx+PiG8FPgoY7pK0hBay5b4bOAFsv1QRERuAh4HbMvNiRDwaEXsz8+nmLXcBDyxabyVJAxl4zj0zjwPn51XvBM5m5sWmfBrYBxAR+4AXgC/0ai8iDkbEVERMzczMLLjjkqT+2u5Q3cTcwD8HbIqIHwGOAD8JHO21YGYey8zxzBwfHR1t2Q1JUre2O1Sngeu6yhuB6cx8DHisZduSpKvUdsv9DLA1ItY35V3AyZZtSpJaGjjcI2IPcAC4OSKORMRIZl4ADgEPRcQE8FzXzlRJ0jIZeFomMyeByR71p4BTi9kpSVI7nqEqSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJU0FBus7dcxg7PvWbZS0f3LVNPJGl5ueUuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUkOEuSQUZ7pJUUKkLh83XfSExLyImaS1xy12SCjLcJakgw12SCjLcJamgoYR7RKyLiN+IiGPDaF+SdHnD2nJ/A/DkENuXJF3GwIdCRsRmYALYlpk7uurvAPYD00Bm5v2Z+eWI+NKi91aSNJCFHOe+GzgBbL9UEREbgIeB2zLzYkQ8GhF7M/PpRe6nJGkBBp42yczjwPl51TuBs5l5sSmfBgY6WygiDkbEVERMzczMDNoNSdIA2s6Jb2Ju4J8DNkVEAD8OvDkivrfXgpl5LDPHM3N8dHS0ZTckSd3aXn5gGriuq7wRmM7MBD7SPCRJS6ztlvsZYGtErG/Ku4CTl3m/JGkJDBzuEbEHOADcHBFHImIkMy8Ah4CHImICeM6dqZK0/AaelsnMSWCyR/0p4NRidkqS1I4nGUlSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQW0vHLZqjB3ufcmbl44OdIViSVpV3HKXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpIIMd0kqyHCXpILWLXaDEbEB+CDwn8AXM/ORxf4MSdLlDRTuEbEZmAC2ZeaOrvo7gP3ANJCZeX9TfiYzH4mIxwDDXZKW2KBb7ruBE8D2SxXNFvrDwG2ZeTEiHo2IvcC3AWeat40sZmclSYMZKNwz83hEfP+86p3A2cy82JRPA/uAzwKjTd2r/dqMiIPAQYAtW7YsoMvDM3b4ZM/6l47uG+pn9Wt/sd6jhXNcNUxL8f1qs0N1E3C+q3yuqfsksCMifgn4RL+FM/NYZo5n5vjo6Gi/t0mSrkKbHarTwHVd5Y3AdGZeAO5t1StJUittttzPAFsjYn1T3gX0nteQJC2pgcI9IvYAB4CbI+JIRIw0W+iHgIciYgJ4LjOfHmJfJUkDGnSH6iQw2aP+FHBqsTslSWrHM1QlqSDDXZIKMtwlqSDDXZIKMtwlqSDDXZIKMtwlqSDDXZIKMtwlqSDDXZIKMtwlqSDDXZIKMtwlqaA2N+voq8+NsyVJS2TRw73fjbO91rskLZ3IzMVtMGIvcF9m7m3KdwO3ZObd8973zRtkA28GPn+VH3kT8PJVLluNYzHLsehwHGZVHIutmdnzJtTDmJbpd+PsOTLzGHCs7YdFxFRmjrdtpwLHYpZj0eE4zFprYzGMHao9b5w9hM+RJPUxjHD3xtmStMwWfVomMy9ExKUbZ88w/Btnt57aKcSxmOVYdDgOs9bUWCz6DlVJ0vLzJCZJKshwl6SChnKG6lKpeCZsRGwGJoBtmbmjqXs98ADwBeBNwNHMfL557aeB7wG+AfxHZv5RUz8G/Cbw78AY8GuZ+UpEXAN8GHgF2Ap8NDP/ZqnWbyEi4o10xuKzwC3AlzLzQxFxI3AUeIHOeNyXmV9slvl1Okdo3QB8OjMfb+q3A78IvEjn0Nx7MvPrlxvblaL5m/058LfA64A3Aj8LjLCGxuGSiBihMxafzsx71urv44oyc1U+gA10/jDrm/KjwN7l7tcirNePAe8CprrqDgP3Ns/fCvxV8/wW4O+Z3XfyDPCm5vmTwNua5+8Dfqt5/hPAHzTPbwSeB75lude7z1jsAN7TVf5n4HY6Z0C/t6l7F/CnzfPvA/6ieX4t8G/A9UAA/whsbl57EPi5y43tSnrQ+Q/7SFf5BPBTa20cutb/QeBPgAcu1/fqv48rPVbztMxO4GxmXmzKp4F9y9ifRZGZx5l7Ehh01utM8/o/ANsiYiNwJ/BsNt/E5j0/FBHXAu+g82WGuWPT3db/AF8FbhvO2rSTmc9k5omuqmuAr9C1Dsxdtx9mdt2+BvwL8HbgO4CRzPzvHsv0G9sVIzNfy8wJgIhYRye0Ps8aGweAiDhAp98vdlWvyd/HlazmcB/oTNgi+q1rv/qbgFe7vtTdY7Mqxy0i7gKeysx/Ze46nANuaEJvoePEFV5bUSLiTuAJ4InMnGKNjUNE3Aq8JTM/Oe+lNf/76GU1h/taOhO237r2q38ZGImImFd/ubZWrIh4B50trfc3Vd3rsBH438z8Ogsfp/ltzX9tRcnMpzLzncC3R8QvsPbG4S7gqxFxGNgNvC0ifpU1/vvoZzWH+1o6E/YknWkoIuKtwOcy8xzwFHB715d0J/Cp5l/xz9CZs4a5Y9Pd1o3A64F/WoqVuBoRsY/Ov9e/AmyOiJ10rQNz1+0JZtdtHXAr8Jd0dji+2uysnr9Mv7FdMSLi1mYcLnmRzhTLmhqHzPztzPxQZh4F/hr4u8z8Pdbw7+NyVvVJTBHxA3R2QM4AX8saR8vsAX4GeCfwh3R2HkHnaID/Ar4T+HDOPRpgnM7RAM/n3KMBPkDnB70FuDtnjwb4HeBCU//HuUKPBoiI24FJYKqpegPw+8DjwEeAs3SOHDmcc48SuaF5fCrnHiXyvmaZG5k9SmSEPmO7UjRHDf0unaOGrgXeAvwy8H+soXG4JCJ+lM4RP6+j8314jDX4+7iSVR3ukqTeVvO0jCSpD8Ndkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpoP8HxI66Rk0+cf8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(joined[\"tpm\"],bins=100)\n",
    "plt.yscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD3CAYAAADmBxSSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAOpklEQVR4nO3dfYxc11nH8e+TOLjrEpOkWWOLYq8oVaChdSDrIitWTXGgRQstCahQwAipkkUEEW1IIhNMlJSluFIiIUtAMKoEavkDxYE4wjTFiqoFLFOyqWh4aQmQxogqdDcNjZ06Xfry8MdcJ7Orme2sZ2d25+H7kVaac3bunefsnv35+tx7ZyIzkSTVcslaFyBJWn2GuyQVZLhLUkGGuyQVZLhLUkEb1roAgKuvvjonJibWugxJGilPPPHEc5k53ul76yLcJyYmmJ2dXesyJGmkRMSZbt9zWUaSCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCloXd6j2Y+LgiZcfP3N4ag0rkaT1wyN3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSpoYOEeER+PiD2D2r8kqbuBhHtE/DDwpUHsW5L0jW3o5UkRsRWYBnZm5q62/huBm4E5IDPz3ogIYBKYHUC9kqQe9BTuwB7gOHDdhY6I2AQ8AFybmQsR8VBE7AOuAB4G3rXcDiPiAHAAYPv27RdRuiSpm56WZTLzGHBuSfdu4ExmLjTtU8AUMAG8hdbR+zsjYrzLPo9m5mRmTo6Pd3yKJOki9bPmvoXFgX8W2JKZ9wOPAl8Hvga80MdrSJIuQq/LMp3MAZe3tTc3fWTmM8A7+ti3JKkP/Ry5nwZ2RMTGpn0DcKL/kiRJ/eop3CNiL7Af2BYRhyJiLDPPA7cARyJiGngyMx8bYK2SpB71tCyTmTPATIf+k8DJ1S5KktQf335Akgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgoy3CWpIMNdkgraMIidRsROYBewGbgiM+8exOtIkjrrOdwjYiswDezMzF1t/TcCNwNzQGbmvZn5qYg4B9wO/Pkq1yxJ+gZWsiyzBzgOxIWOiNgEPAC8LzPvAd4UEfsAMvNp4E7gFzvtLCIORMRsRMzOz89fZPmSpE56DvfMPAacW9K9GziTmQtN+xQwFRFva7Z5Ebi8y/6OZuZkZk6Oj4+vvHJJUlf9rrlvYXHgn236xiPiLuDrwB/1+RqSpBXqN9znWHxkvhmYy8yP9LlfSVIf+r0U8jSwIyI2Nu0bgBN97lOS1Keewz0i9gL7gW0RcSgixjLzPHALcCQipoEnM/OxAdUqSepRz8symTkDzHToPwmcXM2iJEn98Q5VSSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSrIcJekggx3SSpowyB2GhHvAL4LuAx4KjMfHMTrSJI66zncI2IrMA3szMxdbf03AjcDc0Bm5r3AE5n5SER8C/AhwHCXpCFayZH7HuA4cN2FjojYBDwAXJuZCxHxUETsy8zHmqfcBNy3atVKknrS85p7Zh4Dzi3p3g2cycyFpn0KmAKIiCngaeBznfYXEQciYjYiZufn51dcuCSpu35PqG5hceCfBbZExI8Dh4CfAQ532jAzj2bmZGZOjo+P91mGJKldvydU54DL29qbgbnMfBh4uM99S5IuUr9H7qeBHRGxsWnfAJzoc5+SpD71HO4RsRfYD2yLiEMRMZaZ54FbgCMRMQ082XYyVZK0RnpelsnMGWCmQ/9J4ORqFiVJ6o93qEpSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBU0kI/ZWysTBxe/Z9kzh6fWqBJJWlseuUtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVkuEtSQYa7JBVU6o3Dlmp/IzHfREzS/yceuUtSQYa7JBVkuEtSQYa7JBU0kHCPiA0R8esRcXQQ+5ckLW9QR+6vBh4d4P4lScvo+VLIiNgKTAM7M3NXW/+NwM3AHJCZeW9mvhARX1j1aiVJPVnJde57gOPAdRc6ImIT8ABwbWYuRMRDEbEvMx9b5TolSSvQ87JJZh4Dzi3p3g2cycyFpn0K6OluoYg4EBGzETE7Pz/faxmSpB70uya+hcWBfxbYEhEB/BRwTUR8X6cNM/NoZk5m5uT4+HifZUiS2vX79gNzwOVt7c3AXGYm8MHmS5I0ZP0euZ8GdkTExqZ9A3BimedLkoag53CPiL3AfmBbRByKiLHMPA/cAhyJiGngSU+mStLa63lZJjNngJkO/SeBk6tZlCSpP95kJEkFGe6SVJDhLkkFGe6SVJDhLkkFGe6SVJDhLkkFGe6SVJDhLkkF9fvGYSNj4mDnt7x55nBP71AsSSPFI3dJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKshwl6SCDHdJKmjDau8wIjYB9wD/CXw+Mx9c7deQJC2vp3CPiK3ANLAzM3e19d8I3AzMAZmZ9zbtxzPzwYh4GDDcJWnIej1y3wMcB6670NEcoT8AXJuZCxHxUETsA74dON08bWw1i5Uk9aancM/MYxHxA0u6dwNnMnOhaZ8CpoBPAuNN30vd9hkRB4ADANu3b19ByYMzcfBEx/5nDk8N9LW67X+1nqOV8+eqQRrG/OrnhOoW4Fxb+2zT92fAroj4ZeBPum2cmUczczIzJ8fHx7s9TZJ0Efo5oToHXN7W3gzMZeZ54M6+qpIk9aWfI/fTwI6I2Ni0bwA6r2tIkoaqp3CPiL3AfmBbRByKiLHmCP0W4EhETANPZuZjA6xVktSjXk+ozgAzHfpPAidXuyhJUn+8Q1WSCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCjLcJakgw12SCurnwzq66vLB2ZKkIVn1cO/2wdm+17skDU9k5uruMGIfcFdm7mvatwGvzczbljzv5Q/IBq4B/vUiX/Jq4LmL3HYUOL7RVnl8lccGozG+HZnZ8UOoB7Es0+2DsxfJzKPA0X5fLCJmM3Oy3/2sV45vtFUeX+WxweiPbxAnVDt+cPYAXkeS1MUgwt0PzpakNbbqyzKZeT4iLnxw9jyD/+Dsvpd21jnHN9oqj6/y2GDEx7fqJ1QlSWvPm5gkqSDDXZIKGsgdqsMyKnfCRsTrgGngk8BrgS9k5vsj4irgMPA08Hpa9wd8vtnmDlpXGl0J/FVmPtL0Xwf8EvBZWpeY3p6ZX42IVwH3AZ9r9nU4M58a4hjHgE80td6+XD0R8XPA9wJfA/4jM/+g6Z8AfgP4d2AC+NXMfDEiLgE+ALwI7AA+lJl/N8SxXQO8G3gJ2AvcQ2vOrajWbvN1uXkwpPHd0Yzhueb13wOMdatpvc/NiNhK6+9tZ2buavqGMh/XVSZl5kh+AZto/dA3Nu2HgH1rXVeXWncB72xr/wtwPa07ed/V9P0Y8OHm8fcDf9k8vgz4N+AKIIB/ArY237sfeE/z+CBwZ/P4jcDfDHmM9wN/DNy3XD20/nH7B1453/M48Prm8aPAm5vHtwK/2Tz+aeD3msdXAU8Blw5pXJfSutrrkqa9DRhfaa3Lzddu82BI49sKPN82vuPAz47y3AR+sql5tq1v4PNxud/xWnyN8rLMbuBMZi407VPA1BrW01VmPp6Zx9u6LgG+RKve001fe/0/eqE/M78CfBp4C/AdwFhm/neHbabatvlHYGdEbB7IgJaIiP1NLZ9t6+5Wz9uAJ7KZ/c1zfiQiLgPeSuuPC7qP7Xngy8C1AxvQYrtoBdetEfFrtELjixdR63Lztds8GIbzwP/SOhIH+Gbgn5epad3Pzcw8xuIbKZerYTXn47rKpFFelunpTtj1JiJuAj6WmZ+JiPYxnAWujIgNtMbx6bbNLoxtnu5j7vbzOLvqg2gTEW8Avjsz74qIN7V9q1s93fqvBl5q+yPrZWzDsIPWH+27M/OFiPgI8JqLqHW8S//SbV6eB5n51dUezFKZebZZZvnTiHgW+C9aR58jPzeXGMZ8XO53PHSjfOQ+cnfCRsRbaR0NvK/pah/DZuB/mj/obmNbbsxr9fO4CfhyRBwE9gBvjoj3LlNPt/7ngLGIiCX9LLPNMJwFPpOZLzTtvwW+h5XX2uvvrn0eDFyzTn4HMJWZv0Dr93D3MjWN0txsN4z5uB7G+bJRDveRuhM2IqZo/RfwV4CtEbGbVr27m6e01/8XF/qbo6U3AH9N6+TWS80Jo6XbnGjb5o3ApzJz4EdGmflbmfn+zDxMK/j+PjN/Z5l6PgZc3/ZHsxv4aPNf/I/TWgZZbmxXAa+itXQwDJ8AXhMRlzbtHc1rr7TW5eZrt3kwDN8GPN/2j8mztGoe+bm5xDDm47rKpJG+iSkifojWyZN54Cu5fq+WuR6YAWabrlcDvws8AnwQOAO8DjiYi69IuLL5+mguviLh1mabq3jlioQxWlcDPAt8J/CBHO7VMj9B60qJb2rG9nC3epqrEyZpXZ3wVC6+OuFuWkGxHbgtX7k64bdprQ9vB/4wh3u1zE3AD9KaZ9tp/fy/daW1dpuvTUB0nAdDGNulwBFa68ZfpPW/kvcCC91qWu9zMyL2Aj8PvB34fVond+lWw2rOx/WUSSMd7pKkzkZ5WUaS1IXhLkkFGe6SVJDhLkkFGe6SVJDhLkkFGe6SVND/AUbHB9kJP0AxAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(joined[\"new_tpm\"],bins=100)\n",
    "plt.yscale(\"symlog\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
