{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocessing\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import random\n",
    "from scipy import stats\n",
    "import glob\n",
    "import math\n",
    "import csv\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations\n",
    "out_dir = \"/ccb/salz8-1/avaraby/tx_noise/full_analysis_t3_s10_26022020/sim_samples/\"\n",
    "idx_dir = \"/ccb/salz8-1/avaraby/tx_noise/indices/\"\n",
    "annotation_path = \"/ccb/salz3/avaraby/chess/versions/curRelease/chess2.2_assembly.gtf\"\n",
    "hg38_fa = \"/home/avaraby1/genomes/human/hg38/hg38_p8.fa\"\n",
    "\n",
    "num_tissues = 3\n",
    "num_samples = 10\n",
    "\n",
    "readlen = 101\n",
    "\n",
    "hisat_path = \"/ccb/salz8-1/avaraby/agar_manuscript/soft/AGAR2/external/hisat2/hisat2\"\n",
    "kallisto_path = \"/home/avaraby1/soft/kallisto/kallisto\"\n",
    "salmon_path = \"/home/avaraby1/soft/salmon-latest_linux_x86_64/bin/salmon\"\n",
    "sim2sam_path = \"/ccb/salz8-1/avaraby/tx_noise/sim2sam/sim2sam\"\n",
    "stringtie2_path = \"stringtie\"\n",
    "\n",
    "num_threads = 40\n",
    "num_processes = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hisat2_align(ts):\n",
    "    tn = ts[0]\n",
    "    sn = ts[1]\n",
    "    print(\"\\n=================\\nTissue #\"+str(tn)+\" - Sample #\"+str(sn)+\"\\n=================\\n\")\n",
    "    if tn==0 and (sn==4 or sn==6):\n",
    "        if not os.path.exists(out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)):\n",
    "            os.makedirs(out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn))\n",
    "        hisat_cmd = [hisat_path,\n",
    "                     \"-x\",idx_dir+\"annotation.hisat\",\n",
    "                     \"-p\",str(num_threads),\n",
    "                     \"--rna-sensitive\",\"-f\",\n",
    "                     \"-S\",out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\",\n",
    "                     \"-U\",out_dir+\"real.t\"+str(tn)+\"_s\"+str(sn)+\"/sample_01.shuffled.fasta\"]\n",
    "        subprocess.call(hisat_cmd)\n",
    "        sort_cmd = [\"samtools\",\"sort\",\n",
    "                    \"-@\",str(num_threads),\n",
    "                    \"-o\",out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "                    out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\"]\n",
    "        subprocess.call(sort_cmd)\n",
    "        strg_cmd = [\"stringtie\",\n",
    "                    out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "                    \"-p\",str(num_threads),\n",
    "                    \"-G\",annotation_path,\n",
    "                    \"-A\",out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.genes\",\n",
    "                    \"-o\",out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.gtf\"]\n",
    "        subprocess.call(strg_cmd)\n",
    "\n",
    "    #     if not os.path.exists(out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)):\n",
    "    #         os.makedirs(out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn))\n",
    "    #     hisat_cmd = [hisat_path,\n",
    "    #                  \"-x\",idx_dir+\"annotation.hisat\",\n",
    "    #                  \"-p\",str(num_threads),\n",
    "    #                  \"--rna-sensitive\",\"-f\",\n",
    "    #                  \"-S\",out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\",\n",
    "    #                  \"-U\",out_dir+\"real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/sample_01.shuffled.fasta\"]\n",
    "    #     subprocess.call(hisat_cmd)\n",
    "    #     sort_cmd = [\"samtools\",\"sort\",\n",
    "    #                 \"-@\",str(num_threads),\n",
    "    #                 \"-o\",out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\"]\n",
    "    #     subprocess.call(sort_cmd)\n",
    "    #     strg_cmd = [\"stringtie\",\n",
    "    #                 out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 \"-p\",str(num_threads),\n",
    "    #                 \"-G\",annotation_path,\n",
    "    #                 \"-A\",out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.genes\",\n",
    "    #                 \"-o\",out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.gtf\"]\n",
    "    #     subprocess.call(strg_cmd)\n",
    "\n",
    "    #     if not os.path.exists(out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)):\n",
    "    #         os.makedirs(out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn))\n",
    "    #     hisat_cmd = [hisat_path,\n",
    "    #                  \"-x\",idx_dir+\"annotation.hisat\",\n",
    "    #                  \"-p\",str(num_threads),\n",
    "    #                  \"--rna-sensitive\",\"-f\",\n",
    "    #                  \"-S\",out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\",\n",
    "    #                  \"-U\",out_dir+\"real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/sample_01.shuffled.fasta\"]\n",
    "    #     subprocess.call(hisat_cmd)\n",
    "    #     sort_cmd = [\"samtools\",\"sort\",\n",
    "    #                 \"-@\",str(num_threads),\n",
    "    #                 \"-o\",out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\"]\n",
    "    #     subprocess.call(sort_cmd)\n",
    "    #     strg_cmd = [\"stringtie\",\n",
    "    #                 out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 \"-p\",str(num_threads),\n",
    "    #                 \"-G\",annotation_path,\n",
    "    #                 \"-A\",out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.genes\",\n",
    "    #                 \"-o\",out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.gtf\"]\n",
    "    #     subprocess.call(strg_cmd)\n",
    "\n",
    "    #     if not os.path.exists(out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)):\n",
    "    #         os.makedirs(out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn))\n",
    "    #     hisat_cmd = [hisat_path,\n",
    "    #                  \"-x\",idx_dir+\"annotation.hisat\",\n",
    "    #                  \"-p\",str(num_threads),\n",
    "    #                  \"--rna-sensitive\",\"-f\",\n",
    "    #                  \"-S\",out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\",\n",
    "    #                  \"-U\",out_dir+\"real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/sample_01.shuffled.fasta\"]\n",
    "    #     subprocess.call(hisat_cmd)\n",
    "    #     sort_cmd = [\"samtools\",\"sort\",\n",
    "    #                 \"-@\",str(num_threads),\n",
    "    #                 \"-o\",out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\"]\n",
    "    #     subprocess.call(sort_cmd)\n",
    "    #     strg_cmd = [\"stringtie\",\n",
    "    #                 out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 \"-p\",str(num_threads),\n",
    "    #                 \"-G\",annotation_path,\n",
    "    #                 \"-A\",out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.genes\",\n",
    "    #                 \"-o\",out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.gtf\"]\n",
    "    #     subprocess.call(strg_cmd)\n",
    "\n",
    "    #     if not os.path.exists(out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)):\n",
    "    #         os.makedirs(out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn))\n",
    "    #     hisat_cmd = [hisat_path,\n",
    "    #                  \"-x\",idx_dir+\"annotation.hisat\",\n",
    "    #                  \"-p\",str(num_threads),\n",
    "    #                  \"--rna-sensitive\",\"-f\",\n",
    "    #                  \"-S\",out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\",\n",
    "    #                  \"-U\",out_dir+\"all.t\"+str(tn)+\"_s\"+str(sn)+\"/sample_01.shuffled.fasta\"]\n",
    "    #     subprocess.call(hisat_cmd)\n",
    "    #     sort_cmd = [\"samtools\",\"sort\",\n",
    "    #                 \"-@\",str(num_threads),\n",
    "    #                 \"-o\",out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\"]\n",
    "    #     subprocess.call(sort_cmd)\n",
    "    #     strg_cmd = [\"stringtie\",\n",
    "    #                 out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sorted.bam\",\n",
    "    #                 \"-p\",str(num_threads),\n",
    "    #                 \"-G\",annotation_path,\n",
    "    #                 \"-A\",out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.genes\",\n",
    "    #                 \"-o\",out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/strg2.gtf\"]\n",
    "    #     subprocess.call(strg_cmd)\n",
    "\n",
    "        os.remove(out_dir+\"strg.real.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\")\n",
    "    #     os.remove(out_dir+\"strg.real_splicing.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\")\n",
    "    #     os.remove(out_dir+\"strg.real_intronic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\")\n",
    "    #     os.remove(out_dir+\"strg.real_intergenic.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\")\n",
    "    #     os.remove(out_dir+\"strg.all.t\"+str(tn)+\"_s\"+str(sn)+\"/hisat2.sam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = []\n",
    "for tn in range(num_tissues):\n",
    "    for sn in range(num_samples):\n",
    "        samples.append((tn,sn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=================\n",
      "Tissue #0 - Sample #0\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #1\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #2\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #3\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #4\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #5\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #6\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #7\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #8\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #0 - Sample #9\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #0\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #1\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #2\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #3\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #4\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #5\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #6\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #7\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #8\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #1 - Sample #9\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #0\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #1\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #2\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #3\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #4\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #5\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #6\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #7\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #8\n",
      "=================\n",
      "\n",
      "\n",
      "=================\n",
      "Tissue #2 - Sample #9\n",
      "=================\n",
      "\n",
      "Pool: [None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None]\n"
     ]
    }
   ],
   "source": [
    "pool = multiprocessing.Pool(processes=num_processes)\n",
    "pool_outputs = pool.map(hisat2_align, samples)\n",
    "pool.close()\n",
    "pool.join()\n",
    "print('Pool:', pool_outputs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
